---
title: "Get unique plasmids and PPIs per sample"
output: html_document
---

```{r setup, include=FALSE}
# This script gets all the genes per sample called "X_chrom_data.csv"
# etc and puts them in a local folder "OUTPUT7". Per sample this shows their
# index, gloene name (Gene), plasmid status (Plasmid, where 0=plasmid-related),
# and total number of PPIs (Interactions).

# This script gets all the chromosomal genes per sample called "X_chrom_data.csv"
# etc and puts them in a local folder "OUTPUT6". Per sample this shows the
# index, the String IDs for the protein pairs, combined score, the short String
# IDs for the protein pairs and number of chromosomal PPIs (labelled "c") 

# You need the list of plasmid genes: unique_plasmid_Genes_v3.csv
# 
# You need "stringInput.csv" in your local folder. Get it from Figshare doi:
# https://doi.org/10.6084/m9.figshare.19674027.v1

# install.packages("BiocManager")
BiocManager::install("genbankr", ask=F) # v1.0.0
library(genbankr)
install.packages("igraph", ask=F) # v1.0.0
library(igraph)
BiocManager::install(c("STRINGdb"), ask=F) # , version="3.8") #  
# see https://bioconductor.org/packages/devel/bioc/vignettes/STRINGdb/inst/doc/STRINGdb.pdf
#STRINGdb$methods()              # To list all the methods available.
#STRINGdb$help("get_graph")      # To visualize their documentation.
library(STRINGdb)     # activate the STRINGdb library # eg species_id=9606 is Homo sapiens 
install.packages("VennDiagram", ask=F)
library(VennDiagram)
install.packages("rentrez", ask=F)
library(rentrez)
install.packages("tidyverse", ask=F)
library(tidyverse)
install.packages("dplyr", ask=F)
library(dplyr)
install.packages("stringr", ask=F)
library(stringr)
BiocManager::install("GenomicRanges", ask=F)
library(GenomicRanges)
install.packages("readr", ask=F)
library(readr)
install.packages("curl", ask=F)
library(curl)
install.packages("httr", ask=F)
library(httr)
options(warn=-1)
install.packages("ggrepel", ask=F)
library(ggrepel)
install.packages("data.table", ask=F)
library(data.table)
install.packages("openxlsx", ask=F)
library(openxlsx)
install.packages("taxize")
library(taxize)

set_entrez_key("eaf09dca7612332fa2595c903c24d1afef08")
Sys.getenv("ENTREZ_KEY")

# TESTING 
#plasmid <- c()
#try({plasmid <- readGenBank(GBAccession("CP021116.1"), partial=T)}, silent=T) #  ok
#plasmid <- c()
#try({plasmid <- readGenBank(GBAccession("CP020972.1"), partial=T)}, silent=T) # not ok

####### define functions #########
getplasmidgenes <- function(GBAccession_ID){ #  function 
  plasmid_genes <- c("")
  plasmid <- c()
  try({plasmid <- readGenBank(GBAccession(GBAccession_ID), partial=T)}, silent=T)
  # cds(plasmid)$product # gene long names
  # cds(plasmid)$gene    # gene short names
  if(!(is.null(plasmid))){
    plasmid_genes <- unique(sort(tolower(as.vector(na.omit( cds(plasmid)$gene))))) 
    plasmid_genes <- gsub("[()]",  replacement="", plasmid_genes) # remove ( + )
    # unique protein-coding genes all uppercase sorted #  length(plasmid_genes)
    return(tolower(plasmid_genes)) }
  else { return(tolower("-1"))} } # end else

#--- function4 --- get plasmid metadata: (1) plasmid nickname, (2) species,
# (3) strain, (4) length, (5) number of genes
#table <- setDT(table, keep.rownames = TRUE)[] #convert rowname to the first column
getdata <- function(GBAccession_ID){
  plasmid <- readGenBank(GBAccession(GBAccession_ID), partial=TRUE)
  #read in information from plasmid accession number
  w <- elementMetadata(sources(plasmid))
  #get metadata(type,organism,mol_type,strain,db_xref,plasmid,loctype,temp_grouping_id)
  name <- w$plasmid #plasmid name
  species <- w$organism #species
  species <- gsub(" ",  replacement="_", species)
  strain <- w$strain #strain
  if(is.null(strain) == TRUE){   strain <- 0 }
  genes <- elementMetadata(genes(plasmid))$gene 
  genes <- gsub("[()]",  replacement="", genes) 
  genes <- length(genes) #number of plasmid genes
  u.genes <- unique(sort(tolower(as.vector(na.omit(cds(plasmid)$gene))))) #remove redundant genes
  u.genes <- gsub("[()]",  replacement="", u.genes) 
  u.genes <- length(u.genes)    #number of unique plasmid genes
  element <- length(elementMetadata(exons(plasmid))$type)
  info <- seqinfo(plasmid)  #get plasmid information
  df_plasmid <- as.data.frame(info)
  length <- df_plasmid$seqlengths #plasmid length
  metadata <- c(name,species,strain,genes,u.genes,element,length) #combine all data
  return(metadata)
}
######## end set up functions #########

setwd("~/Google Drive/TDA_Summer2021/RMD_FILES")
input <- data.frame(read.csv("stringInput.csv")) # species, proteins, String_ID
input$Species <- gsub("/",  replacement="-", input$Species)
input$interactions <- rep(0, length(input$Species))

# https://datadryad.org/stash/dataset/doi:10.15146/R33X2J # Plasmids database
# Brooks et al 2019 doi: 10.1128/MRA.01325-18 #  complete and assembled plasmids
accessID <- data.frame(read.csv("plasmids__2020_11_19__v0.4.1/plsdb.tsv", sep = "\t"))
str(accessID$ACC_NUCCORE) # 27939 entries
PLSDB1 <- subset(accessID, select=c(ACC_NUCCORE, Description_NUCCORE)) # all are plasmids
       # "Enterobacter hormaechei strain 3804 plasmid p3804_4, complete sequence"
colnames(PLSDB1) <- c("Accession", "Sample")
str(PLSDB1) # 27939 obs: accesssion, species, family, order, lineage, strain & plasmid

# PLSDB v2020_11_19 https://ccb-microbe.cs.uni-saarland.de/plsdb/plasmids/download
accessID1 <- data.frame(read.csv("PLSDB/Metadata.csv"))
PLSDB2 <- subset(accessID1, select=c(Locus_ID, Assembly_organism))
PLSDB2 <- PLSDB2[-c(3867),] # remove one bad sample
str(PLSDB2) # 6641 plasmids with accession and organism 
colnames(PLSDB2) <- c("Accession", "Sample")
for (i in 1:length(PLSDB2$Accession)){ # add Brooks into PLSDB, which has more metadata
     if((PLSDB2$Accession[i] %in% PLSDB1$Accession)==F){ # if not known in PLSDB
          if(!(is.na(PLSDB1$Accession[i]))){
          PLSDB1[nrow(PLSDB1) + 1,] = c(PLSDB1$Accession[i], PLSDB1$Sample[i])  } }
} # end for
str(PLSDB1)  # 32838 plasmids as Accession and Sample

vector1 <- data.frame(matrix(ncol=9))
uniqueGeneList <- c()
colnames(vector1) <- c("Accession", "Genes", "Plasmid", "Sample", "Strain",
                       "UniqueGenes", "CDSs", "Length_bp", "Gene_List") 
 vector1 <- vector1[-c(1),]
for (jj in 21413:length(PLSDB1$Accession)){ # 17
     temp1 <- getplasmidgenes(PLSDB1$Accession[jj]) # genes, single vector of names
      print(jj)
      Sys.sleep(0.1) # to prevent too many requests
     if(length(temp1) > 1){ # plasmid, species,strain,genes,u.genes,element,length
        zz <- getdata(PLSDB1$Accession[jj])
        vector1[nrow(vector1)+1,] <- c(PLSDB1$Accession[jj], length(temp1), zz[1],
                    zz[2], zz[3], zz[5], zz[6], zz[7], paste(temp1,collapse=" "))
        write.xlsx(vector1 %>% distinct(), "plasmid_list_1.xlsx")
        uniqueGeneList <- c(uniqueGeneList, temp1) # add list # end if enough entries
        uniqueGeneList <- unique(sort(uniqueGeneList)) # how many genes
        write.csv(uniqueGeneList, "uniqueGeneList.csv")
        print(length(uniqueGeneList))                   # 9k ?
        print(length(unique(sort(vector1$Accession)))) # 18k unique plasmids?
     } # end if
} # end for each plasmid
str(vector1) # 18,628 plasmids potentially 
# View(vector1)
vector1$Genes <- as.numeric(vector1$Genes)
#write.xlsx(vector1 %>% distinct(), "plasmid_list_1.xlsx")
#hist(log10(vector1$Genes), breaks=66, ylab="Number of plasmids", col="red", 
#     xlab="Log10 of number of genes", main="", xlim=c(0.5,3)) + grid()
uniqueGeneList <- unique(sort(uniqueGeneList)) # how many genes
# write.csv(uniqueGeneList, "uniqueGeneList.csv")
length(uniqueGeneList)                        # 8932 
length(unique(sort(vector1$Accession))) # 16382 unique plasmids?

#######################
# plasmidGenes needs to include genes AND sample
# make table of plasmidGenes, add gene presence/absence
# make table of plasmidGenes PPIs, add PPI presence/absence
# add number of PPIs per protein
vector1  <- data.frame(read.xlsx("plasmid_list_1.xlsx", sheet=1))
dim(vector1)

# oo <- c(279, 355, 691, 818, 1108, 1153, 1211) # 1-800, 1000-1199 done
for (k in 1:4445){  #     # change k   # 5 =>18 not work # 3726
     if((k != 25)&&(k != 99)&&(k != 235)&&(k != 246)&&(k !=266)&&(k !=345)&&(k !=349)
        &&(k !=376)&&(k !=546)&&(k !=403)&&(k !=407)&&(k !=416)&&(k !=419)&&(k !=485)
        &&(k !=529)&&(k !=559)&&(k !=622)&&(k !=691)&&(k !=4377)&&(k !=4378) 
     ){  # end if
       print(k)#       Sys.sleep(0.1) # to prevent too many requests
       print(input$Species[k])
      string_db <- STRINGdb$new(version="11.5", species=input$ID[k], # new STRINGdb object
                                score_threshold=400, input_directory="")
      # string_db$proteins$preferred_name contains the list of genes, eg "DR97_1"
      mapped <- string_db$map(data.frame(gene_name = string_db$proteins$preferred_name),
                                'gene_name', removeUnmappedRows=T) 
      # mapped is a table with gene_name and STRING_id      # str(mapped) } } 
      links_all1 <- string_db$get_interactions(string_db$mp(string_db$proteins$preferred_name))
      links_all <- links_all1[!duplicated(links_all1[,c('from','to')]),]
      # str(links_all) # from: String_ID, to: string_ID, combined_score
      # links_all is a table with the STRING_ids of the 
      # connected proteins (cols 1 & 2) & their score (col 3)
      # links_all does not contain duplicates (unlike links_all1)

      # Now create a table for all genes with STRING_ids and combined score
      from_vector <- c()			# empty vector for gene names rather than STRING_ids
      to_vector <- c()				# empty vector for gene names rather than STRING_ids  
      for (i in 1:length(links_all$from)){		# for each STRING_id, extract the gene name
        from_vector[i] <- subset(mapped, STRING_id == links_all$from[i])$gene_name
        to_vector[i] <- subset(mapped, STRING_id == links_all$to[i])$gene_name }
      links_all$from_ID <- tolower(from_vector)
      links_all$to_ID <- tolower(to_vector) 
      # str(links_all) # now: from: String_ID, to: string_ID, combined_score,
        # from_ID gene, to_ID, gene 

      ######## get plasmids' gene interactions # select plasmid-associated interactions
      links_genes <- data.frame(unique(sort(c(links_all$from_ID,links_all$to_ID)))) # all genes
      links_genes$Plasmid <- rep(0,length(links_genes[,1]))
      links_genes$Interactions <- rep(0,length(links_genes[,1]))
      colnames(links_genes) <- c("Gene","Plasmid","Interactions")
      str(links_genes) # eg 7,203 genes
      #  View(data.frame(unique(tolower(mapped$gene_name))))
      
      mapped_p <- data.frame() # has from/to data # set of plasmid genes = species
      plasmidGenes <- c()
      for( hh in 1:length(vector1$Sample)){ # for each of 16,393 plasmids
       if(grepl(vector1$Sample[hh], input$Species[k], fixed=T)){ # see if names match
        plasmidGenes <- sort(c(plasmidGenes, strsplit(vector1$Gene_List[hh],"\\s+")[[1]]))}}
      plasmidGenes <- unique(plasmidGenes) # add to plasmid genes
      # str(plasmidGenes) # list of species' plasmid genes 
      
      for (m in 1:length(plasmidGenes)){	# for all plasmid genes
        mapped_p <- rbind(mapped_p, # add in mapped genes
          data.frame(subset(links_all, (to_ID==plasmidGenes[m])|(from_ID==plasmidGenes[m]))))
          if(length(subset(links_genes, Gene==plasmidGenes[m])[,1]) > 0){ # if match
             links_genes[links_genes$Gene == plasmidGenes[m],2] <- 1    }  } # end if
      str(mapped_p) # plasmid PPI list
      
     for (m1 in 1:length(links_genes$Gene)){	# for all host proteins # c = number of PPIs
       counter=0 
       for (eee in links_all$to_ID){ 
              if(eee == links_genes$Gene[m1]){ counter = counter + 1 } }
      for (eee in links_all$from_ID){ 
              if(eee == links_genes$Gene[m1]){ counter = counter + 1 } }    
       links_genes$Interactions[m1] <- counter  } # count the PPIs # issue
          # grepl on angn 
      #  length(subset(links_all, grepl(links_genes$Gene[m1], to_ID, fixed=T))$to_ID) + 
       # length(subset(links_all, grepl(links_genes$Gene[m1], from_ID, fixed=T))$from_ID) }
     str(links_genes)
     
      # pdf("Serratia_intra_subnetwork.pdf", width=16, height=20)
     # pdf("Staph_eq_intra_subnetwork.2.pdf", width=16, height=20)
     # string_db$plot_network(string_db$mp(plasmidGenes)) # make plot 
     # string_db$plot_network(string_db$mp(unique(c(mapped_p$from_ID, mapped_p$to_ID))))
     # dev.off()   # make a plot of the network for fun
          
     # links_genes is the issue 
     write.csv(links_all, paste("OUTPUT6/", input$Species[k], "_chrom_data.csv", sep=""))
     write.csv(links_genes, paste("OUTPUT7/", input$Species[k], "_genes_data.csv", sep=""))
  } # end if k is ok
} # 4244:4499 ongoing
# end for all species

# add plasmid status from OUTPUT6 files to OUTPUT7 files
setwd("~/Google Drive/TDA_Summer2021/RMD_FILES/OUTPUT6/")
names2 <- list.files(pattern = "\\.csv$")  # 4414 samples
str(names2)
for (n2 in 1:1){ # 3333:length(names2)){ # n2 <- 1568 # E. coli  
     print(n2)
     name <- names2[n2]  # 4393 samples
   #       print(name)
     setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/RMD_FILES/OUTPUT6/")
     try ( in2 <- data.frame(read.csv(name)) ) # Escherichia_coli_CFT073__chrom_data.csv
     in2 <- in2[,-c(1)] # from, to, score, from_ID, to_ID, c
     in2$c <- c(0)
     #name = gsub("\\.(?=[^.]*\\.)", "", name, perl=T)
     #name = gsub("\\:", "", name, perl=T)
     name = gsub("_chrom_data.csv", "_genes_data.csv", name, perl=T) # genes_data
     setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/RMD_FILES/OUTPUT7/")
     ingenes <- 0
     try ( ingenes <- data.frame(read.csv( name)) )
  #   print(name)
     ingenes <- ingenes[,-c(1)]  
     str(ingenes) # Gene Plasmid Interactions
     if( sum(ingenes$Plasmid) >0){  # check if any plasmid genes
        for (n4 in 1:length(in2$from_ID)){ # n4 <- 13933 = aaea
          v1 <- 0
          v2 <- 0
          if(length(subset(ingenes, grepl(in2$to_ID[n4], Gene, fixed=T))$Plasmid)==0){ v1=0
          } else { v1 <- subset(ingenes, grepl(in2$to_ID[n4], Gene, fixed=T))$Plasmid }
          if(length(subset(ingenes, grepl(in2$from_ID[n4], Gene, fixed=T))$Plasmid)==0){ v2=0
          } else { v2 <- subset(ingenes, grepl(in2$from_ID[n4], Gene, fixed=T))$Plasmid }
          in2$c[n4] <- v1 + v2     } 
       } # end check
     name = gsub("_genes_data.csv", "_chrom_data.csv", name, perl=T)
   #  View(in2)
  #   View(ingenes)
     write.csv(in2, paste("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/RMD_FILES/OUTPUT8/",
                          name, sep=""))
     # write.csv(in2, paste("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/RMD_FILES/OUTPUT6/", name, sep=""))
} # end for

# install.packages("dplyr")
library(dplyr)
samplenames <- data.frame(read.csv("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/RMD_FILES/OLDFILES/sample_names_497.csv"))
setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/RMD_FILES/OUTPUT8/") # ?
names <- list.files(pattern = "\\.csv$")  # 4393 samples
for (n2 in 1:length(names)){ # n2 <- 1568 # E. coli  
     print(n2)
     name <- names[n2]  # 4393 samples # "Actinoplanes_sp._N902-109__chrom_data.csv"
     try ( ingenes <- data.frame(read.csv(name)) )
     name = gsub("_chrom_data.csv", "", name, perl=T)
     if(name %in% samplenames){ # add _comma_ in gene names to replace   
        ingenes <- ingenes[,-c(1)]  # str(ingenes) # Gene Plasmid Interactions
        # View(ingenes)
        ingenes[1,]
        print(n2)
        print(name)
        ingenes$from_ID = str_replace(ingenes$from_ID, ",", "_comma_")
        ingenes$to_ID = str_replace(ingenes$to_ID, ",", "_comma_")
     } # end if sample in list
} # end for files

h <- list() # give a name to the hash
hint <- list() # give a name to the hash
genelist <- c()
samplelist <- c()
setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/RMD_FILES/OUTPUT7/") # ??
names <- unique(list.files(pattern = "\\.csv$"))  # 4414 samples
str(names)
for (n2 in 1:length(names)){ # n2 <- 1568 # E. coli  
     name <- names[n2]  # 4393 samples # "Actinoplanes_sp._N902-109__chrom_data.csv"
     print(n2)
     try ( ingenes <- data.frame(read.csv(name)) )
     name = gsub("_genes_data.csv", "", name, perl=T)
     ingenes <- ingenes[,-c(1)]  # str(ingenes) # Gene Plasmid Interactions
     ingenes$Plasmid <- as.numeric(ingenes$Plasmid)
     ingenes$Interactions <- as.numeric(ingenes$Interactions)
      # str(ingenes) # Gene Plasmid Interactions
     # print(length(subset(ingenes, Plasmid>0)$Gene) ) 
     if(length(subset(ingenes, Plasmid>0)$Gene) > 0){ # if genes
       htemp <- list() 
       htempint <- list()
       ingenes2 <- subset(ingenes, Plasmid>0)
       samplelist <- c(samplelist, name)
       for (i1 in 1:length(ingenes2$Gene)){
         genelist <- c(genelist, ingenes2$Gene[i1])
         htemp[[ingenes2$Gene[i1]]] <- 1
         htempint[[ingenes2$Gene[i1]]] <- ingenes2$Interactions[i1] }
       h[[name]] <- htemp
       hint[[name]] <- htempint
       } # end if genes
} # end for files

genelist <- unique(sort(genelist))
str(genelist) # 3023 genes
str(samplelist) # 497 samples
finaltable <- data.frame(matrix(nrow=length(samplelist), ncol=length(genelist)))
colnames(finaltable) <- genelist #  
rownames(finaltable) <- samplelist #   
str(finaltable)
finaltable_int <- data.frame(matrix(nrow=length(samplelist), ncol=length(genelist)))
colnames(finaltable_int) <- genelist
rownames(finaltable_int) <- samplelist

for (m1 in 1:length(samplelist)){ # for each sample
   geneset <- rep(0,length(genelist)) # assume all absent
   int_set <- rep(0,length(genelist)) # assume all absent interactions
   templist <- h[[samplelist[m1]]] # access sample -> list of genes
   templist2 <- hint[[samplelist[m1]]] # access sample -> list of genes
   for (m2 in 1:length(genelist)){ #   for each gene
        if(genelist[m2] %in% names(templist)){  if(templist[[genelist[m2]]]>0){ 
             geneset[m2] <- templist[[genelist[m2]]]# add as plasmid-encoded
             int_set[m2] <- templist2[[genelist[m2]]]# add as plasmid-encoded
        } } # end if valid plasmid-encoded gene
   } # end for each gene   
   finaltable[m1,] <- geneset
   finaltable_int[m1,] <- int_set
} # end for each sample  # "numbers" %in% names(my_list) 

finaltable <- data.frame(finaltable)
rownames(finaltable) 
finaltable_int <- data.frame(finaltable_int)
dim(finaltable) # 364 samples x 2921 genes
dim(finaltable_int)  #  364 samples x 2921 genes
#View( (finaltable))
#View( (finaltable_int))
finaltable[1:3,1:4]
write.csv(finaltable, "plasmidGenes.new.csv")
write.csv(finaltable_int, "plasmidGenes.new.int.csv")

# get all genes
```

