---
title: "Plasmid-related gene & PPI main analysis"
---

```{r setup, include=FALSE}
# Potential input files [1] - [16] read from the local folder
# There are 17 listed, only number [1] is essential
# 
# all_PPIs_4377.tar -> 4,377 samples' CSV files for all proteins showing the
# index, gene name (Gene), plasmid status (Plasmid, where 0=plasmid-related),
# and total number of PPIs (Interactions) - for each sample
# 
# [1]  stringInput.csv = the species, proteins & String_ID - get it from
#      Figshare doi: https://doi.org/10.6084/m9.figshare.19674027.v1
# [2]  plasmidGenesTable.i.csv for i=0 to i=23 inclusive, containing the matrix 
#      of all 9,172 plasmid-related genes across the 4,445 samples - don't try
#      to run this unless you have a long time to wait for it to work - each has
#      partial data that is amalgamated together for the final correct matrix
#      because it was too difficult to run it all together, try with caution
# [3]  plasmidGenesTable2.csv = the final version of [2]
# [4]  GeneName_presence.csv = 4,445 samples with their, stringID, CorrectNames,
#      and LongNames to test for gene name length
# [5]  Acinetobacter.xlsx = example data to test the gene name comparison
# [6]  Gene_rates_4393.csv = rates of plasmid & chrom genes for 4,393 samples -
#      this is not essential, the code with generate this.
# [7]  Plasmid_PPI.xlsx = 4,377 samples' plasmid & chromosomal PPI rates
# [8]  Plasmid_PPI_data_PMNLE.csv = 4,377 samples with their String ID,
#      chromosomal aPMNLE, chromosomal cThreshold, chromosomal cTriangles, 
#      chromosomal cEdges, chromosomal cVertices, chromosomal cNonTrivialLoops,
#      and chromosomal cConnectedComponents
# [9]  approximatePMNLE_purely_chromosomal.csv = like [8] but with 4,445 samples
#      shown where some contain no relevant data (chromosomal PPIs)
# [10] approximatePMNLE_fullPPIN.csv = like [8] but with 4,422 samples shown
#      where some contain no relevant data (for all PPIs)
# [11] Table_S3.xlsx = 4,377 samples with the plasmid-related 
#      PPIs (Plasmid_PPI), chromosomal PPIs (Chrom_PPI), fraction of PPIs that 
#      were plasmid-related (Fraction), numbers of proteins per sample
#      (inputProteins), number of PPIs (interactions), number of plasmid genes
#      (Plasmid_genes), chromosomal aPMNLE (aPMNLE), chromosomal trios (cTri),
#      chromosomal indirect loops (loops), chromosomal connected components (ccs),
#      all proteins aPMNLE (aPMNLE_all), all proteins trios (cTri_all), all 
#      proteins indirect loops (loops_all), and all proteins connected components
#      (ccs_all)
# [12] OUTPUT7 -> points to chromosomal_PPIs_4377.tar.gz -> 4,377 samples' CSV
#      files for all proteins showing the index, the String IDs for the protein
#      pairs, combined score, the short String IDs for the protein pairs and
#      number of chromosomal PPIs (c) - for each sample - see it at 
#   https://drive.google.com/file/d/1gTOucAvkuqDaS14ah7Xzt4Q_CcWdMV4C/view?usp=sharing
# [13] Data_for_Fig_S7_metasamples.csv = metasamples from NMF on plasmid gene
#      presence-absence matrix
# [14] Data_for_Fig_S11_metasamples.interactions.all.csv = mixture gene
#      coefficients from NMF on plasmid gene presence-absence matrix
# [15] Data_for_Fig_S8_mixture_coefficients.interactions.all.csv = metasamples
#      from NMF on plasmid gene PPI rates matrix
# [16] Data_for_Fig_S12_mixture_coefficients.interactions.all.csv = mixture 
#      gene coefficients from NMF on plasmid gene PPI rates matrix


# install.packages("genbankr", ask=F) # v1.0.0
library(genbankr)
# install.packages("igraph", ask=F) # v1.0.0
library(igraph)
# BiocManager::install(c("STRINGdb"), ask=F) # , version="3.8") #  
# see https://bioconductor.org/packages/devel/bioc/vignettes/STRINGdb/inst/doc/STRINGdb.pdf
#STRINGdb$methods()              # To list all the methods available.
#STRINGdb$help("get_graph")      # To visualize their documentation.
library(STRINGdb)     # activate the STRINGdb library # eg species_id=9606 is Homo sapiens 
#install.packages("VennDiagram", ask=F)
library(VennDiagram)
#install.packages("rentrez", ask=F)
library(rentrez)
#install.packages("tidyverse", ask=F)
library(tidyverse)
# install.packages("dplyr", ask=F)
library(dplyr)
# install.packages("stringr", ask=F)
library(stringr)
#BiocManager::install("GenomicRanges", ask=F)
library(GenomicRanges)
#install.packages("readr", ask=F)
library(readr)
# install.packages("curl", ask=F)
library(curl)
# install.packages("httr", ask=F)
library(httr)
# install.packages("openxlsx")
library(openxlsx)
library(ggrepel)
options(warn=-1)
# READ IN K AS USER INPUT
library(dplyr)
library(ggpubr)
#install.packages("hash")
library(hash)
library(factoextra)

set_entrez_key("eaf09dca7612332fa2595c903c24d1afef08")
Sys.getenv("ENTREZ_KEY")

####### define functions #########

# Multiple plot function
multiplot <- function(..., plotlist=NULL, file, cols=1, layout=NULL) {
# Make a list from the ... arguments and plotlist
  plots <- c(list(...), plotlist)
  numPlots = length(plots)
  # If layout is NULL, then use 'cols' to determine layout
  if (is.null(layout)) { # Make the panel
  # ncol: Number of columns of plots
  # nrow: Number of rows needed, calculated from # of cols
  layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
  ncol = cols, nrow = ceiling(numPlots/cols))}

    if (numPlots==1) { print(plots[[1]]) }
    else { # Set up the page
      grid.newpage()
       pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))
      # Make each plot, in the correct location
      for (i in 1:numPlots) { # Get the i,j matrix positions of the regions that contain this subplot
        matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))
        print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,layout.pos.col = matchidx$col)) }
     }
}

getplasmidgenes <- function(GBAccession_ID){ #  function 
  plasmid_genes <- c("")
  plasmid <- c()
  try({plasmid <- readGenBank(GBAccession(GBAccession_ID), partial=T)}, silent=T)
  # cds(plasmid)$product # gene long names
  # cds(plasmid)$gene    # gene short names
  if(!(is.null(plasmid))){
    plasmid_genes <- unique(sort(tolower(as.vector(na.omit( cds(plasmid)$gene))))) 
    plasmid_genes <- gsub("[()]",  replacement="", plasmid_genes) # remove ( + )
    # unique protein-coding genes all uppercase sorted #  length(plasmid_genes)
    return(tolower(plasmid_genes)) }
  else { return(tolower("-1"))} } # end else
######## end set up functions #########

########################################################################
########### Import gene names #### #### #### #### #### #### #### #### ####
########################################################################

setwd("~/Google Drive/TDA_Summer2021/RMD_FILES")
input <- data.frame(read.csv("stringInput.csv")) # species, proteins, String_ID
input$Species <- gsub("/",  replacement="-", input$Species)
input$interactions <- rep(0, length(input$Species))

########################################################################
########### Create sample-gene matrix for all plasmid-related genes ####
########################################################################
# make PPI data across samples, merging from different files yay
# on server /media/newdrive/tim/PPIN_files ###
# need to split files up to run jobs, putting them back together again
# humpty dumpty style as a series of partially filled matrices

########################################################################
########### Load plasmid-related genes table  ########### ####
########################################################################

# Need to make new table # real plasmid-associated genes vs samples
vector1  <- data.frame(read.xlsx("plasmid_list_1.xlsx"))
vector1$Length_bp <- as.numeric(vector1$Length_bp)
vector1$Genes <- as.numeric(vector1$Genes)
vector1$CDSs <- as.numeric(vector1$CDSs)
median(vector1$Genes)
str(vector1) # Accession	Genes	Plasmid	Sample	Strain	UniqueGenes	CDSs	
             # Length_bp  Gene_List
# LR883052.1	65	3	Escherichia_coli	0	65	98	82909 ampr arsc assf ...
hist( (vector1$Genes), breaks=99, ylab="Number of plasmids", col="red", 
     xlab="Number of genes", main="", xlim=c(0,400)) + grid()

hist.data = hist(vector1$Genes, plot=F, breaks=99)
hist.data$counts = log10(hist.data$counts)
plot(hist.data, ylim=c(0,4), ylab="Log10-scaled number of plasmids",
     xlim=c(0,320), xlab="Number of genes", col="red", main="") + grid()

vector2 <- data.frame(matrix(ncol=2)) # make tidy data of samples x genes
colnames(vector2) = c("Sample", "Gene")
for(hh in 1:length(vector1$Sample)){ # for each plasmid
   genes1 <- sort(strsplit(vector1$Gene_List[hh],"\\s+")[[1]])
   for (ee in 1:length(genes1)){
      vector2 <- rbind(vector2, c(vector1$Sample[hh], genes1[ee])) }
} # end for plasmids
vector2 <- vector2[-c(1),]
vector2 <- data.frame(distinct(vector2))
str(vector2) # very loooong 
#View(vector2)
 
for (i in 1:length(rownames(new))){ # for each sample
  new[i,] <- 0 # assign as zeros
  for(hh in 1:length(vector2$Sample)){ # for each sample x gene
    if(grepl(vector2$Sample[hh], rownames(new)[i], fixed=T)){ # match
      # print(rownames(new)[i])
       for (i2 in 1:length(colnames(new))){  # for each gene
          if(colnames(new)[i2] == vector2$Gene[hh]){ new[i,i2] <- 1} # end if
       } # end for matching genes
    } # end if match
  } # end for plasmids
} # end for samples
yy <- t(new[1211,])
length(yy[yy>0]) # genes on plasmids 
#View(data.frame(yy)) # 94 

# remove empty columns from table
# get all unique genes
setwd("~/Google Drive/TDA_Summer2021/RMD_FILES")
uniqueGeneList  <- data.frame(read.csv("uniqueGeneList.csv"))
uniqueGeneList <- uniqueGeneList[,-c(1)]
str(uniqueGeneList) # 16393 obs. of  9 variables

# read in data after having made it already above and exported it
setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/RMD_FILES")
new <- data.frame(read.csv("plasmidGenes.new.csv"))  # not useful anymore
rownames(new) <- new[,1]
new <- new[-c(1)] # drop unneeded column
dim(new) # 494 obs. of  3023 variables:

hist(log10 (rowSums(new)), breaks=99) + grid()
hist(log10 (colSums(new)), breaks=99) + grid()
median(colSums(new)) # 3 median number of samples in which each gene is found
IQR(colSums(new))    # 4 IQR of number of samples in which each gene is found
median(rowSums(new)) # 5 median genes per sample
IQR(rowSums(new))    # 12 IQR of number of genes per sample
summary(rowSums(new)) # mean = 27.04 +- 112.0
sd(rowSums(new)) 
summary(colSums(new)) #  
sd(colSums(new)) # mean = 4.42 +- 5.75
View(new)

# interactions
new_int <- data.frame(read.csv("plasmidGenes.new.int.csv"))  # not useful anymore
rownames(new_int) <- new_int[,1]
new_int <- new_int[-c(1)] # drop unneeded column
dim(new_int) # 494 obs. of  3023 variables:
median(colSums(new_int)) # 86 median number of PPIs per gene  
IQR(colSums(new_int))    # 190 IQR of number of PPIs per gene  
summary(colSums(new_int)) # mean = 346.5 +- 1,273.5
sd(colSums(new_int)) 

########################################################################
########### Gene name analysis #########################################
########################################################################

#create an empty table
#table <- data.frame(k=integer(), species=character(), StringID=character(),
#                    CorrectNames=character(), LongNames=character())
setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/RMD_FILES")
newset <- c()
for (jk in 1:length(rownames(new))){
  newset <- c(newset, subset(input, Species==rownames(new)[jk])$ID) }
str(newset)

tablen <- data.frame(k=integer(), Species=character(), ID=integer(), 
                     four=integer(), long=integer())
###########
for (k in 1:685){     # change k  1:4445
    print(k)
      k <- newset[k]
       print(k) 
      string_db <- STRINGdb$new(version="11.5", species=k, # new STRINGdb object
                                score_threshold=400, input_directory="")
      # string_db$proteins$preferred_name contains the list of genes, eg "DR97_1"
      mapped <- string_db$map(data.frame(gene_name = string_db$proteins$preferred_name),
                                'gene_name', removeUnmappedRows=T) 
      mapped$gene_name <- str_to_lower(mapped$gene_name)
      str(mapped$gene_name)
      removePat <- "[_-].{1}$" # remove weird digits
      mapped$gene_name1 <- gsub(removePat, "", mapped$gene_name)
      four=0;
      long=0;
      for (i in 1:length(mapped$gene_name)){
          if(nchar(mapped$gene_name[i]) > 4){ long = long+1 }} # if > 4 letters
      four = length(mapped$gene_name) - long
      tablen[nrow(tablen)+1,] <- c(k, input$Species[k], input$ID[k], four, long)  
} # end for all species
###
View(tablen)
# Check gene names lengths - most are short
write.csv(tablen, "GeneName_presence.csv") 
#tablen <- data.frame(read.csv( "GeneName_presence.csv"))
str(tablen) # 4445 samples x species, stringID, CorrectNames, LongNames
tablen_old <- tablen
tablen <- subset(tablen_old, four>310)
tablen <- tablen[,-c(1)]
str(tablen) # now  samples
tablen$four <- as.numeric(tablen$four)
hist(tablen$four, breaks=42) + grid()

#table1_old <- table1_old[-c(1,2)]
str(na.omit(tablen_old))
summary(na.omit(tablen_old))
sd(na.omit(tablen_old$four))
sd(na.omit(tablen_old$long))

pdf("Gene_Name_Matches2.pdf", width=6, height=11, bg="white")
par(mfrow=c(2,1))
plot(log10(tablen$long), log10(tablen$four),
     pch=".", col="red",   xlab="Log10 of number of long gene names",
     ylab="Log10 of number of short gene names") + grid()
plot((tablen$long), (tablen$four),
     pch=".", col="blue", xlab="Number of long gene names",
     ylab="Number of short gene names") + grid()
dev.off()

pdf("Gene_Name_Matches2.pdf", width=6, height=6, bg="white")
par(mfrow=c(1,1))
plot((tablen$long), (tablen$four),
     pch=".", col="blue", xlab="Number of long gene names",
     ylab="Number of short gene names") + grid()
dev.off()

summary(na.omit(tablen_old$four))
#    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#        0.0     0.0   411.0   428.7   574.0  4075.0 
summary(na.omit(tablen_old$long))
#    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#       4    1935    2931    3203    4046   11207 

write.csv(table1, "Valid_samples_Gene_Names.csv")

############## OLD ######################################
# check gene names in Acinetobacter only
ex1 <- read.xlsx("~/GoogleDrive/TDA_Summer2021/Key_Results_Figures/OLDFILES/Acinetobacter.xlsx")
# str(ex1)
 
pdf("OLDFILES/Acinetobacter_check.pdf", width=6, height=11, bg="white")
par(mfrow=c(2,1))
plot(log10(ex1$CorrectNames), log10(ex1$Plasmid_PPI), col="red") + grid()
cor.test(log10(ex1$CorrectNames), log10(ex1$Plasmid_PPI))
plot((ex1$Fraction_CorrectNames), (ex1$Fraction_PlasmidPPIs), xlim=c(0,1), col="blue") + grid()
cor.test((ex1$Fraction_CorrectNames), (ex1$Fraction_PlasmidPPIs))
dev.off()

################################################################
########################################################################
########### Amalgamate summary data into Table S3 ################# ####
########################################################################

# OUTPUT7 # "","Gene","Plasmid","Interactions"
# "1","acee",1,41

setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/RMD_FILES/OUTPUT7")
# make correct numbers of plasmid-related genes per sample in "Gene_rates"
# input = Gene	Plasmid	Interactions
# need Gene_rates: Sample	=> Plasmid_genes
Gene_rates <- as.data.frame(matrix(0, ncol=3, nrow =length(list.files(pattern = "\\.csv$"))))
colnames(Gene_rates) <- c("Sample","Plasmid_genes", "Chrom_genes")
str(Gene_rates) # 4429 samples 
names3 <- list.files(pattern = "\\.csv$")
length(names3) # 4429 samples
for (k in 1:length(names3)){ # read in 4429 files
     name <- names3[k]
     try ( in2 <- data.frame(read.csv(name)) )
     in2 <- in2[,-c(1)]
     name = gsub("\\.(?=[^.]*\\.)", "", name, perl=T)
     name = gsub("\\:", "", name, perl=T)
     name = gsub("_genes_data.csv", "", name, perl=T)
     print(k) #     str(in2)
     Gene_rates$Sample[k] <- name
     Gene_rates$Plasmid_genes[k] <- length(subset(in2, Plasmid==1)$Gene)
     Gene_rates$Chrom_genes[k] <- length(subset(in2, Plasmid==0)$Gene)
} # end for
View(Gene_rates)
write.csv(Gene_rates, "../Gene_rates_new.csv")
setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/RMD_FILES")
# Gene_rates <- read.csv("../Gene_rates_4393.csv")
# Gene_rates <- Gene_rates[,-c(1)]
str(Gene_rates) # 4429 samples 
summary(Gene_rates$Plasmid_genes/Gene_rates$Chrom_genes) # mean = 0.0812%
sd(Gene_rates$Plasmid_genes/Gene_rates$Chrom_genes)  # SD = 1.12%
summary(Gene_rates$Chrom_genes)  # median = 3,333, mean =3,540
sd(Gene_rates$Chrom_genes) # SD=1,656.0
summary(Gene_rates$Plasmid_genes)  # mean = 3.02 median=0
sd(Gene_rates$Plasmid_genes)  # SD = 38.3

hist(subset(Gene_rates, Plasmid_genes>0)$Plasmid_genes, breaks=99) + grid()
hist(log10(Gene_rates$Chrom_genes), breaks=69) + grid()
plot(Gene_rates$Plasmid_genes, Gene_rates$Chrom_genes, cex=0.1) + grid()

length(subset(Gene_rates, Plasmid_genes>0)$Sample) # 497
# write.csv(subset(Gene_rates, Plasmid_genes>0)$Sample, "sample_names_497.csv")

Gene_rates <- data.frame(read.csv("Gene_rates_new.csv"))
Gene_rates <- Gene_rates[,-c(1)]
str(Gene_rates) # Sample, Plasmid_genes, Chrom_genes

sample_names_497 <- data.frame(read.csv("sample_names_497.csv"))
sample_names_497 <- sample_names_497[,-c(1)]
str(sample_names_497) # 497 genes with plasmid genes

setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/RMD_FILES")
# Plasmid_PPI_data <- read.xlsx("Plasmid_PPI.xlsx") # , sheetIndex = 1)
Plasmid_PPI_data <- Gene_rates
Final_Table  <- Gene_rates

#################################################################
########################################################################
########### Examine aPMNLE for all proteins, not too useful ####### ####
########################################################################

# Full aPMNLE
setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/")
fullPMNLE <- data.frame(read.csv("Plasmid_PPI_data_PMNLE.csv"))
fullPMNLE <- fullPMNLE[,-c(1)]
#fullPMNLE$inputProteins <- (Gene_rates$Plasmid_genes + Gene_rates$Chrom_genes)
colnames(fullPMNLE)[1] <- "STRING.bacteria.name"
str(fullPMNLE) # 4363 obs. of  16 variables
#fullPMNLE$Plasmid_Plasmid_PPI <- c(0)
#fullPMNLE$Plasmid_PPI <- c(0)
#fullPMNLE$Chrom_PPI <- c(0)
#fullPMNLE$interactions <- c(0)
# str(Gene_rates) # Sample Plasmid_genes Chrom_genes 
#fullPMNLE$Plasmid_genes <- c(0) # burkholderia_sp -> no data
#fullPMNLE$Chrom_genes <- c(0) # Actinoplanes_sp_SE50-110 -> no data

for (k in 1:length(fullPMNLE$STRING.bacteria.name)){
  fullPMNLE$STRING.bacteria.name[k]= gsub("_equals_", "_",
                                fullPMNLE$STRING.bacteria.name[k], perl=T) } # end for

fullPMNLE$STRING.bacteria.name[4] = "Escherichia_coli_O157H7_str_EDL933"
fullPMNLE$STRING.bacteria.name[396] = "Staphylococcus_caprae_M23864W1"
fullPMNLE$STRING.bacteria.name[3232] = "Pseudomonas_sp_URMO17WK12I8"
fullPMNLE[1843,16] <- 1681
fullPMNLE[1843,2] <- 1681
fullPMNLE[2938,16] <- 2202
fullPMNLE[2938,2] <- 2202
fullPMNLE[2939,16] <- 1739
fullPMNLE[2939,2] <- 1739
fullPMNLE[3232,16] <- 4970
fullPMNLE[3232,2] <- 4970
fullPMNLE[396,15] <- 10
fullPMNLE[396,16] <- 2490
fullPMNLE[396,2] <- 2500

# add fullPMNLE: inputProteins, Plasmid_genes, Chrom_genes
for (k in 1:length(fullPMNLE$STRING.bacteria.name)){
  if(fullPMNLE$Chrom_genes[k] <1){
  for (kk in 1:length(Gene_rates$Sample)){
     Gene_rates$Sample[kk] = gsub("\\.(?=[^.]*\\.)", "", Gene_rates$Sample[kk], perl=T)
     Gene_rates$Sample[kk] = gsub("\\:", "", Gene_rates$Sample[kk], perl=T)
     Gene_rates$Sample[kk] = gsub("\\[|\\]", "", Gene_rates$Sample[kk], perl=T)
      Gene_rates$Sample[kk] = gsub("\'", "", Gene_rates$Sample[kk], perl=T)
    Gene_rates$Sample[kk] = gsub('_{1}$', '', Gene_rates$Sample[kk], perl=T)
    if(fullPMNLE$STRING.bacteria.name[k] == Gene_rates$Sample[kk]){
       fullPMNLE$Plasmid_genes[k] <- Gene_rates$Plasmid_genes[kk]
       fullPMNLE$Chrom_genes[k] <- Gene_rates$Chrom_genes[kk]
       fullPMNLE$inputProteins[k] <- (Gene_rates$Plasmid_genes[kk] + 
                              Gene_rates$Chrom_genes[kk])} } # end if # end for
   print(fullPMNLE$Chrom_genes[k])
      print(fullPMNLE$STRING.bacteria.name[k]) }
} # for each entry
length(subset(fullPMNLE, Plasmid_genes >0)$Plasmid_genes)
subset(fullPMNLE, Chrom_genes <1)$Plasmid_genes
str(fullPMNLE)
#write.csv(fullPMNLE, "fullPMNLE2.csv")
#fullPMNLE2 <- data.frame(read.csv("fullPMNLE2.csv"))
#str(fullPMNLE2)

# add fullPMNLE: Plasmid_Plasmid_PPI, Plasmid_PPI, Chrom_PPI, interactions
setwd("~/Desktop/OUTPUT8/")
names4 <- list.files(pattern="_chrom_data.csv")
str(names4) # 4418
for (k in 1:length(names4)){ 
     print(k)
     name <- names4[k]
     try ( in2 <- data.frame(read.csv(name)) )
     in2 <- in2[,-c(1)]
     name = gsub("\\.(?=[^.]*\\.)", "", name, perl=T)
     name = gsub("\\:", "", name, perl=T)
     name = gsub("_chrom_data.csv", "", name, perl=T)
     name = gsub("\\[|\\]", "", name, perl=T)
     name = gsub('_{1}$', '', name, perl=T)
     name = gsub("_=_", "_", name, perl=T)
     name = gsub("\'", "", name, perl=T)
     name = gsub("_equals_", "_",  name, perl=T) 
     print(name)
     for (kk in 1:length(fullPMNLE$STRING.bacteria.name)){
       if(fullPMNLE$Plasmid_PPI[kk] < 1){ 
       if(name == fullPMNLE$STRING.bacteria.name[kk]){
         fullPMNLE$Plasmid_Plasmid_PPI[kk] <- length(subset(in2, c==2)$c)
         fullPMNLE$Plasmid_PPI[kk] <- length(subset(in2, c==1)$c)
         fullPMNLE$Chrom_PPI[kk] <- length(subset(in2, c==0)$c)  
         fullPMNLE$interactions[kk] <- (fullPMNLE$Chrom_PPI[kk] + 
         fullPMNLE$Plasmid_Plasmid_PPI[kk] + fullPMNLE$Plasmid_PPI[kk])
         print(fullPMNLE[kk,])  
          } }  } # end if # end for
} # end for
View(fullPMNLE)

 (subset(fullPMNLE, Chrom_PPI <1)) # For missing ones with equals
str(fullPMNLE) # 4363
length(subset(fullPMNLE, Plasmid_genes >0)$Plasmid_genes) # 491
length(subset(fullPMNLE, Plasmid_genes >0 & fullPMNLE$Plasmid_PPI<1)$Plasmid_genes) # 2

setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/")
BettiNumber <- data.frame(read.csv("resultsOnFirstBettiNumber.csv"))
str(BettiNumber) # Sample cTri_all	loop_all	ccs_all

for (kk in 1:length(BettiNumber$Sample)){
     BettiNumber$Sample[kk] = gsub("_chrom_data/resultsOnFirstBettiNumber.csv",
                                  "", BettiNumber$Sample[kk], perl=T)
     BettiNumber$Sample[kk] = gsub("\\.(?=[^.]*\\.)", "", BettiNumber$Sample[kk], perl=T)
     BettiNumber$Sample[kk] = gsub("\\:", "",     BettiNumber$Sample[kk], perl=T)
     BettiNumber$Sample[kk] = gsub("\\[|\\]", "", BettiNumber$Sample[kk], perl=T)
     BettiNumber$Sample[kk] = gsub("\'", "",      BettiNumber$Sample[kk], perl=T)
     BettiNumber$Sample[kk] = gsub("\\.", "",      BettiNumber$Sample[kk], perl=T)
     BettiNumber$Sample[kk] = gsub('_{1}$', '',   BettiNumber$Sample[kk], perl=T) }
View(BettiNumber) # Sample cTri_all	loop_all	ccs_all

setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/")
sample_names_497 <- data.frame(read.csv("sample_names_497.csv"))
#sample_names_497 <- sample_names_497[,-c(1)]
str(sample_names_497) # 497 genes with plasmid genes

for (kk in 1:length(sample_names_497$Sample)){
     sample_names_497$Sample[kk] = gsub("\\.(?=[^.]*\\.)", "", sample_names_497$Sample[kk], perl=T)
     sample_names_497$Sample[kk] = gsub("\\:", "",     sample_names_497$Sample[kk], perl=T)
     sample_names_497$Sample[kk] = gsub("\\[|\\]", "", sample_names_497$Sample[kk], perl=T)
     sample_names_497$Sample[kk] = gsub("\'", "",      sample_names_497$Sample[kk], perl=T)
     sample_names_497$Sample[kk] = gsub("\\.", "",      sample_names_497$Sample[kk], perl=T)
     sample_names_497$Sample[kk] = gsub('_{1}$', '',   sample_names_497$Sample[kk], perl=T) }

setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/RMD_FILES/")
approximatePMNLE <- data.frame(read.csv("approximatePMNLE.290622.csv"))
approximatePMNLE <- approximatePMNLE[,-c(1)]
str(approximatePMNLE) # STRING.bacteria.name , aPMNLE for all genes
for (j in 1:length(approximatePMNLE$STRING.bacteria.name)){
     approximatePMNLE$STRING.bacteria.name[j] = gsub("\\.(?=[^.]*\\.)", "", approximatePMNLE$STRING.bacteria.name[j] , perl=T)
     approximatePMNLE$STRING.bacteria.name[j]  = gsub("\\:", "", approximatePMNLE$STRING.bacteria.name[j] , perl=T)
     approximatePMNLE$STRING.bacteria.name[j]  = gsub("_chrom_data.csv", "", approximatePMNLE$STRING.bacteria.name[j] , perl=T)
     approximatePMNLE$STRING.bacteria.name[j]  = gsub("\\[|\\]", "", approximatePMNLE$STRING.bacteria.name[j] , perl=T)
   approximatePMNLE$STRING.bacteria.name[j]  = gsub("\\.", "", approximatePMNLE$STRING.bacteria.name[j] , perl=T)

     approximatePMNLE$STRING.bacteria.name[j]  = gsub('_{1}$', '', approximatePMNLE$STRING.bacteria.name[j] , perl=T)
     approximatePMNLE$STRING.bacteria.name[j]  = gsub("_=_", "_", approximatePMNLE$STRING.bacteria.name[j] , perl=T)
     approximatePMNLE$STRING.bacteria.name[j]  = gsub("\'", "", approximatePMNLE$STRING.bacteria.name[j] , perl=T)
     approximatePMNLE$STRING.bacteria.name[j]  = gsub("_equals_", "_",
                                approximatePMNLE$STRING.bacteria.name[j] , perl=T)  }
     
fullPMNLE$Entry <- c(0)
for (k in 1:length(fullPMNLE$STRING.bacteria.name)){
  print(fullPMNLE$STRING.bacteria.name[k])
  print(k)
  fullPMNLE$cTri_all[k] <- fullPMNLE$cTri[k]
  fullPMNLE$loops_all[k] <- fullPMNLE$loops[k]
  fullPMNLE$ccs_all[k] <- fullPMNLE$ccs[k]
  for (kk in 1:length(sample_names_497$Sample)){
     if(fullPMNLE$STRING.bacteria.name[k] == sample_names_497$Sample[kk]){
        for (kk2 in 1:length(BettiNumber$Sample)){
         if(fullPMNLE$STRING.bacteria.name[k] == BettiNumber$Sample[kk2]){
         fullPMNLE$cTri_all[k] <- BettiNumber$cTri_all[kk2]
         fullPMNLE$loops_all[k] <- BettiNumber$loop_all[kk2]
         fullPMNLE$ccs_all[k] <- BettiNumber$ccs_all[kk2]
         fullPMNLE$Entry[k] <- 1 } } # end  BettiNumber
     }# end if 
  } # end for sample
}

# Repeat
for (k in 1:length(fullPMNLE$STRING.bacteria.name)){
  print(fullPMNLE$STRING.bacteria.name[k])
  print(k)
  for (kk in 1:length(sample_names_497$Sample)){
     if(fullPMNLE$STRING.bacteria.name[k] == sample_names_497$Sample[kk]){
        for (kk3 in 1:length(approximatePMNLE$STRING.bacteria.name)){
         if(fullPMNLE$STRING.bacteria.name[k] == approximatePMNLE$STRING.bacteria.name[kk3]){
         fullPMNLE$aPMNLE_all[k] <- approximatePMNLE$aPMNLE[kk3] } } } # end  approximatePMNLE
  } # end for sample
}

# Repeat
fullPMNLE$Fraction <- ((fullPMNLE$Plasmid_PPI + fullPMNLE$Plasmid_Plasmid_PPI)/
    (fullPMNLE$Plasmid_PPI + fullPMNLE$Plasmid_Plasmid_PPI + fullPMNLE$Chrom_PPI))
fullPMNLE$Fraction[is.na(fullPMNLE$Fraction)] <- 0
hist(log10(fullPMNLE$Fraction), breaks=111) + grid()
summary(subset(fullPMNLE, Fraction >0)$Fraction) #  
# Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
# 0.0000454 0.0013542 0.0035293 0.0138450 0.0095459 0.5197107
sd(subset(fullPMNLE, Fraction >0)$Fraction)  

# fullPMNLE$cTri fullPMNLE$loops fullPMNLE$ccs fullPMNLE$aPMNLE 
# fullPMNLE$cTri_all fullPMNLE$loops_all fullPMNLE$ccs_all fullPMNLE$aPMNLE_all

for (i in 1:length(fullPMNLE$cTri)){
  fullPMNLE$cTri[i] <- str_remove(fullPMNLE$cTri[i], "c\\(")
  fullPMNLE$cTri[i] <- str_remove(fullPMNLE$cTri[i], "\\)")
  fullPMNLE$cTri[i] <- as.numeric(tail(unlist(strsplit(fullPMNLE$cTri[i], ",")), n=1))
} # Full aPMNLE

for (i in 1:length(fullPMNLE$loops)){
  fullPMNLE$loops[i] <- str_remove(fullPMNLE$loops[i], "c\\(")
  fullPMNLE$loops[i] <- str_remove(fullPMNLE$loops[i], "\\)")
  fullPMNLE$loops[i] <- as.numeric(tail(unlist(strsplit(fullPMNLE$loops[i], ",")), n=1))
} # Full aPMNLE

for (i in 1:length(fullPMNLE$ccs)){
  fullPMNLE$ccs[i] <- str_remove(fullPMNLE$ccs[i], "c\\(")
  fullPMNLE$ccs[i] <- str_remove(fullPMNLE$ccs[i], "\\)")
  fullPMNLE$ccs[i] <- as.numeric(tail(unlist(strsplit(fullPMNLE$ccs[i], ",")), n=1))
} # Full aPMNLE

for (i in 1:length(fullPMNLE$cTri_all)){
  fullPMNLE$cTri_all[i] <- str_remove(fullPMNLE$cTri_all[i], "c\\(")
  fullPMNLE$cTri_all[i] <- str_remove(fullPMNLE$cTri_all[i], "\\)")
  fullPMNLE$cTri_all[i] <- as.numeric(tail(unlist(strsplit(fullPMNLE$cTri_all[i], ",")), n=1))
} # Full aPMNLE

for (i in 1:length(fullPMNLE$loops_all)){
  fullPMNLE$loops_all[i] <- str_remove(fullPMNLE$loops_all[i], "c\\(")
  fullPMNLE$loops_all[i] <- str_remove(fullPMNLE$loops_all[i], "\\)")
  fullPMNLE$loops_all[i] <- as.numeric(tail(unlist(strsplit(fullPMNLE$loops_all[i], ",")), n=1))
} # Full aPMNLE

for (i in 1:length(fullPMNLE$ccs_all)){
  fullPMNLE$ccs_all[i] <- str_remove(fullPMNLE$ccs_all[i], "c\\(")
  fullPMNLE$ccs_all[i] <- str_remove(fullPMNLE$ccs_all[i], "\\)")
  fullPMNLE$ccs_all[i] <- as.numeric(tail(unlist(strsplit(fullPMNLE$ccs_all[i], ",")), n=1))
} # Full aPMNLE

# Repeat
# View(fullPMNLE) # Full aPMNLE
write.csv(fullPMNLE, "Plasmid_PPI_data_PMNLE.csv")
#fullPMNLE <- data.frame(read.csv("Plasmid_PPI_data_PMNLE.csv"))
str(fullPMNLE) # 4363  

length(subset(fullPMNLE, Plasmid_Plasmid_PPI > 0)$Plasmid_Plasmid_PPI) # 283
summary(subset(fullPMNLE, Plasmid_Plasmid_PPI > 0)$Plasmid_Plasmid_PPI)
#    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#    1.0     1.0     5.0   165.3    16.0 11065.0 

length(subset(fullPMNLE, Plasmid_PPI > 0)$Plasmid_PPI) # 489
summary(subset(fullPMNLE, Plasmid_PPI > 0)$Plasmid_PPI)
#    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#    2.0    68.0   168.0   798.8   554.0 34629.0 
sd(subset(fullPMNLE, Plasmid_PPI > 0)$Plasmid_PPI) # 2877.302
summary(subset(fullPMNLE, Plasmid_PPI > 0)$Chrom_PPI)
#    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#     8431   38356   53947   58391   72569  179015 
sd(subset(fullPMNLE, Plasmid_PPI > 0)$Chrom_PPI) # 26722.16

length(subset(fullPMNLE, Chrom_PPI > 0)$Chrom_PPI) # 4363
summary(subset(fullPMNLE, Chrom_PPI > 0)$Chrom_PPI)
#   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 #  851   33156   48834   57267   70793  377867 
sd(subset(fullPMNLE, Chrom_PPI > 0)$Chrom_PPI) # 46614.8

tt <-(subset(fullPMNLE, Plasmid_PPI > 0))
summary(tt$Plasmid_Plasmid_PPI + tt$Plasmid_PPI) # +- 3592.446
#    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#   2.0    69.0   169.0   894.4   559.0 45694.0 

summary((fullPMNLE$Plasmid_Plasmid_PPI)/(fullPMNLE$Plasmid_Plasmid_PPI + 
                        fullPMNLE$Plasmid_PPI + fullPMNLE$Chrom_PPI)) # 0.01543% 
summary((fullPMNLE$Plasmid_PPI)/(fullPMNLE$Plasmid_Plasmid_PPI + 
                        fullPMNLE$Plasmid_PPI + fullPMNLE$Chrom_PPI)) # 0.1398% 
summary((fullPMNLE$Plasmid_PPI + fullPMNLE$Plasmid_Plasmid_PPI)/
          (fullPMNLE$Plasmid_Plasmid_PPI + 
                        fullPMNLE$Plasmid_PPI + fullPMNLE$Chrom_PPI)) # 0.1552% 
summary(fullPMNLE$Fraction)
sum(fullPMNLE$Chrom_PPI) #  286364425
sum(fullPMNLE$Plasmid_Plasmid_PPI) # 46772
sum(fullPMNLE$Plasmid_PPI) # 390592

 cor ( (fullPMNLE$Plasmid_PPI + fullPMNLE$Plasmid_Plasmid_PPI),
              (fullPMNLE$inputProteins)) # r=0.05

fullPMNLE_new <- subset(fullPMNLE, Plasmid_PPI >0)
str(fullPMNLE_new)
r2d <- cor (log10(fullPMNLE_new$Plasmid_PPI + fullPMNLE_new$Plasmid_Plasmid_PPI),
             log10(fullPMNLE_new$inputProteins)) # r=0.283
fig4 <- ggplot(fullPMNLE_new, aes(y= log10(fullPMNLE_new$Plasmid_PPI + 
      fullPMNLE_new$Plasmid_Plasmid_PPI), x=log10(fullPMNLE_new$Chrom_PPI))) +
  geom_point(color ="blue", size=0.2, shape=1, (aes(alpha=0.1)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of chromosomal PPIs", limits =c(4.3,5.2)) +
  scale_y_continuous(name="Log10 of number of plasmid PPIs", limits =c(1,4))  + 
  geom_smooth(method='lm', col="red") + 
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE_new$Chrom_PPI),
    " bacterial samples",sep=""),x=0.67,y=.1,hjust=0, gp=gpar(col="black",fontsize=9)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r=", round(r2d,3),sep=""),
  x=0.67, y=.075, hjust=0, gp=gpar(col="black", fontsize=9)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r2=", round(r2d**2,4),sep=""),
  x=0.67, y=.05, hjust=0, gp=gpar(col="black", fontsize=9))))  

fig4_2 <- ggExtra::ggMarginal(fig4, type="densigram",
                    size=5,colour="grey", bins=50)  
pdf("Fig_SS4.pdf", width=6, height=6, bg="white")
ggarrange(fig4_2, ncol=1, nrow=1)
dev.off()

########################################################################
#################################################################
# Full aPMNLE

png("FigS_Triangles_distribution_all.png", width=500, height=600, units="px")
hist(log10(as.numeric(fullPMNLE$cTri)), breaks=123, col="grey",
     xlab="Log10-scaled numbers of all protein trios", main="", cex=1.2,
     cex.axis=1.4) + grid()
dev.off()
png("FigS_Loops_distribution_all.png", width=500, height=600, units="px")
hist(log10(as.numeric(fullPMNLE$loops)), breaks=123, col="grey",
     xlab="Log10-scaled numbers of all indirect connections", main="",
     cex=1.2, cex.axis=1.4) + grid()
dev.off()
png("FigS_Components_distribution_all.png", width=500, height=600, units="px")
hist(log10(fullPMNLE$ccs), breaks=123, col="grey",
     xlab="PPI network components - all genes", main="", cex=1.2,
     cex.axis=1.4) + grid()
dev.off()

pvalues <- c(0)
cor.test(as.numeric(fullPMNLE$loops), fullPMNLE$aPMNLE)[4]
cor.test(as.numeric(fullPMNLE$loops), fullPMNLE$aPMNLE)[3]
cor.test(as.numeric(fullPMNLE$loops), fullPMNLE$aPMNLE)$conf.int[1:2]
# Loops: r=0.3521433 raw p=2.823332e-123   # p value 1
pvalues <- c(2.823332e-123, pvalues)

cor.test(as.numeric(fullPMNLE$ccs), fullPMNLE$aPMNLE)[4]
cor.test(as.numeric(fullPMNLE$ccs), fullPMNLE$aPMNLE)[3]
cor.test(as.numeric(fullPMNLE$ccs), fullPMNLE$aPMNLE)$conf.int[1:2]
# Connected Components: r=-0.1937999  raw p=6.111757e-37 95% CI -0.2226910 -0.1645687
pvalues <- c(6.111757e-37, pvalues) # p value 2

cor.test(as.numeric(fullPMNLE$loops_all), fullPMNLE$aPMNLE_all)[4]
cor.test(as.numeric(fullPMNLE$loops_all), fullPMNLE$aPMNLE_all)[3]
cor.test(as.numeric(fullPMNLE$loops_all), fullPMNLE$aPMNLE_all)$conf.int[1:2]
# Loops: r=0.2994389  raw p=1.828104e-88 95% CI 0.2717708 0.3266125
pvalues <- c(1.828104e-88, pvalues) # p value 3

cor.test(as.numeric(fullPMNLE$ccs_all), fullPMNLE$aPMNLE_all)[4]
cor.test(as.numeric(fullPMNLE$ccs_all), fullPMNLE$aPMNLE_all)[3]
cor.test(as.numeric(fullPMNLE$ccs_all), fullPMNLE$aPMNLE_all)$conf.int[1:2]
# Connected Components: r=-0.02929255 raw p=0.05663542 95% CI 
pvalues <- c(0.05663542, pvalues) # p value 4

# Compile final info from aPMNLEs for Table S3
write.xlsx(fullPMNLE, "Key_Results_Figures/Table_S3_new.xlsx")

########################################################################
########### Examine change in aPMNLE for chrom vs all, not useful # ####
########################################################################

# % fall in from full PNMLE to partial PMNLE
fullPMNLE$diffPMNLE <-(fullPMNLE$aPMNLE_all - fullPMNLE$aPMNLE)/(fullPMNLE$aPMNLE_all) 
hist(na.omit(fullPMNLE$diffPMNLE), breaks=123, col="grey", main="", 
     xlab="Difference in PMNLE", cex=1.4, cex.axis=1.4,  cex.lab=1.4) + grid()
summary(fullPMNLE$diffPMNLE)
str(fullPMNLE$diffPMNLE) # 4363  
#     Min.   1st Qu.    Median      Mean   3rd Qu.  Max.      NA's 
# -0.11488  0.00000  0.00000  0.00448  0.00000  0.76252      139 
subset(fullPMNLE, fullPMNLE$diffPMNLE > 0.5)[c(1:3,10:13)]
sum(is.na(fullPMNLE$aPMNLE) ) # 136 NAs
sum(!(is.na(fullPMNLE$aPMNLE) ))#4227 not NAs

# fullPMNLE4
fullPMNLE3 <- subset(fullPMNLE, Plasmid_genes>0)
fullPMNLE3$Plasmid_PPI_log <- log10(fullPMNLE3$Plasmid_PPI)
dim(fullPMNLE3)  # 491
fullPMNLE4 <- subset(fullPMNLE3, fullPMNLE3$Plasmid_PPI_log >0)
fullPMNLE4$Sample <- fullPMNLE4$STRING.bacteria.name
dim(fullPMNLE4) # 489

summary(fullPMNLE4$Plasmid_PPI/fullPMNLE4$Plasmid_genes)
sd(fullPMNLE4$Plasmid_PPI/fullPMNLE4$Plasmid_genes)
#   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#   1.40   19.00   29.90   39.15   49.00  452.00 
summary(fullPMNLE4$Chrom_PPI/fullPMNLE4$Chrom_genes)
sd(fullPMNLE4$Chrom_PPI/fullPMNLE4$Chrom_genes)
#   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#   7.164  12.364  13.889  14.715  16.151  28.973 
t.test(fullPMNLE4$Plasmid_PPI/fullPMNLE4$Plasmid_genes, 
        fullPMNLE4$Chrom_PPI/fullPMNLE4$Chrom_genes)
wilcox.test(fullPMNLE4$Plasmid_PPI/fullPMNLE4$Plasmid_genes, 
        fullPMNLE4$Chrom_PPI/fullPMNLE4$Chrom_genes)


png("Fig_S19_PMNLE_distribution_all.png", width=600, height=1200, units="px")
par(mfrow=c(3,1))
hist(fullPMNLE4$aPMNLE_all, breaks=123, xlab="PMNLE values - all genes",
     main="", cex=1.8, col="red", cex.axis=1.5, cex.lab=1.5, xlim=c(0,0.7)) + grid()
hist(fullPMNLE$aPMNLE, breaks=123, xlab="PMNLE values - chromosomal genes",
     main="", cex=1.8, col="blue", cex.axis=1.5, cex.lab=1.5, xlim=c(0,0.7)) + grid()
hist(fullPMNLE4$diffPMNLE, breaks=123, xlab="Change in PMNLE values", # ylim=c(0,50),
     main="", cex=1.8, col="green", cex.axis=1.5, cex.lab=1.5) + grid()
dev.off()

# Association of chrom PMNLE with numbers of plasmid-related PPIs
plot(log10(fullPMNLE$Plasmid_PPI), fullPMNLE$aPMNLE) + grid()

# Association of full PMNLE with numbers of plasmid-related PPIs
plot( log10(fullPMNLE$Plasmid_PPI), fullPMNLE$aPMNLE_all, ylim=c(0.48,0.63)) + grid()

# Association of difference in PMNLE with numbers of plasmid-related PPIs
plot(log10(fullPMNLE$Plasmid_PPI), fullPMNLE$diffPMNLE) + grid()

########
# Association of chrom PMNLE with numbers of chrom-related PPIs - weaker
plot(log10(fullPMNLE$Chrom_PPI), fullPMNLE$aPMNLE) + grid()
abline(lm(fullPMNLE$aPMNLE ~ log10(fullPMNLE$Chrom_PPI)), col="green")
cor.test(log10(fullPMNLE$Chrom_PPI), fullPMNLE$aPMNLE)
cor.test(fullPMNLE$Chrom_PPI, fullPMNLE$aPMNLE)[3]
cor.test(fullPMNLE$Chrom_PPI, fullPMNLE$aPMNLE)$conf.int[1:2]
# Chrom_PPI r=0.4404846   p=3.0668e-89  
pvalues <- c(3.0668e-98, pvalues) # p value 8

# Association of full PMNLE with numbers of all-related PPIs - weaker
plot(log10(fullPMNLE$Chrom_PPI), fullPMNLE$aPMNLE_all) + grid()
abline(lm(fullPMNLE$aPMNLE_all ~ log10(fullPMNLE$Chrom_PPI)), col="green")
cor.test(log10(fullPMNLE$Chrom_PPI), fullPMNLE$aPMNLE_all)
cor.test(fullPMNLE$Chrom_PPI, fullPMNLE$aPMNLE_all)[3]
cor.test(fullPMNLE$Chrom_PPI, fullPMNLE$aPMNLE_all)$conf.int[1:2]
# Chrom_PPI r=0.4212498 p=6.785402e-79  
pvalues <- c(6.785402e-79, pvalues) # p value 9

# Association of difference in PMNLE with numbers of chrom-related PPIs NO
plot(log10(fullPMNLE$Chrom_PPI), fullPMNLE$diffPMNLE) + grid()
 
# ########
# Key obvious finding
# Association of chrom PMNLE with numbers of total PPIs
plot(log10(fullPMNLE$interactions), fullPMNLE$aPMNLE) + grid()
abline(lm(fullPMNLE$aPMNLE ~ log10(fullPMNLE$interactions)),col="red")
cor.test(log10(fullPMNLE$interactions), fullPMNLE$aPMNLE)
cor.test(log10(fullPMNLE$interactions), fullPMNLE$aPMNLE)[3]
cor.test(log10(fullPMNLE$interactions), fullPMNLE$aPMNLE)$conf.int[1:2]
# interactions r=0.44  p=2.150176e-198  
pvalues <- c(2.150176e-198, pvalues) # p value 11

# Association of full PMNLE with numbers of total PPIs
plot(log10(fullPMNLE$interactions), fullPMNLE$aPMNLE_all) + grid()
abline(lm(fullPMNLE$aPMNLE_all ~ log10(fullPMNLE$interactions)),col="red")
cor.test(log10(fullPMNLE$interactions), fullPMNLE$aPMNLE_all)
cor.test(log10(fullPMNLE$interactions), fullPMNLE$aPMNLE_all)[3]
cor.test(log10(fullPMNLE$interactions), fullPMNLE$aPMNLE_all)$conf.int[1:2]
# interactions r=0.42  p= 3.316022e-183 
pvalues <- c( 3.316022e-183, pvalues) # p value 12

# Association of difference in PMNLE with numbers of total PPIs
plot(log10(fullPMNLE$interactions), fullPMNLE$diffPMNLE) + grid()

fullPMNLE3 <- subset(fullPMNLE, Plasmid_genes>0)
fullPMNLE3$Plasmid_PPI_log <- log10(fullPMNLE3$Plasmid_PPI)
dim(fullPMNLE3)  # 491
fullPMNLE4 <- subset(fullPMNLE3, fullPMNLE3$Plasmid_PPI_log >0)
fullPMNLE4$Sample <- fullPMNLE4$STRING.bacteria.name
dim(fullPMNLE4) # 489

#########################
# plot to visualise number of PPIs vs chrom aPMNLE - not used
rvalue <- cor.test(fullPMNLE4$aPMNLE, fullPMNLE4$Plasmid_PPI_log)$estimate
ppi_pmnle <- ggplot(fullPMNLE4, aes(x=log10(fullPMNLE4$interactions),
                                   y=fullPMNLE4$aPMNLE)) +
  geom_point(color ="blue", size=0.3, shape=1, (aes(alpha=0.1)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of PPIs") +
  scale_y_continuous(name="Chromosomal PMNLE indirect connections")  + 
  geom_smooth(method='lm') +
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE4$STRING.bacteria.name),
                                            " bacterial samples",sep=""),
              x=0.6, y=.08, hjust=0, gp=gpar(col="black", fontsize=11)))) +
  annotation_custom(grobTree(textGrob(paste("adj_r=", round(rvalue,3),sep=""),
  x=0.7, y=.05, hjust=0, gp=gpar(col="black", fontsize=11)))) +
  annotation_custom(grobTree(textGrob(paste("adj_r2=", round(rvalue**2,3),sep=""),
  x=0.69, y=.02, hjust=0, gp=gpar(col="black", fontsize=11))))
ppi_pmnle1 <- ggExtra::ggMarginal(ppi_pmnle, type="densigram", size=6,
                                  colour="grey", bins=50)
#pdf("Fig_S13_PPI_PMNLE_chrom.pdf", width=6, height=6, bg="white")
ggarrange(ppi_pmnle1, ncol=1, nrow=1)
#dev.off()

#########################
# plot to visualise number of PPIs vs all proteins aPMNLE - not used
# ggplot it
rvalue <- sqrt(summary(lm((fullPMNLE4$aPMNLE_all) ~
                            log10(fullPMNLE4$interactions)))$adj.r.squared)
ppi_pmnle2 <- ggplot(fullPMNLE4, aes(x=log10(fullPMNLE4$interactions),
                                   y=fullPMNLE4$aPMNLE_all)) +
  geom_point(color ="blue", size=0.3, shape=1, (aes(alpha=0.1)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of PPIs") +
  scale_y_continuous(name="Full PMNLE indirect connections")  + 
  geom_smooth(method='lm') +
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE4$Sample),
                                            " bacterial samples",sep=""),
              x=0.6, y=.08, hjust=0, gp=gpar(col="black", fontsize=11)))) +
  annotation_custom(grobTree(textGrob(paste("adj_r=", round(rvalue,3),sep=""),
  x=0.7, y=.05, hjust=0, gp=gpar(col="black", fontsize=11)))) +
  annotation_custom(grobTree(textGrob(paste("adj_r2=", round(rvalue**2,3),sep=""),
  x=0.69, y=.02, hjust=0, gp=gpar(col="black", fontsize=11))))
ppi_pmnle12 <- ggExtra::ggMarginal(ppi_pmnle2, type="densigram", size=6,
                                  colour="grey", bins=50)
#pdf("Fig_5_PPI_PMNLE_all.pdf", width=6, height=6, bg="white")
ggarrange(ppi_pmnle12, ncol=1, nrow=1)
#dev.off()

######################### Below was not published, not useful #########
# plot to visualise number of PPIs vs change in aPMNLE - not used
# difference in PMNLE
# ggplot it
rvalue <- sqrt(summary(lm((fullPMNLE4$diffPMNLE) ~
                            log10(fullPMNLE4$Plasmid_PPI)))$adj.r.squared)
ppi_pmnle3<- ggplot(fullPMNLE4, aes(x=log10(fullPMNLE4$interactions),
                                   y=fullPMNLE4$diffPMNLE)) +
  geom_point(color ="blue", size=0.3, shape=1, (aes(alpha=0.1)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of PPIs") +
  scale_y_continuous(name="Difference in PMNLE indirect connections")  + 
  geom_smooth(method='lm') +
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE4$Sample),
                                            " bacterial samples",sep=""),
              x=0.6, y=.08, hjust=0, gp=gpar(col="black", fontsize=11)))) +
  annotation_custom(grobTree(textGrob(paste("adj_r=", round(rvalue,3),sep=""),
  x=0.7, y=.05, hjust=0, gp=gpar(col="black", fontsize=11)))) +
  annotation_custom(grobTree(textGrob(paste("adj_r2=", round(rvalue**2,3),sep=""),
  x=0.69, y=.02, hjust=0, gp=gpar(col="black", fontsize=11))))
ppi_pmnle13 <- ggExtra::ggMarginal(ppi_pmnle3, type="densigram", size=6,
                                  colour="grey", bins=50)
#pdf("Fig_5_PPI_PMNLE_diff.pdf", width=6, height=6, bg="white") # CHECK
ggarrange(ppi_pmnle13, ncol=1, nrow=1)
#dev.off()
####################

###################
# make ggplot of PPI, chrom_PPI, plasmid_PPI, proteins across:
# full PMNLE, chrom PMNLE and difference in PMNLE

r1 <- sqrt(summary(lm((fullPMNLE$aPMNLE) ~ log10(fullPMNLE$interactions)))$adj.r.squared) # r=0.45
r2 <- sqrt(summary(lm((fullPMNLE4$aPMNLE_all) ~ log10(fullPMNLE4$interactions)))$adj.r.squared) # r=0.19
r3 <- sqrt(summary(lm((fullPMNLE4$diffPMNLE) ~ log10(fullPMNLE4$interactions)))$adj.r.squared) # r=0.52

figure5_a_1 <- ggplot(fullPMNLE4, aes(x=log10(fullPMNLE4$interactions),
                                          y=fullPMNLE4$aPMNLE_all)) +
  geom_point(color ="black", size=0.1, shape=1, (aes(alpha=0.1)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of PPIs per sample", limits =c(4,5.4)) +
  scale_y_continuous(name="aPMNLE (indirect connectivity)", limits =c(-0.1,0.7))  + 
  geom_smooth(method='lm', col="black") + 
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE4$Sample),
    " bacterial samples",sep=""),x=0.6,y=.65,hjust=0, gp=gpar(col="black",fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Full PMNLE adj_r=", round(r2,3),sep=""),
  x=0.6, y=.45, hjust=0, gp=gpar(col="black", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Full PMNLE adj_r2=", round(r2**2,3),sep=""),
  x=0.6, y=.42, hjust=0, gp=gpar(col="black", fontsize=7))))  
  
figure5_a_2 <- figure5_a_1 + geom_point(data=fullPMNLE,
        aes(x=log10(fullPMNLE$interactions), y=fullPMNLE$aPMNLE, alpha=0.1),
             color ="red", size=0.1, shape=2, show.legend=F) + 
  geom_smooth(data=fullPMNLE,
        aes(x=log10(fullPMNLE$interactions), y=fullPMNLE$aPMNLE),
        method='lm', col="red") +
  annotation_custom(grobTree(textGrob(paste("Chrom PMNLE adj_r=", round(r1,3),sep=""),
  x=0.6, y=.52, hjust=0, gp=gpar(col="red", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Chrom PMNLE adj_r2=", round(r1**2,3),sep=""),
  x=0.6, y=.49, hjust=0, gp=gpar(col="red", fontsize=7)))) 

figure5_a_3 <- figure5_a_2 + geom_point(data=fullPMNLE4,
    aes(x=log10(fullPMNLE4$interactions), y=fullPMNLE4$diffPMNLE, alpha=0.1),
             color ="blue", size=0.1, shape=3, show.legend=F) + 
  geom_smooth(data=fullPMNLE4,
    aes(x=log10(fullPMNLE4$interactions), y=fullPMNLE4$diffPMNLE),
    method='lm', col="blue") +
  annotation_custom(grobTree(textGrob(paste("Change in PMNLE adj_r=-", round(r3,3),sep=""),
  x=0.6, y=.60, hjust=0, gp=gpar(col="blue", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Change in PMNLE adj_r2=", round(r3**2,3),sep=""),
  x=0.6, y=.57, hjust=0, gp=gpar(col="blue", fontsize=7))))  + labs(tag="A")

### Proteins
r1p <- sqrt(summary(lm((fullPMNLE$aPMNLE) ~ log10(fullPMNLE$inputProteins)))$adj.r.squared) # 0.39
r2p <- sqrt(summary(lm((fullPMNLE4$aPMNLE_all) ~ log10(fullPMNLE4$inputProteins)))$adj.r.squared) # 0.05
r3p <- sqrt(summary(lm((fullPMNLE4$diffPMNLE) ~ log10(fullPMNLE4$inputProteins)))$adj.r.squared) # 0.52

figure5_b_1 <- ggplot(fullPMNLE4, aes(x=log10(fullPMNLE4$inputProteins),
                                          y=fullPMNLE4$aPMNLE_all)) +
  geom_point(color ="black", size=0.1, shape=1, (aes(alpha=0.1)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of proteins per sample", limits =c(3,4)) +
  scale_y_continuous(name="aPMNLE (indirect connectivity)", limits =c(-0.1,0.7))  + 
  geom_smooth(method='lm', col="black") + 
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE4$Sample),
    " bacterial samples",sep=""),x=0.6,y=.65,hjust=0, gp=gpar(col="black",fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Full PMNLE adj_r=", round(r2p,3),sep=""),
  x=0.6, y=.45, hjust=0, gp=gpar(col="black", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Full PMNLE adj_r2=", round(r2p**2,3),sep=""),
  x=0.6, y=.42, hjust=0, gp=gpar(col="black", fontsize=7))))  
  
figure5_b_2 <- figure5_b_1 + geom_point(data=fullPMNLE,
        aes(x=log10(fullPMNLE$inputProteins), y=fullPMNLE$aPMNLE, alpha=0.1),
             color ="red", size=0.1, shape=2, show.legend=F) + 
  geom_smooth(data=fullPMNLE,
        aes(x=log10(fullPMNLE$inputProteins), y=fullPMNLE$aPMNLE),
        method='lm', col="red") +
  annotation_custom(grobTree(textGrob(paste("Chrom PMNLE adj_r=", round(r1p,3),sep=""),
  x=0.6, y=.52, hjust=0, gp=gpar(col="red", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Chrom PMNLE adj_r2=", round(r1p**2,3),sep=""),
  x=0.6, y=.49, hjust=0, gp=gpar(col="red", fontsize=7)))) 

figure5_b_3 <- figure5_b_2 + geom_point(data=fullPMNLE4,
    aes(x=log10(fullPMNLE4$inputProteins), y=fullPMNLE4$diffPMNLE, alpha=0.1),
             color ="blue", size=0.1, shape=3, show.legend=F) + 
  geom_smooth(data=fullPMNLE4,
    aes(x=log10(fullPMNLE4$inputProteins), y=fullPMNLE4$diffPMNLE),
    method='lm', col="blue") +
  annotation_custom(grobTree(textGrob(paste("Change in PMNLE adj_r=-", round(r3p,3),sep=""),
  x=0.6, y=.60, hjust=0, gp=gpar(col="blue", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Change in PMNLE adj_r2=", round(r3p**2,3),sep=""),
  x=0.6, y=.57, hjust=0, gp=gpar(col="blue", fontsize=7)))) + labs(tag="B")

### Chrom_PPI
r1c <- sqrt(summary(lm((fullPMNLE$aPMNLE) ~ log10(fullPMNLE$Chrom_PPI)))$adj.r.squared) # 0.45
r2c <- sqrt(summary(lm((fullPMNLE4$aPMNLE_all) ~ log10(fullPMNLE4$Chrom_PPI)))$adj.r.squared) # 0.16
r3c <- sqrt(summary(lm((fullPMNLE4$diffPMNLE) ~ log10(fullPMNLE4$Chrom_PPI)))$adj.r.squared) # 0.54

figure5_c_1 <- ggplot(fullPMNLE4, aes(x=log10(fullPMNLE4$Chrom_PPI),
                                          y=fullPMNLE4$aPMNLE_all)) +
  geom_point(color ="black", size=0.1, shape=1, (aes(alpha=0.1)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of chromosomal PPIs per sample", limits=c(3.9,5.3)) +
  scale_y_continuous(name="aPMNLE (indirect connectivity)", limits =c(-0.1,0.7))  + 
  geom_smooth(method='lm', col="black") + 
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE4$Sample),
    " bacterial samples",sep=""),x=0.61,y=.65,hjust=0, gp=gpar(col="black",fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Full PMNLE adj_r=", round(r2c,3),sep=""),
  x=0.6, y=.45, hjust=0, gp=gpar(col="black", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Full PMNLE adj_r2=", round(r2c**2,3),sep=""),
  x=0.6, y=.43, hjust=0, gp=gpar(col="black", fontsize=7))))  
  
figure5_c_2 <- figure5_c_1 + geom_point(data=fullPMNLE,
        aes(x=log10(fullPMNLE$Chrom_PPI), y=fullPMNLE$aPMNLE, alpha=0.1),
             color ="red", size=0.1, shape=2, show.legend=F) + 
  geom_smooth(data=fullPMNLE,
        aes(x=log10(fullPMNLE$Chrom_PPI), y=fullPMNLE$aPMNLE),
        method='lm', col="red") +
  annotation_custom(grobTree(textGrob(paste("Chrom PMNLE adj_r=", round(r1c,3),sep=""),
  x=0.6, y=.4, hjust=0, gp=gpar(col="red", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Chrom PMNLE adj_r2=", round(r1c**2,3),sep=""),
  x=0.6, y=.37, hjust=0, gp=gpar(col="red", fontsize=7)))) 

figure5_c_3 <- figure5_c_2 + geom_point(data=fullPMNLE4,
    aes(x=log10(fullPMNLE4$Chrom_PPI), y=fullPMNLE4$diffPMNLE, alpha=0.1),
             color ="blue", size=0.1, shape=3, show.legend=F) + 
  geom_smooth(data=fullPMNLE4,
    aes(x=log10(fullPMNLE4$Chrom_PPI), y=fullPMNLE4$diffPMNLE),
    method='lm', col="blue") +
  annotation_custom(grobTree(textGrob(paste("Change in PMNLE adj_r=-", round(r3c,3),sep=""),
  x=0.6, y=.32, hjust=0, gp=gpar(col="blue", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Change in PMNLE adj_r2=", round(r3c**2,3),sep=""),
  x=0.6, y=.29, hjust=0, gp=gpar(col="blue", fontsize=7))))  + labs(tag="C")

### Plasmid_PPI
r1d <- sqrt(summary(lm((fullPMNLE4$aPMNLE) ~ log10(fullPMNLE4$Plasmid_PPI)))$adj.r.squared) # 0.09
r2d <- sqrt(summary(lm((fullPMNLE4$aPMNLE_all) ~ log10(fullPMNLE4$Plasmid_PPI)))$adj.r.squared) # 0.12
r3d <- sqrt(summary(lm((fullPMNLE4$diffPMNLE) ~ log10(fullPMNLE4$Plasmid_PPI)))$adj.r.squared) # 0.01

figure5_d_1 <- ggplot(fullPMNLE4, aes(x=log10(fullPMNLE4$Plasmid_PPI),
                                          y=fullPMNLE4$aPMNLE_all)) +
  geom_point(color ="black", size=0.1, shape=1, (aes(alpha=0.1)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of plasmid PPIs per sample", limits =c(1,4.5)) +
  scale_y_continuous(name="aPMNLE (indirect connectivity)", limits =c(-0.1,0.7))  + 
  geom_smooth(method='lm', col="black") + 
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE4$Sample),
    " bacterial samples",sep=""),x=0.61,y=.65,hjust=0, gp=gpar(col="black",fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Full PMNLE adj_r=", round(r2d,3),sep=""),
  x=0.6, y=.45, hjust=0, gp=gpar(col="black", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Full PMNLE adj_r2=", round(r2d**2,3),sep=""),
  x=0.6, y=.42, hjust=0, gp=gpar(col="black", fontsize=7))))  
  
figure5_d_2 <- figure5_d_1 + geom_point(data=fullPMNLE4,
        aes(x=log10(fullPMNLE4$Plasmid_PPI), y=fullPMNLE4$aPMNLE, alpha=0.1),
             color ="red", size=0.1, shape=2, show.legend=F) + 
  geom_smooth(data=fullPMNLE4,
        aes(x=log10(fullPMNLE4$Plasmid_PPI), y=fullPMNLE4$aPMNLE),
        method='lm', col="red") +
  annotation_custom(grobTree(textGrob(paste("Chrom PMNLE adj_r=", round(r1d,3),sep=""),
  x=0.6, y=.52, hjust=0, gp=gpar(col="red", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Chrom PMNLE adj_r2=", round(r1d**2,3),sep=""),
  x=0.6, y=.49, hjust=0, gp=gpar(col="red", fontsize=7)))) 

figure5_d_3 <- figure5_d_2 + geom_point(data=fullPMNLE4,
    aes(x=log10(fullPMNLE4$Plasmid_PPI), y=fullPMNLE4$diffPMNLE, alpha=0.1),
             color ="blue", size=0.1, shape=3, show.legend=F) + 
  geom_smooth(data=fullPMNLE4,
    aes(x=log10(fullPMNLE4$Plasmid_PPI), y=fullPMNLE4$diffPMNLE),
    method='lm', col="blue") +
  annotation_custom(grobTree(textGrob(paste("Change in PMNLE adj_r=-", round(r3d,3),sep=""),
  x=0.6, y=.60, hjust=0, gp=gpar(col="blue", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Change in PMNLE adj_r2=", round(r3d**2,3),sep=""),
  x=0.6, y=.57, hjust=0, gp=gpar(col="blue", fontsize=7)))) + labs(tag="D")

#####
figure5_a1 <- ggExtra::ggMarginal(figure5_a_3, type="densigram",
                    size=5,colour="grey", bins=50)  
figure5_b1 <- ggExtra::ggMarginal(figure5_b_3, type="densigram",
                    size=5,colour="grey", bins=50)  
figure5_c1 <- ggExtra::ggMarginal(figure5_c_3, type="densigram",
                    size=5,colour="grey", bins=50)  
figure5_d1 <- ggExtra::ggMarginal(figure5_d_3, type="densigram",
                    size=5,colour="grey", bins=50)  
par(mfrow=c(2,2))
pdf("Fig_S20_indirectconnectivity.pdf", width=10, height=12, bg="white")
ggarrange(figure5_a1, figure5_b1, figure5_c1, figure5_d1, ncol=2, nrow=2)
dev.off()
#png("Fig_5_PMNLE.png", width=2500, height=2600, units="px")
#ggarrange(figure5_a1, figure5_b1, figure5_c1, figure5_d1, ncol=2, nrow=2)
#dev.off()

########################################################################
########### Numbers of connections vs proteins & other metadata -> Fig_S21 ##
########################################################################
###################
# make ggplot of PPI, chrom_PPI, plasmid_PPI, proteins across:
# indirect connections

par(mfrow=c(1,1))
fullPMNLE$indirect_chrom <- (fullPMNLE$loops/fullPMNLE$inputProteins)
fullPMNLE$indirect_all <- fullPMNLE$loops_all/fullPMNLE$inputProteins
#fullPMNLE$changeLoops <- (fullPMNLE$loops_all - fullPMNLE$loops)/
#                           (fullPMNLE$loops*fullPMNLE$inputProteins)
  
hist(fullPMNLE$indirect_all, breaks=111, xlim=c(0,15)) + grid()
hist(fullPMNLE$indirect_chrom, breaks=111, xlim=c(0,15)) + grid()
boxplot(fullPMNLE$loops, fullPMNLE$loops_all, ylim=c(0,30000)) + grid()
hist(fullPMNLE$inputProteins, breaks=111) + grid()
hist(log10(fullPMNLE$loops), breaks=111, col="red", xlim=c(3,5)) + grid()
hist(log10(fullPMNLE$loops_all), breaks=111, add=T, col="blue") + grid()

fullPMNLE$proteins_ccs_all = log10(fullPMNLE$inputProteins/fullPMNLE$ccs_all)
fullPMNLE$proteins_ccs_chrom = log10(fullPMNLE$Chrom_genes/fullPMNLE$ccs)

fullPMNLE5 <- subset(fullPMNLE, Plasmid_genes==0)
fullPMNLE5_high <- subset(fullPMNLE5, indirect_all > 6.9)
fullPMNLE5_low <- subset(fullPMNLE5, indirect_all <= 6.9)

dim(fullPMNLE5) # 3872
dim(fullPMNLE4) # 489
t.test(fullPMNLE4$indirect_all, fullPMNLE5$indirect_all) # 0.8884783 5.7117042 
t.test(fullPMNLE4$ccs_all, fullPMNLE5$ccs_all) # 222.47010  40.47081 
t.test(fullPMNLE4$interactions, fullPMNLE5$interactions) # 59285.07  66511.59 

p.adjust(2.2e-16, method="BH", 22) # 4.4e-15
r21 <- sqrt(summary(lm((fullPMNLE5_high$indirect_all) ~ log10(fullPMNLE5_high$interactions)))$adj.r.squared)  
r31 <- sqrt(summary(lm((fullPMNLE5_low$indirect_all) ~ log10(fullPMNLE5_low$interactions)))$adj.r.squared)  
r51 <- sqrt(summary(lm((fullPMNLE4$indirect_chrom) ~ log10(fullPMNLE4$interactions)))$adj.r.squared) # 0.712
r61 <- sqrt(summary(lm((fullPMNLE4$indirect_all) ~ log10(fullPMNLE4$interactions)))$adj.r.squared)   # 0.061

x5_a_1 <- ggplot(fullPMNLE4, aes(x=log10(fullPMNLE4$interactions),
                                          y=fullPMNLE4$indirect_all)) +
  geom_point(color ="blue", size=0.1, shape=1, (aes(alpha=0.1)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of PPIs per sample", limits =c(3.9,5.2)) +
  scale_y_continuous(name="Indirect connections per protein", limits =c(0,13))  + 
  geom_smooth(method='lm', col="blue") + 
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE4$STRING.bacteria.name),
    " samples",sep=""),x=0.01,y=.89,hjust=0, gp=gpar(col="blue",fontsize=10)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r=", round(r61,3),sep=""),
  x=0.01,y=.87, hjust=0, gp=gpar(col="blue", fontsize=10)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r2=", round(r61**2,3),sep=""),
  x=0.01,y=.85, hjust=0, gp=gpar(col="blue", fontsize=10))))  
  
x5_a_2 <- x5_a_1 + geom_point(data=fullPMNLE4,
        aes(x=log10(fullPMNLE4$interactions), y=fullPMNLE4$indirect_chrom, alpha=0.1),
             color ="red", size=0.1, shape=2, show.legend=F) + 
  geom_smooth(data=fullPMNLE4,
        aes(x=log10(fullPMNLE4$interactions), y=fullPMNLE4$indirect_chrom),
        method='lm', col="red") +
  annotation_custom(grobTree(textGrob(paste("Chrom proteins adj_r=", round(r51,3),sep=""),
  x=0.01,y=.82, hjust=0, gp=gpar(col="red", fontsize=10)))) +
  annotation_custom(grobTree(textGrob(paste("Chrom proteins adj_r2=", round(r51**2,3),sep=""),
  x=0.01,y=.79, hjust=0, gp=gpar(col="red", fontsize=10)))) 

x5_b_1 <- ggplot(fullPMNLE5, aes(x=log10(fullPMNLE5$interactions),
                                          y=fullPMNLE5$indirect_all)) +
  geom_point(color ="darkgreen", size=0.1, shape=1, (aes(alpha=0.1)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of PPIs per sample", limits =c(3.9,5.2)) +
  scale_y_continuous(name="Indirect connections per protein", limits =c(0,13))  + 
    geom_smooth(data=fullPMNLE5_high,
        aes(x=log10(fullPMNLE5_high$interactions), y=fullPMNLE5_high$indirect_all),
        method='lm', col="darkgreen") + 
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE5_high$STRING.bacteria.name),
    " samples",sep=""),x=0.1,y=.95,hjust=0, gp=gpar(col="darkgreen",fontsize=10)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r=", round(r21,3),sep=""),
  x=0.1,y=.92, hjust=0, gp=gpar(col="darkgreen", fontsize=10)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r2=", round(r21**2,3),sep=""),
  x=0.1,y=.89, hjust=0, gp=gpar(col="darkgreen", fontsize=10))))    + 
    geom_smooth(data=fullPMNLE5_low,
        aes(x=log10(fullPMNLE5_low$interactions), y=fullPMNLE5_low$indirect_all),
        method='lm', col="darkgreen") + 
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE5_low$STRING.bacteria.name),
    " samples",sep=""),x=0.61,y=.13,hjust=0, gp=gpar(col="darkgreen",fontsize=10)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r=", round(r31,3),sep=""),
  x=0.61,y=.1, hjust=0, gp=gpar(col="darkgreen", fontsize=10)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r2=", round(r31**2,3),sep=""),
  x=0.61,y=.07, hjust=0, gp=gpar(col="darkgreen", fontsize=10))))  

aa1 <- ggExtra::ggMarginal(x5_a_2, type="densigram",size=5,colour="grey", bins=50)  
aa2 <- ggExtra::ggMarginal(x5_b_1, type="densigram",size=5,colour="grey", bins=50)  

pdf("Fig_5_IndirectConnections.pdf", width=12, height=6, bg="white")
ggarrange(aa1, aa2, ncol=2, nrow=1)
dev.off()

####
dim(subset(fullPMNLE, indirect_all >= 7)) # 1302
dim(subset(fullPMNLE, indirect_all <= 7 & indirect_all > 1.5)) # 2299
dim(subset(fullPMNLE, indirect_all <= 1.5)) # 637
dim(subset(fullPMNLE, indirect_chrom >= 7)) # 1328
dim(subset(fullPMNLE, indirect_chrom < 7)) # 2886
fullPMNLE_high <- subset(fullPMNLE4, indirect_all > 7 ) # 1302 (31%)
fullPMNLE_mid <- subset(fullPMNLE, indirect_all <= 7 & indirect_all > 1.5 ) # 2299 (54%)
fullPMNLE_low <- subset(fullPMNLE, indirect_all < 1.5)  # 637 (15%) # 

r21 <- sqrt(summary(lm((fullPMNLE_high$indirect_all) ~ log10(fullPMNLE_high$interactions)))$adj.r.squared) # 0.36
r41 <- sqrt(summary(lm((fullPMNLE_mid$indirect_all) ~ log10(fullPMNLE_mid$interactions)))$adj.r.squared)   # 0.671
r51 <- sqrt(summary(lm((fullPMNLE_low$indirect_chrom) ~ log10(fullPMNLE_low$interactions)))$adj.r.squared) # 0.712
r61 <- sqrt(summary(lm((fullPMNLE_low$indirect_all) ~ log10(fullPMNLE_low$interactions)))$adj.r.squared)   # 0.061

x5_a_1 <- ggplot(fullPMNLE_low, aes(x=log10(fullPMNLE_low$interactions),
                                          y=fullPMNLE_low$indirect_all)) +
  geom_point(color ="blue", size=0.1, shape=1, (aes(alpha=0.1)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of PPIs per sample", limits =c(3.9,5.2)) +
  scale_y_continuous(name="Indirect connections per protein", limits =c(0,13))  + 
  geom_smooth(method='lm', col="blue") + 
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE_low$STRING.bacteria.name),
    " samples",sep=""),x=0.01,y=.89,hjust=0, gp=gpar(col="blue",fontsize=10)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r=", round(r61,3),sep=""),
  x=0.01,y=.87, hjust=0, gp=gpar(col="blue", fontsize=10)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r2=", round(r61**2,3),sep=""),
  x=0.01,y=.85, hjust=0, gp=gpar(col="blue", fontsize=10))))  
  
x5_a_2 <- x5_a_1 + geom_point(data=fullPMNLE_low,
        aes(x=log10(fullPMNLE_low$interactions), y=fullPMNLE_low$indirect_chrom, alpha=0.1),
             color ="red", size=0.1, shape=2, show.legend=F) + 
  geom_smooth(data=fullPMNLE_low,
        aes(x=log10(fullPMNLE_low$interactions), y=fullPMNLE_low$indirect_chrom),
        method='lm', col="red") +
  annotation_custom(grobTree(textGrob(paste("Chrom proteins adj_r=", round(r51,3),sep=""),
  x=0.01,y=.83, hjust=0, gp=gpar(col="red", fontsize=10)))) +
  annotation_custom(grobTree(textGrob(paste("Chrom proteins adj_r2=", round(r51**2,3),sep=""),
  x=0.01,y=.81, hjust=0, gp=gpar(col="red", fontsize=10)))) 

x5_b_1 <- ggplot(fullPMNLE_mid, aes(x=log10(fullPMNLE_mid$interactions),
                                          y=fullPMNLE_mid$indirect_all)) +
  geom_point(color ="darkgreen", size=0.1, shape=1, (aes(alpha=0.1)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of PPIs per sample", limits =c(3.9,5.2)) +
  scale_y_continuous(name="Indirect connections per protein", limits =c(0,13))  + 
  geom_smooth(method='lm', col="darkgreen") + 
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE_mid$STRING.bacteria.name),
    " samples",sep=""),x=0.61,y=.15,hjust=0, gp=gpar(col="darkgreen",fontsize=10)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r=", round(r41,3),sep=""),
  x=0.61,y=.13, hjust=0, gp=gpar(col="darkgreen", fontsize=10)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r2=", round(r41**2,3),sep=""),
  x=0.61,y=.11, hjust=0, gp=gpar(col="darkgreen", fontsize=10))))  

x5_c_1 <- x5_b_1 + geom_point(data=fullPMNLE_high,
        aes(x=log10(fullPMNLE_high$interactions), y=fullPMNLE_high$indirect_all, alpha=0.1),
             color ="red", size=0.1, shape=2, show.legend=F) + 
  geom_smooth(data=fullPMNLE_high,
        aes(x=log10(fullPMNLE_high$interactions), y=fullPMNLE_high$indirect_all),
        method='lm', col="red") + 
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE_high$STRING.bacteria.name),
    " samples",sep=""),x=0.01,y=.85,hjust=0, gp=gpar(col="red",fontsize=10)))) +
  annotation_custom(grobTree(textGrob(paste("Chrom proteins adj_r=", round(r51,3),sep=""),
  x=0.01,y=.83, hjust=0, gp=gpar(col="red", fontsize=10)))) +
  annotation_custom(grobTree(textGrob(paste("Chrom proteins adj_r2=", round(r51**2,3),sep=""),
  x=0.01,y=.81, hjust=0, gp=gpar(col="red", fontsize=10)))) 

pdf("Fig_5_IndirectConnections.pdf", width=12, height=6, bg="white")
ggarrange(x5_a_2, x5_c_1, ncol=2, nrow=1)
dev.off()

### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###
### 
r1 <- sqrt(summary(lm((fullPMNLE$indirect_chrom) ~ log10(fullPMNLE$interactions)))$adj.r.squared) # 0.36
r2 <- sqrt(summary(lm((fullPMNLE$indirect_all) ~ log10(fullPMNLE$interactions)))$adj.r.squared) # 0.36
figure5_a_1 <- ggplot(fullPMNLE, aes(x=log10(fullPMNLE$interactions),
                                          y=fullPMNLE$indirect_all)) +
  geom_point(color ="blue", size=0.1, shape=1, (aes(alpha=0.1)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of PPIs per sample", limits =c(3.9,5.4)) +
  scale_y_continuous(name="Indirect connections per protein", limits =c(0,15))  + 
  geom_smooth(method='lm', col="blue") + 
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE$STRING.bacteria.name),
    " bacterial samples",sep=""),x=0.01,y=.89,hjust=0, gp=gpar(col="blue",fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r=", round(r1,3),sep=""),
  x=0.01,y=.87, hjust=0, gp=gpar(col="blue", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r2=", round(r1**2,3),sep=""),
  x=0.01,y=.85, hjust=0, gp=gpar(col="blue", fontsize=7))))  
  
figure5_a_2 <- figure5_a_1 + geom_point(data=fullPMNLE,
        aes(x=log10(fullPMNLE$interactions), y=fullPMNLE$indirect_chrom, alpha=0.1),
             color ="red", size=0.1, shape=2, show.legend=F) + 
  geom_smooth(data=fullPMNLE,
        aes(x=log10(fullPMNLE$interactions), y=fullPMNLE$indirect_chrom),
        method='lm', col="red") +
  annotation_custom(grobTree(textGrob(paste("Chrom proteins adj_r=", round(r2,3),sep=""),
  x=0.01,y=.83, hjust=0, gp=gpar(col="red", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Chrom proteins adj_r2=", round(r2**2,3),sep=""),
  x=0.01,y=.81, hjust=0, gp=gpar(col="red", fontsize=7)))) 

### Proteins
r1p <- sqrt(summary(lm((fullPMNLE$indirect_chrom) ~ log10(fullPMNLE$inputProteins)))$adj.r.squared)
r2p <- sqrt(summary(lm((fullPMNLE$indirect_all) ~ log10(fullPMNLE$inputProteins)))$adj.r.squared)

figure5_b_1 <- ggplot(fullPMNLE, aes(x=log10(fullPMNLE$inputProteins),
                                          y=fullPMNLE$indirect_all)) +
  geom_point(color ="blue", size=0.1, shape=1, (aes(alpha=0.1)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of proteins per sample", limits =c(2.7,4)) +
  scale_y_continuous(name="Indirect connections per protein", limits =c(0,15))  + 
  geom_smooth(method='lm', col="blue") + 
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE$STRING.bacteria.name),
    " bacterial samples",sep=""),x=0.01,y=.89,hjust=0, gp=gpar(col="blue",fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r=", round(r2p,3),sep=""),
  x=0.01,y=.87, hjust=0, gp=gpar(col="blue", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r2=", round(r2p**2,3),sep=""),
  x=0.01,y=.85, hjust=0, gp=gpar(col="blue", fontsize=7))))  
  
figure5_b_2 <- figure5_b_1 + geom_point(data=fullPMNLE,
    aes(x=log10(fullPMNLE$inputProteins), y=fullPMNLE$indirect_chrom, alpha=0.1),
             color ="red", size=0.1, shape=2, show.legend=F) + 
  geom_smooth(data=fullPMNLE,
        aes(x=log10(fullPMNLE$inputProteins), y=fullPMNLE$indirect_chrom),
        method='lm', col="red") +
  annotation_custom(grobTree(textGrob(paste("Chrom proteins adj_r=", round(r1p,3),sep=""),
  x=0.01,y=.83, hjust=0, gp=gpar(col="red", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Chrom proteins adj_r2=", round(r1p**2,3),sep=""),
  x=0.01,y=.81, hjust=0, gp=gpar(col="red", fontsize=7)))) 

### Chrom_PPI -> Fig_S22.pdf
r1c <- sqrt(summary(lm((fullPMNLE$indirect_chrom) ~ log10(fullPMNLE$Chrom_PPI)))$adj.r.squared)
r2c <- sqrt(summary(lm((fullPMNLE$indirect_all) ~ log10(fullPMNLE$Chrom_PPI)))$adj.r.squared)

figure5_c_1 <- ggplot(fullPMNLE, aes(x=log10(fullPMNLE$Chrom_PPI),
                                          y=fullPMNLE$indirect_all)) +
  geom_point(color ="blue", size=0.1, shape=1, (aes(alpha=0.1)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of chromosomal PPIs per sample", limits =c(3.9,5.4)) +
  scale_y_continuous(name="Indirect connections per protein", limits =c(0,15))  + 
  geom_smooth(method='lm', col="blue") + 
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE$STRING.bacteria.name),
    " bacterial samples",sep=""),x=0.01,y=.89,hjust=0, gp=gpar(col="blue",fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r=", round(r2c,3),sep=""),
  x=0.02, y=.87, hjust=0, gp=gpar(col="blue", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r2=", round(r2c**2,3),sep=""),
  x=0.02, y=.85, hjust=0, gp=gpar(col="blue", fontsize=7))))  
  
figure5_c_2 <- figure5_c_1 + geom_point(data=fullPMNLE,
        aes(x=log10(fullPMNLE$Chrom_PPI), y=fullPMNLE$indirect_chrom, alpha=0.1),
             color ="red", size=0.1, shape=2, show.legend=F) + 
  geom_smooth(data=fullPMNLE,
        aes(x=log10(fullPMNLE$Chrom_PPI), y=fullPMNLE$indirect_chrom),
        method='lm', col="red") +
  annotation_custom(grobTree(textGrob(paste("Chrom proteins adj_r=", round(r1c,3),sep=""),
  x=0.02, y=.83, hjust=0, gp=gpar(col="red", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Chrom proteins adj_r2=", round(r1c**2,3),sep=""),
  x=0.02, y=.81, hjust=0, gp=gpar(col="red", fontsize=7)))) 

### Plasmid_PPI
r1d <- sqrt(summary(lm((fullPMNLE$indirect_chrom) ~ log10(fullPMNLE$Plasmid_PPI)))$adj.r.squared)
r2d <- sqrt(summary(lm((fullPMNLE$indirect_all) ~ log10(fullPMNLE$Plasmid_PPI)))$adj.r.squared)

figure5_d_1 <- ggplot(fullPMNLE, aes(x=log10(fullPMNLE$Plasmid_PPI),
                                          y=fullPMNLE$indirect_all)) +
  geom_point(color ="blue", size=0.1, shape=1, (aes(alpha=0.1)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of plasmid PPIs per sample", limits =c(1,4.5)) +
  scale_y_continuous(name="Indirect connections per protein", limits =c(0,15))  + 
  geom_smooth(method='lm', col="blue") + 
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE$STRING.bacteria.name),
    " bacterial samples",sep=""),x=0.01,y=.89,hjust=0, gp=gpar(col="blue",fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r=", round(r2d,3),sep=""),
  x=0.01,y=.87, hjust=0, gp=gpar(col="blue", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r2=", round(r2d**2,3),sep=""),
  x=0.01,y=.85, hjust=0, gp=gpar(col="blue", fontsize=7))))  
  
figure5_d_2 <- figure5_d_1 + geom_point(data=fullPMNLE,
        aes(x=log10(fullPMNLE$Plasmid_PPI), y=fullPMNLE$indirect_chrom, alpha=0.1),
             color ="red", size=0.1, shape=2, show.legend=F) + 
  geom_smooth(data=fullPMNLE,
        aes(x=log10(fullPMNLE$Plasmid_PPI), y=fullPMNLE$indirect_chrom),
        method='lm', col="red") +
  annotation_custom(grobTree(textGrob(paste("Chrom proteins adj_r=", round(r1d,3),sep=""),
  x=0.01,y=.83, hjust=0, gp=gpar(col="red", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Chrom proteins adj_r2=", round(r1d**2,3),sep=""),
  x=0.01,y=.81, hjust=0, gp=gpar(col="red", fontsize=7)))) 

#####
figure5_a1 <- ggExtra::ggMarginal(figure5_a_2, type="densigram",
                    size=5,colour="grey", bins=50)  
figure5_b1 <- ggExtra::ggMarginal(figure5_b_2, type="densigram",
                    size=5,colour="grey", bins=50)  
figure5_c1 <- ggExtra::ggMarginal(figure5_c_2, type="densigram",
                    size=5,colour="grey", bins=50)  
figure5_d1 <- ggExtra::ggMarginal(figure5_d_2, type="densigram",
                    size=5,colour="grey", bins=50)  
par(mfrow=c(2,2))
pdf("Fig_S19.pdf", width=10, height=12, bg="white")
ggarrange(figure5_a1, figure5_b1, figure5_c1, figure5_d1, ncol=2, nrow=2)
dev.off()

########################################################################
########### Numbers of connections vs proteins & other metadata -> Fig_S22 ##
########################################################################

fullPMNLE$proteins_ccs_all = log10(fullPMNLE$inputProteins/fullPMNLE$ccs_all)
fullPMNLE$proteins_ccs_chrom = log10(fullPMNLE$Chrom_genes/fullPMNLE$ccs)
fullPMNLE[mapply(is.infinite, fullPMNLE)] <- NA

r1 <- sqrt(summary(lm((fullPMNLE$proteins_ccs_chrom) ~ log10(fullPMNLE$interactions)))$adj.r.squared)
r2 <- sqrt(summary(lm((fullPMNLE$proteins_ccs_all) ~ log10(fullPMNLE$interactions)))$adj.r.squared)

figure5_a_1 <- ggplot(fullPMNLE, aes(x=log10(fullPMNLE$interactions),
                                          y=fullPMNLE$proteins_ccs_all)) +
  geom_point(color ="black", size=0.1, shape=1, (aes(alpha=0.1)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of PPIs per sample", limits =c(4,5.5) ) +
  scale_y_continuous(name="Proteins per connected component", limits =c(1,3))  + 
  geom_smooth(method='lm', col="black") + 
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE$STRING.bacteria.name),
    " bacterial samples",sep=""),x=0.01,y=.89,hjust=0, gp=gpar(col="black",fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r=", round(r2,3),sep=""),
  x=0.01,y=.87, hjust=0, gp=gpar(col="black", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r2=", round(r2**2,3),sep=""),
  x=0.01,y=.85, hjust=0, gp=gpar(col="black", fontsize=7))))  
  
figure5_a_2 <- figure5_a_1 + geom_point(data=fullPMNLE,
        aes(x=log10(fullPMNLE$interactions), y=fullPMNLE$proteins_ccs_chrom, alpha=0.1),
             color ="red", size=0.1, shape=2, show.legend=F) + 
  geom_smooth(data=fullPMNLE,
        aes(x=log10(fullPMNLE$interactions), y=fullPMNLE$proteins_ccs_chrom),
        method='lm', col="red") +
  annotation_custom(grobTree(textGrob(paste("Chrom proteins adj_r=", round(r1,3),sep=""),
  x=0.01,y=.83, hjust=0, gp=gpar(col="red", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Chrom proteins adj_r2=", round(r1**2,3),sep=""),
  x=0.01,y=.81, hjust=0, gp=gpar(col="red", fontsize=7)))) 

### Proteins
r1p <- sqrt(summary(lm((fullPMNLE$proteins_ccs_chrom) ~ log10(fullPMNLE$inputProteins)))$adj.r.squared)
r2p <- sqrt(summary(lm((fullPMNLE$proteins_ccs_all) ~ log10(fullPMNLE$inputProteins)))$adj.r.squared)

figure5_b_1 <- ggplot(fullPMNLE, aes(x=log10(fullPMNLE$inputProteins),
                                          y=fullPMNLE$proteins_ccs_all)) +
  geom_point(color ="black", size=0.1, shape=1, (aes(alpha=0.1)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of proteins per sample", limits =c(2.8,4)) +
  scale_y_continuous(name="Proteins per connected component", limits =c(1,3))  + 
  geom_smooth(method='lm', col="black") + 
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE$STRING.bacteria.name),
    " bacterial samples",sep=""),x=0.01,y=.89,hjust=0, gp=gpar(col="black",fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r=", round(0.00,3),sep=""),
  x=0.01,y=.87, hjust=0, gp=gpar(col="black", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r2=", round(0.00,3),sep=""),
  x=0.01,y=.85, hjust=0, gp=gpar(col="black", fontsize=7))))  
  
figure5_b_2 <- figure5_b_1 + geom_point(data=fullPMNLE,
    aes(x=log10(fullPMNLE$inputProteins), y=fullPMNLE$proteins_ccs_chrom, alpha=0.1),
             color ="red", size=0.1, shape=2, show.legend=F) + 
  geom_smooth(data=fullPMNLE,
        aes(x=log10(fullPMNLE$inputProteins), y=fullPMNLE$proteins_ccs_chrom),
        method='lm', col="red") +
  annotation_custom(grobTree(textGrob(paste("Chrom proteins adj_r=", round(r1p,3),sep=""),
  x=0.01,y=.83, hjust=0, gp=gpar(col="red", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Chrom proteins adj_r2=", round(r1p**2,3),sep=""),
  x=0.01,y=.81, hjust=0, gp=gpar(col="red", fontsize=7)))) 

### Chrom_PPI
r1c <- sqrt(summary(lm((fullPMNLE$proteins_ccs_chrom) ~ log10(fullPMNLE$Chrom_PPI)))$adj.r.squared)
r2c <- sqrt(summary(lm((fullPMNLE$proteins_ccs_all) ~ log10(fullPMNLE$Chrom_PPI)))$adj.r.squared)

figure5_c_1 <- ggplot(fullPMNLE, aes(x=log10(fullPMNLE$Chrom_PPI),
                                          y=fullPMNLE$proteins_ccs_all)) +
  geom_point(color ="black", size=0.1, shape=1, (aes(alpha=0.1)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of chromosomal PPIs per sample", limits =c(4,5.5)) +
  scale_y_continuous(name="Proteins per connected component", limits =c(1,3))  + 
  geom_smooth(method='lm', col="black") + 
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE$STRING.bacteria.name),
    " bacterial samples",sep=""),x=0.01,y=.89,hjust=0, gp=gpar(col="black",fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r=", round(r2c,3),sep=""),
  x=0.02, y=.87, hjust=0, gp=gpar(col="black", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r2=", round(r2c**2,3),sep=""),
  x=0.02, y=.85, hjust=0, gp=gpar(col="black", fontsize=7))))  
  
figure5_c_2 <- figure5_c_1 + geom_point(data=fullPMNLE,
        aes(x=log10(fullPMNLE$Chrom_PPI), y=fullPMNLE$proteins_ccs_chrom, alpha=0.1),
             color ="red", size=0.1, shape=2, show.legend=F) + 
  geom_smooth(data=fullPMNLE,
        aes(x=log10(fullPMNLE$Chrom_PPI), y=fullPMNLE$proteins_ccs_chrom),
        method='lm', col="red") +
  annotation_custom(grobTree(textGrob(paste("Chrom proteins adj_r=", round(r1c,3),sep=""),
  x=0.02, y=.83, hjust=0, gp=gpar(col="red", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Chrom proteins adj_r2=", round(r1c**2,3),sep=""),
  x=0.02, y=.81, hjust=0, gp=gpar(col="red", fontsize=7)))) 

### Plasmid_PPI
r1d <- sqrt(summary(lm((fullPMNLE$proteins_ccs_chrom) ~ log10(fullPMNLE$Plasmid_PPI)))$adj.r.squared)
r2d <- sqrt(summary(lm((fullPMNLE$proteins_ccs_all) ~ log10(fullPMNLE$Plasmid_PPI)))$adj.r.squared)

figure5_d_1 <- ggplot(fullPMNLE, aes(x=log10(fullPMNLE$Plasmid_PPI),
                                          y=fullPMNLE$proteins_ccs_all)) +
  geom_point(color ="black", size=0.1, shape=1, (aes(alpha=0.1)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of plasmid PPIs per sample", limits =c(1.1,4.5)) +
  scale_y_continuous(name="Proteins per connected component", limits =c(1,3))  + 
  geom_smooth(method='lm', col="black") + 
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE$STRING.bacteria.name),
    " bacterial samples",sep=""),x=0.01,y=.89,hjust=0, gp=gpar(col="black",fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r=", round(r2d,3),sep=""),
  x=0.02, y=.87, hjust=0, gp=gpar(col="black", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("All proteins adj_r2=", round(r2d**2,3),sep=""),
  x=0.02, y=.85, hjust=0, gp=gpar(col="black", fontsize=7))))  
  
figure5_d_2 <- figure5_d_1 + geom_point(data=fullPMNLE,
        aes(x=log10(fullPMNLE$Plasmid_PPI), y=fullPMNLE$proteins_ccs_chrom, alpha=0.1),
             color ="red", size=0.1, shape=2, show.legend=F) + 
  geom_smooth(data=fullPMNLE,
        aes(x=log10(fullPMNLE$Plasmid_PPI), y=fullPMNLE$proteins_ccs_chrom),
        method='lm', col="red") +
  annotation_custom(grobTree(textGrob(paste("Chrom proteins adj_r=", round(r1d,3),sep=""),
  x=0.02, y=.83, hjust=0, gp=gpar(col="red", fontsize=7)))) +
  annotation_custom(grobTree(textGrob(paste("Chrom proteins adj_r2=", round(r1d**2,3),sep=""),
  x=0.02, y=.81, hjust=0, gp=gpar(col="red", fontsize=7)))) 

#####
figure5_a1 <- ggExtra::ggMarginal(figure5_a_2, type="densigram",
                    size=5,colour="grey", bins=50)  
figure5_b1 <- ggExtra::ggMarginal(figure5_b_2, type="densigram",
                    size=5,colour="grey", bins=50)  
figure5_c1 <- ggExtra::ggMarginal(figure5_c_2, type="densigram",
                    size=5,colour="grey", bins=50)  
figure5_d1 <- ggExtra::ggMarginal(figure5_d_2, type="densigram",
                    size=5,colour="grey", bins=50)  
par(mfrow=c(2,2))
pdf("Fig_S20_2.pdf", width=10, height=12, bg="white")
ggarrange(figure5_a1, figure5_b1, figure5_c1, figure5_d1, ncol=2, nrow=2)
dev.off()
#png("Fig_5_PMNLE.png", width=2500, height=2600, units="px")
#ggarrange(figure5_a1, figure5_b1, figure5_c1, figure5_d1, ncol=2, nrow=2)
#dev.off()

###################
# Association of chrom PMNLE with numbers of input proteins is present 
plot(log10(fullPMNLE$inputProteins), fullPMNLE$aPMNLE) + grid()
abline(lm(fullPMNLE$aPMNLE ~ log10(fullPMNLE$inputProteins)),col="green")
cor.test(log10(fullPMNLE$inputProteins), fullPMNLE$aPMNLE)$estimate
cor.test(fullPMNLE$inputProteins, fullPMNLE$aPMNLE)[3]
cor.test(fullPMNLE$inputProteins, fullPMNLE$aPMNLE)$conf.int[1:2]
# inputProteins r=0.39  p=2.831362e-81  
pvalues <- c(2.831362e-81 , pvalues) # p value 14

plot(log10(fullPMNLE$inputProteins), fullPMNLE$aPMNLE_all) + grid()
abline(lm(fullPMNLE$aPMNLE_all ~ log10(fullPMNLE$inputProteins)),col="green")
cor.test(log10(fullPMNLE$inputProteins), fullPMNLE$aPMNLE_all)$estimate
cor.test(fullPMNLE$inputProteins, fullPMNLE$aPMNLE_all)[3]
cor.test(fullPMNLE$inputProteins, fullPMNLE$aPMNLE_all)$conf.int[1:2]
# inputProteins r=0.38 p=6.132256e-81  
pvalues <- c(6.132256e-81, pvalues) # p value 15

plot(log10(fullPMNLE$inputProteins), fullPMNLE$diffPMNLE) + grid()
abline(lm(fullPMNLE$diffPMNLE ~ log10(fullPMNLE$inputProteins)),col="green")
cor.test(log10(fullPMNLE$inputProteins), fullPMNLE$diffPMNLE)$estimate
cor.test(fullPMNLE$inputProteins, fullPMNLE$diffPMNLE)[3]
cor.test(fullPMNLE$inputProteins, fullPMNLE$diffPMNLE)$conf.int[1:2]
# inputProteins r=-0.07 p=0.0009648791
pvalues <- c(0.0009648791, pvalues) # p value 16
###################
#################################################################

########################################################################
########### Change in aPMNLE, not useful ###############################
########################################################################
########## % fall in from full PNMLE to partial PMNLE
par(mfrow=c(1,1))
Plasmid_PPI_dat3 <- fullPMNLE4
hist(Plasmid_PPI_dat3$diffPMNLE, breaks=123, col="grey", main="", 
     xlab="Difference in PMNLE", cex=1.4, cex.axis=1.4,  cex.lab=1.4) + grid()
summary(Plasmid_PPI_dat3$diffPMNLE)
sd(na.omit(Plasmid_PPI_dat3$diffPMNLE)) # SD = 0.077
#     Min.   1st Qu.    Median      Mean  3rd Qu.  Max  NAs
# -0.114881 -0.002151  0.024038  0.041272  0.067758  0.762518        30 

dim(Plasmid_PPI_dat3)  # 489
t3 <- subset(Plasmid_PPI_dat3, diffPMNLE >  2*sd(na.omit(Plasmid_PPI_dat3$diffPMNLE) ))
# Top 5% of difference in aPMNLE where plasmid has big effect -> lower connectivity
dim(t3) # 18 samples
View(t3[,c(1:8,12,16)] %>%  mutate_if(is.numeric, round, digits=3))
write.xlsx(t3, "Table_S6.xlsx")  
 
main_set1 <-  Plasmid_PPI_dat3[ !(Plasmid_PPI_dat3$Sample %in% t3$Sample), ]
main_set2 <-  main_set1 
dim(main_set2) # 471

# compare groups' features
par(mfrow=c(5,3))
pdf("mainset_v_small.pdf", width=6, height=6, bg="white")
p_set <- c()
t_set <- c()
for (i in 2:16){
   t3[,i] <- as.numeric(t3[,i])
   ttt <- t.test(as.numeric(t3[,i]), as.numeric(na.omit(main_set2[,i])))
   print(hist(as.numeric(main_set2[,i]), col=alpha("blue", 0.4), breaks=65,
              xlab=colnames(t3)[i], main=""))
   print(grid())
   print(hist(as.numeric(t3[,i]), col=alpha("red", 0.4) , breaks=65, add=T)) 
   p_set <- c(p_set, ttt$p.value)
   t_set <- c(t_set, ttt$statistic)
} # end for -> 11 items
dev.off()

# dim(main_set2)
output1 <- matrix(nrow=4, ncol=15)
colnames(output1) <- colnames(main_set2)[2:16]
output1[1,] <- (t_set)
output1[2,] <- p.adjust(p_set, method="BH")
output1[3,] <- colMeans(main_set2[2:16])
output1[4,] <- colMeans(t3[2:16])
rownames(output1) <- c("t_stat", "adj_p", "main", "new")
View(t(output1))

# compare connectec components as all +ve vs chrom -ve scaled by all
all_ccs <- subset(main_set2, ccs_all>0 & ccs>0) # 71.3% more
summary((all_ccs$ccs_all - all_ccs$ccs)/all_ccs$ccs_all)
#     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
# -0.07463  0.66667  0.77295  0.71312  0.85043  0.95227 
t3_ccs <- subset(t3, ccs_all>0 & ccs>0)
mean((t3_ccs$ccs_all - t3_ccs$ccs)/t3_ccs$ccs_all)
# 327.3% difference

# Although most showed statistically extreme differences between the 18 and
# the main group, these were artefactual and not real (see plots). The key one
# was the change from the aPMLNE for all proteins compared to chromosomal ones
# only (see below).

# compare groups' features
par(mfrow=c(3,1))
pdf("mainset_v_small.pdf", width=6, height=6, bg="white")
hist(as.numeric(main_set2[,8]), col=alpha("blue", 0.4), breaks=65,
              xlab="aPMNLE values for all proteins", main="") + grid()
hist(as.numeric(t3[,8]), col=alpha("red", 0.4), breaks=65, add=T)
hist(as.numeric(main_set2[,12]), col=alpha("blue", 0.4), breaks=65,
       xlab="aPMNLE values for chromosomal proteins", main="") + grid()
hist(as.numeric(t3[,12]), col=alpha("red", 0.4), breaks=65, add=T)
hist(as.numeric(main_set2[,16]), col=alpha("blue", 0.4), breaks=65,
              xlab="Difference in aPMNLE values", main="") + grid()
hist(as.numeric(t3[,16]), col=alpha("red", 0.4), breaks=65, add=T)
dev.off()

par(mfrow=c(1,1))
pdf("Figure_S19_diffPMNLE.pdf", width=6, height=6, bg="white")
# NEED to fix
plot(as.numeric(main_set2$aPMNLE_all), as.numeric(main_set2$aPMNLE), col=alpha("blue",
      0.4), cex=0.6, ylab="aPMNLE values for chromosomal proteins", main="",
      xlab="aPMNLE values for all proteins", xlim=c(0.48,0.61), ylim=c(0.11,0.65))
points(as.numeric(t3$aPMNLE_all), as.numeric(t3$aPMNLE), col=alpha("red",
      0.6), cex=0.6) + grid()
abline(lm(main_set2$aPMNLE ~ main_set2$aPMNLE_all), col=alpha("blue",
      0.4), lty="dotted")
#abline(lm(t3$aPMNLE ~ t3$aPMNLE_all), col=alpha("red", 0.4), lty="dotted")
text(0.526,0.64, paste0("For the robust set of ",length(main_set2$Sample)," samples",
                      sep=""), col=alpha("blue",0.7))
text(0.526,0.60, paste0("r=", round(cor.test(main_set2$aPMNLE, main_set2$aPMNLE_all)$estimate,3),
     sep=""),  col=alpha("blue",0.7))
text(0.526,0.18, paste0("For the less robust set of ",length(t3$Sample)," samples",
                      sep=""), col=alpha("red",0.7))
text(0.526,0.15, paste0("r=", round(cor(t3$aPMNLE, t3$aPMNLE_all),3), sep=""), col=alpha("red",0.7))
 dev.off()

# Examine metrics for whole group first
par(mfrow=c(1,1))
pdf("Figure_S14.pdf", width=7, height=7, bg="white")
xx = fullPMNLE4$loops_all/(fullPMNLE4$Chrom_genes + fullPMNLE4$Plasmid_genes)
yy = fullPMNLE4$loops/fullPMNLE4$Chrom_genes
plot(xx, yy, col=alpha("blue", 0.4), cex=0.2, main="", xlab="All: Indirect PPIs per protein",
    ylab="Chromosomal: Indirect PPIs per protein",  cex.lab=1.2, cex.axis=1.2) + grid()
abline(lm(yy ~ xx), col=alpha("blue", 0.6), lty="dotted")
abline(coef = c(0,1), col=alpha("red", 0.6), lty="dotted")
legend("topright", legend=c("Expected rate", "Observed rate"),
       col=c("red", "blue"), lty="dotted", cex=1.1)
text(27,2, "Samples with more indirect PPIs when",
     col=alpha("black",0.7), cex=1.1)
text(27,0.8, "plasmid-encoded proteins are included",
     col=alpha("black",0.7), cex=1.1)
text(14,33, "Samples with more indirect PPIs when",
     col=alpha("black",0.7), cex=1.1)
text(14,31.8, "plasmid-encoded proteins are excluded",
     col=alpha("black",0.7), cex=1.1)
dev.off()

summary(fullPMNLE4$loops_all )
#   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 #     0    2081    2888    3101    3853   14062 
sd(fullPMNLE4$loops_all ) # 14682.255
summary(fullPMNLE4$loops )
#   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 #    14    7408   12667   15446   20506   65171      27 
sd(na.omit(fullPMNLE4$loops) ) # 11217.89
t.test(fullPMNLE4$loops,fullPMNLE4$loops_all )

summary(fullPMNLE4$ccs_all )
#   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 # 7.0   155.0   227.0   222.5   282.0   494.0       4 
sd(na.omit(fullPMNLE4$ccs_all )) # 95.87443
summary(fullPMNLE4$ccs )
#   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 #    3.00   31.00   50.00   55.59   70.00  217.00      27 
sd(na.omit(fullPMNLE4$ccs) ) # 34.59954

summary(fullPMNLE4$Plasmid_Plasmid_PPI )
#   Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
 #   0.00     0.00     1.00    95.65     6.00 11065.00 


# plot change in genes vs change in ppis
# plot change in ppis vs change in indirect connections
# plot change in genes vs change in indirect connections
fullPMNLE4$loops[is.na(fullPMNLE4$loops)] <- 0
rr1 <- summary(lm(log10(fullPMNLE4$Plasmid_genes) ~
                (fullPMNLE4$Plasmid_PPI +fullPMNLE4$Plasmid_Plasmid_PPI)))$adj.r.squared # 0.07
f7a <- ggplot(fullPMNLE4, aes(x=log10(fullPMNLE4$Plasmid_genes),
                y=log10(fullPMNLE4$Plasmid_PPI+ fullPMNLE4$Plasmid_Plasmid_PPI))) +
  geom_point(color ="black", size=0.5, shape=1, (aes(alpha=0.3)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of plasmid genes per sample", limits=c(1,3.1)) +
  scale_y_continuous(name="Log10 of number of plasmid PPIs per sample", limits=c(2,4.5))  + 
  geom_smooth(method='lm', col="black") + 
  annotation_custom(grobTree(textGrob(paste("For ",length(fullPMNLE4$STRING.bacteria.name),
    " samples",sep=""),x=0.41,y=.09,hjust=0, gp=gpar(col="black",fontsize=11)))) +
  annotation_custom(grobTree(textGrob(paste("    adj_r=", round(rr1,2),sep=""),
  x=0.44, y=.06, hjust=0, gp=gpar(col="black", fontsize=11))))  
rr2 <- summary(lm(log10(fullPMNLE4$Plasmid_genes) ~
                (fullPMNLE4$loops_all -fullPMNLE4$loops)))$adj.r.squared # 0.06
f7b <- ggplot(fullPMNLE4, aes(x=log10(fullPMNLE4$Plasmid_genes),
            y= (fullPMNLE4$loops_all -fullPMNLE4$loops))) +
  geom_point(color ="blue", size=0.5, shape=1, (aes(alpha=0.3)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of plasmid genes per sample", limits=c(1,3.1)) +
  scale_y_continuous(name="Change in indirect connections due to plasmid-related proteins",
                     limits=c(-50000,4000))  + 
  geom_smooth(method='lm', col="blue") +  
  annotation_custom(grobTree(textGrob(paste("    adj_r=", round(rr2,3),sep=""),
  x=0.44, y=.06, hjust=0, gp=gpar(col="blue", fontsize=11))))  
rr3 <- summary(lm(log10(fullPMNLE4$Plasmid_PPI+ fullPMNLE4$Plasmid_Plasmid_PPI) ~
                (fullPMNLE4$loops_all -fullPMNLE4$loops)))$adj.r.squared # 0.08
f7c <- ggplot(fullPMNLE4, aes(x=log10(fullPMNLE4$Plasmid_PPI+ fullPMNLE4$Plasmid_Plasmid_PPI),
            y= (fullPMNLE4$loops_all -fullPMNLE4$loops))) +
  geom_point(color ="darkgreen", size=0.5, shape=1, (aes(alpha=0.5)), show.legend=F) +
  scale_x_continuous(name="Log10 of number of plasmid PPIs per sample",limits=c(1,4.5)) +
  scale_y_continuous(name="Change in indirect connections due to plasmid-related proteins",
                     limits=c(-50000,4000))  + 
  geom_smooth(method='lm', col="darkgreen") +  
  annotation_custom(grobTree(textGrob(paste("    adj_r=-", round(rr3,3),sep=""),
  x=0.44, y=.06, hjust=0, gp=gpar(col="darkgreen", fontsize=11))))  

f7aa <- ggExtra::ggMarginal(f7a, type="densigram",size=5,colour="grey", bins=50)  
f7bb <- ggExtra::ggMarginal(f7b, type="densigram",size=5,colour="grey", bins=50)  
f7cc <- ggExtra::ggMarginal(f7c, type="densigram",size=5,colour="grey", bins=50)  
pdf("Fig_SS.pdf", width=14, height=6, bg="white")
ggarrange(f7aa, f7bb, f7cc, ncol=3, nrow=1)
dev.off()

library(ppcor)
datatemp <- data.frame(log10(fullPMNLE4$Plasmid_genes),
            log10(fullPMNLE4$Plasmid_PPI+ fullPMNLE4$Plasmid_Plasmid_PPI),
            (fullPMNLE4$loops_all -fullPMNLE4$loops))
 (pcor(datatemp))
p.adjust(c(4.662631e-145, 3.889774e-01,2.476967e-02 ), method="BH", 24)

#####

xx = fullPMNLE$loops_all/fullPMNLE$inputProteins
yy = fullPMNLE$loops/fullPMNLE$Chrom_genes
xx = log10(fullPMNLE4$loops_all )
yy = log10(fullPMNLE4$loops )
pdf("Figure_6.pdf", width=13, height=7, bg="white")
par(mfrow=c(1,2))
plot(xx, yy, col=alpha("blue", 0.3), cex=0.5, main="",   cex.axis=1.2,
     xlab="All proteins: Log10 of indirect connections", ylim=c(3,5),xlim=c(3,4.6),
    ylab="Chromosomal proteins: Log10 of indirect connections", cex.lab=1.2) + grid()
abline(lm(yy ~ xx), col=alpha("blue", 1), lty="dashed", lwd=2)
abline(coef = c(0,1), col=alpha("red", 1), lty="dashed", lwd=2)
legend("bottomright", legend=c("Expected rate", "Observed rate"),
       col=c("red", "blue"), lty="dashed", cex=1.2)
text(4.2,3.5, "Samples with more indirect", col=alpha("black",0.7), cex=1.1)
text(4.15,3.415, "connections when plasmid-encoded", col=alpha("black",0.7), cex=1.1)
text(4.2,3.33, "proteins are included", col=alpha("black",0.7), cex=1.1)
text(3.45,5.03, "Samples with more indirect", col=alpha("black",0.7), cex=1.1)
text(3.45,4.96, "connections when plasmid-encoded", col=alpha("black",0.7), cex=1.1)
text(3.45,4.88, "proteins are excluded", col=alpha("black",0.7), cex=1.1)

xx2 = log10(fullPMNLE4$ccs_all)
yy2 = log10(fullPMNLE4$ccs)
temp2 <- (na.omit(data.frame(xx2,yy2)))
temp2[mapply(is.infinite, temp2)] <- NA
plot(xx2, yy2, col=alpha("darkgreen", 0.3), cex=0.5, main="", cex.lab=1.2, xlim=c(1,2.7),
     xlab="All: log10-scaled number of connected components", cex.axis=1.2, ylim=c(1,2.7),
     ylab="Chromosomal: log10-scaled number of connected components") + grid()
abline(lm(temp2$yy2 ~ temp2$xx2), col=alpha("darkgreen", 1), lty="dashed", lwd=2)
abline(coef = c(0,1), col=alpha("red", 1), lty="dashed", lwd=2)
legend("topleft", legend=c("Expected rate", "Observed rate"),
       col=c("red", "darkgreen"), lty="dashed", cex=1.2)
text(2.25,1.11, "Samples with more connected", col=alpha("black",0.7), cex=1.1)
text(2.2,1.05, "components when plasmid-encoded", col=alpha("black",0.7), cex=1.1)
text(2.25,0.99, "proteins are included", col=alpha("black",0.7), cex=1.1)
text(1.5,2.4, "Samples with more connected", col=alpha("black",0.7), cex=1.1)
text(1.55, 2.34, "components when plasmid-encoded", col=alpha("black",0.7), cex=1.1)
text(1.5,2.28, "proteins are excluded", col=alpha("black",0.7), cex=1.1)
dev.off()

xx2 = log10(fullPMNLE4$ccs_all)
yy2 = log10(fullPMNLE4$ccs)
temp2 <- (na.omit(data.frame(xx2,yy2)))
temp2[mapply(is.infinite, temp2)] <- NA
plot(xx2, yy2, col=alpha("darkgreen", 0.3), cex=0.5, main="", cex.lab=1.2, xlim=c(1,2.7),
     xlab="All: log10-scaled number of connected components", cex.axis=1.2, ylim=c(1,2.7),
     ylab="Chromosomal: log10-scaled number of connected components") + grid()
abline(lm(temp2$yy2 ~ temp2$xx2), col=alpha("darkgreen", 1), lty="dashed", lwd=2)
abline(coef = c(0,1), col=alpha("red", 1), lty="dashed", lwd=2)
legend("topleft", legend=c("Expected rate", "Observed rate"),
       col=c("red", "darkgreen"), lty="dashed", cex=1.2)
text(2.25,1.11, "Samples with more connected", col=alpha("black",0.7), cex=1.1)
text(2.2,1.05, "components when plasmid-encoded", col=alpha("black",0.7), cex=1.1)
text(2.25,0.99, "proteins are included", col=alpha("black",0.7), cex=1.1)
text(1.5,2.4, "Samples with more connected", col=alpha("black",0.7), cex=1.1)
text(1.55, 2.34, "components when plasmid-encoded", col=alpha("black",0.7), cex=1.1)
text(1.5,2.28, "proteins are excluded", col=alpha("black",0.7), cex=1.1)
##

########################################################################
########### plasmid-associated genes & per sample rates ###############################
########################################################################
setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/")
dim(new) # check
pdf("Key_Results_Figures/Figure_3_hist_plasmidgenes.pdf", width=10, height=6, bg="white")
par(mfrow=c(1,2)) # Gene_rates$Plasmid_genes
hist.data <- hist(colSums(new), plot=F)
hist.data$counts <- log10(hist.data$counts + 1)
plot(hist.data, breaks=99, col="red", main="", cex=1.1,
     ylab="Log10-scaled number of plasmid-encoded genes",  
     xlab="Number of samples with each plasmid-encoded gene") + grid() 
hist.data <- hist(Gene_rates$Plasmid_genes, plot=F)
hist.data$counts <- log10(hist.data$counts +1)
plot(hist.data, breaks=111, col="lightgray", main="",
     ylab="Log10-scaled number of samples", ylim=c(0,4),
     xlab="Numbers of plasmid-encoded genes per sample", cex=1.1) + grid() 
dev.off()

rownames(new) <- str_remove(rownames(new), "\\.") # seems to only remove one at a time!?
rownames(new) <- str_remove(rownames(new), "\\.") # seems to only remove one at a time!?
rownames(new) <- str_remove(rownames(new), "\\:") # seems to only remove one at a time!?
rownames(new) <- str_remove(rownames(new), "\\:") # seems to only remove one at a time!?

plasmidPPI <- data.frame(new[ fullPMNLE$Sample %in% rownames(new), ])
plasmidPPI[1:3,1:4]
dim(plasmidPPI) # should be 4433 9172

rows1_PPI <- data.frame(rowSums(plasmidPPI)) #  plasmid genes per sample
summary(rows1_PPI$rowSums.plasmidPPI.)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
sums_PPI <- data.frame(colSums(plasmidPPI))  
logsums_PPI <- log10(sums_PPI$colSums.plasmidPPI.) # plasmid genes 

######## heatmap ######
# do heatmap of plasmid genes with matches only
# install.packages("gplots")
library(gplots)
dim(new) # 497 samples by 3023 genes

#heatmap.2(as.matrix(new[1:100,1:100]))
pdf("Figure_S5_heatmap_plasmidgenes.pdf", width=36, height=36, bg="white")
heatmap.2(as.matrix(new), dendrogram="both", trace="none",
          margin=c(5,16), cex.lab=0.1, cex.axis=0.1, col=hcl.colors(10,"BluYl"))
dev.off()

#heatmap.2(as.matrix(new[1:100,1:100]))
pdf("Figure_S5_heatmap_plasmidgenes.interactions.3.pdf", width=36, height=36, bg="white")
heatmap.2(as.matrix(log10(new_int +1)), dendrogram="both", trace="none",  
          margin=c(5,16), cex.lab=0.1, cex.axis=0.1, col=redblue(100))
dev.off()
pdf("Figure_S5_heatmap_plasmidgenes.interactions.col.pdf", width=36, height=36, bg="white")
heatmap.2(as.matrix( (new_int +1)), dendrogram="both", trace="none", scale=c("column"),
          margin=c(5,16), cex.lab=0.1, cex.axis=0.1, col=redblue(100))
dev.off()

#### get dendrograms of samples col=redblue(100)
new_map <- heatmap.2(as.matrix(new), dendrogram="both", trace="none",
          margin=c(8,11), cex.lab=0.2, cex.axis=0.2, col=hcl.colors(50,"BluYl"))
str(new_map)
plot(new_map$rowDendrogram)
plot(new_map$colDendrogram)

########################################################################
# compare groups
# new_3 <- new_2[-c(67, 87, 174, 204, 259, 264, 272),] #without E. coli

new_33 <- new[-c(165:170,361:363),]  
dim(new_33) # 485 samples x 3023 genes
median(colSums(new_33)) # 1 median number of samples in which each gene is found
IQR(colSums(new_33))    # 2 IQR of number of samples in which each gene is found
median(rowSums(new_33)) # 5 median genes per sample
IQR(rowSums(new_33))    # 11 IQR of number of genes per sample
summary(rowSums(new_33)) # mean =14.4 +- 28.6
sd(rowSums(new_33))

# E. coli
new_4 <- new[c(165:170,361:363),] # only E. coli
#new_4 <- subset(new_4, select = -c(rows1))
new_44 <- new_4 %>% select_if(colSums(.) > 0)
median(colSums(new_44)) # 4 median number of samples in which each gene is found
median(rowSums(new_44)) # 1035 median genes per sample
summary(rowSums(new_44)) # 931 

dim(new_44) # 6 samples x 1438 genes
pca_proc3333 <- prcomp((new_44)) # across genes # All_PCA_Viz_AcrossSamples 
outall_33 <- fviz_pca_ind(pca_proc3333, col.ind="cos2", repel=T, cex=0.1,
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"))+
  theme(legend.position = "none") + labs(tag="A") + 
  ggtitle("All: PC1 v PC2 across E. coli")
pca_proc555 <- prcomp(t(new_44)) # All_PCA_Viz_AcrossGenes
outall_44 <- fviz_pca_ind(pca_proc555, col.ind="cos2", repel=T, cex=0.1,
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"))+
  theme(legend.position = "none") + labs(tag="B") + 
  ggtitle("All: PC1 v PC2 across genes  - E. coli")
png("../Key_Results_Figures/Figure_S4_All_PCA_justEcoli.png", width=1311,height=911,
    bg="white", units="px")
  par(mfrow = c(1, 2))
  multiplot(outall_33, outall_44)
  dev.off()
  
  # need to exclude E. coli
pca_proc_3 <- prcomp((new)) # across genes # All_PCA_Viz_AcrossSamples 
outall_3_3 <- fviz_pca_ind(pca_proc_3, col.ind="cos2", repel=T, cex=0.1,
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"))+
  theme(legend.position = "none") + labs(tag="A") + 
  ggtitle("All: PC1 v PC2 across samples")
pca_proc_5 <- prcomp(t(new)) # All_PCA_Viz_AcrossGenes
outall_4_4 <- fviz_pca_ind(pca_proc_5, col.ind="cos2", repel=T, cex=0.1,
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"))+
  theme(legend.position = "none") + labs(tag="B") + 
  ggtitle("All: PC1 v PC2 across genes")
png("../Key_Results_Figures/Figure_S4_All_PCA.png", width=1311,height=911,
    bg="white", units="px")
  par(mfrow = c(1, 2))
  multiplot(outall_3_3, outall_4_4)
  dev.off()
  
  # plasmid-related genes per sample

########### giant heatmap of plasmid-related genes #####################
############ check sorted samples
new_sorted <- new[ order(row.names(new)), ]
pdf("heatmap_plasmidgenes.sorted.pdf", width=150, height=70, bg="white")
heatmap.2(as.matrix(new_sorted), dendrogram="none", trace="none",
          Rowv=NULL, Colv=NULL, key=F, keysize=0, margin=c(4,14),
          col=hcl.colors(10,"BluYl"), cexRow=0.1, cexCol=0.1)
dev.off()

########################################################################
########### PCA ################################## #####################
########################################################################

pca_proc5 <- prcomp(t(new_33)) # across genes
pcvalue1 = round(100*pca_proc5$sdev[1]^2/(sum(pca_proc5$sdev^2)),1)
pcvalue2 = round(100*pca_proc5$sdev[2]^2/(sum(pca_proc5$sdev^2)),1)
pcvalue1hw = 30*pcvalue1 # pixel PC1
pcvalue2hw = 30*pcvalue2 # pixel PC2
mycol<- c(rainbow(dim(new_33)[2], start=1, end=.1)) # group samples
png(file="Key_Results_Figures/All_PCA_1v2_AcrossGenes.png", width=pcvalue1hw*5,
    height=pcvalue2hw*5, units="px")
par(mar = c(5, 5, 5, 5))
plot(pca_proc5$x[,1], pca_proc5$x[,2], col=mycol, xlab=pcvalue1, ylab=pcvalue2,
     main='All: PC1 v PC2 across genes', cex=1, cex.axis=2,
     cex.lab=2) + grid()
dev.off() #  

# All across samples 
pca_proc33 <- prcomp((new_33)) # across samples
pcvalue1 = round(100*pca_proc33$sdev[1]^2/(sum(pca_proc33$sdev^2)),1)
pcvalue2 = round(100*pca_proc33$sdev[2]^2/(sum(pca_proc33$sdev^2)),1)
pcvalue3 = round(100*pca_proc33$sdev[3]^2/(sum(pca_proc33$sdev^2)),1)
pcvalue1hw = 30*pcvalue1 # pixel PC1
pcvalue2hw = 30*pcvalue2 # pixel PC2
pcvalue3hw = 30*pcvalue3 # pixel PC3
mycol<- c(rainbow(dim(new_33)[1], start=1, end=.1)) # group samples
png(file="Key_Results_Figures/All_PCA_1v2_AcrossSamples.png", width=1.5*pcvalue1hw,
    height=3*pcvalue2hw, units="px")
par(mar = c(5, 5, 5, 5))
plot(pca_proc33$x[,1], pca_proc33$x[,2], col=mycol, xlab=pcvalue1, ylab=pcvalue2,
     main='All: PC1 v PC2 across samples', cex=1, cex.axis=2,
     cex.lab=2) + grid()
dev.off()

# PC1 vs PC3 #  
png(file="Key_Results_Figures/All_PCA_1v3_AcrossSamples.png", width=2.5*pcvalue1hw,
    height=2.5*pcvalue3hw, units="px")
par(mar = c(5, 5, 5, 5))
plot(pca_proc33$x[,1], pca_proc33$x[,3], col=mycol,  xlab=pcvalue1, ylab=pcvalue3,
     main='All: PC1 v PC3 across samples', cex=1, cex.axis=2,
     cex.lab=2) + grid()
dev.off()

# PC2 vs PC3
png(file="Key_Results_Figures/AllI_PCA_2v3_AcrossGenes.png", width=4*pcvalue2hw,
    height=4*pcvalue3hw, units="px")
par(mar = c(5, 5, 5, 5))
plot(pca_proc33$x[,2], pca_proc33$x[,3], col="red",  xlab=pcvalue2, ylab=pcvalue3,
     main='High PPI: PC3 v PC3 across genes', cex=0.3, cex.axis=2,
     cex.lab=2) + grid()
dev.off()

yy <- data.frame(pca_proc5$x) # genes
hist(yy$PC1) #WTF???
View(rownames(subset(yy, PC1>1))) # 2,447 genes
str(rownames(subset(yy, PC1<1))) # 3,091 genes

yy2 <- data.frame(pca_proc33$x) # genes
hist(yy2$PC1) #WTF???
str(rownames(subset(yy2, PC1>-20))) # 678 genes
subset(yy2, PC1< -20, select=c(PC1, PC2)) # 8 genes

# All_PCA_Viz_AcrossSamples 
outall_1 <- fviz_pca_ind(pca_proc33, col.ind="cos2", repel=T, cex=0.1,
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"))+
  theme(legend.position = "none") + labs(tag="A") + 
  ggtitle("All: PC1 v PC2 across samples")

# All_PCA_Viz_AcrossGenes
outall_2 <- fviz_pca_ind(pca_proc5, col.ind="cos2", repel=T, cex=0.1,
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"))+
  theme(legend.position = "none") + labs(tag="B") + 
  ggtitle("All: PC1 v PC2 across genes")

# plot them
png("Key_Results_Figures/Figure_S4_All_PCA.png", width=1311,height=911,
    bg="white", units="px")
  par(mfrow = c(1, 2))
  multiplot(outall_1, outall_2)
  dev.off()
  
pca_proc333 <- prcomp((new_33)) # across genes # All_PCA_Viz_AcrossSamples 
outall_3 <- fviz_pca_ind(pca_proc333, col.ind="cos2", repel=T, cex=0.1,
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"))+
  theme(legend.position = "none") + labs(tag="C") + 
  ggtitle("All: PC1 v PC2 across samples")
pca_proc55 <- prcomp(t(new_33)) # All_PCA_Viz_AcrossGenes
outall_4 <- fviz_pca_ind(pca_proc55, col.ind="cos2", repel=T, cex=0.1,
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"))+
  theme(legend.position = "none") + labs(tag="D") + 
  ggtitle("All: PC1 v PC2 across genes")
png("Key_Results_Figures/Figure_S4_All_PCA_noEcoli.png", width=1311,height=911,
    bg="white", units="px")
  par(mfrow = c(1, 2))
  multiplot(outall_3, outall_4)
  dev.off()

  # plot all 4 plots  
  png("../Key_Results_Figures/Figure_S4_Both.png", width=1811,height=1411,
    bg="white", units="px")
  par(mfrow = c(2, 2))
  ggarrange(outall_3_3, outall_4_4, outall_3, outall_4, ncol=2, nrow=2)
  dev.off()
  
   par(mfrow = c(1, 1)) 
# fim pap pil
newtemp <- new[,c(720:730,1842:1847)]
dim(newtemp) # 494 samples x 17 genes
newtemp$rows1 <- rowSums(newtemp)
newtemp_3 <- (subset(newtemp, rows1>0))
newtemp_3 <- subset(newtemp_3, select = -c(rows1))
dim(newtemp_3) # 17 samples x 33 genes
write.csv(newtemp_3, "adhesins.csv")

setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021")
#getppis <- data.frame(read.csv("interactions_table_full.csv"))
#rownames(getppis) <- new3$list.files.pattern.......csv...
#dim(getppis)
# fim: getppis[1:4,2395:2417] 
# pap: getppis[1:4,5271:5275]
# pil: getppis[1:4,5584:5605]
#write.csv(getppis[,c(2395:2417,5271:5275, 5585:5606)], "adhesins.interactions.csv")

########################################################################
########### NMF of plasmid-related genes #####################
########################################################################

# Given a u x v matrix A 
# find nonnegative, rank-k matrices W (u x k) and H (k x v) such that A  WH
# column j of A = linear combination of columns of W, with the coefficients
# being column j of H. W  forms a pseudobasis for the columns of A.
# The larger the rank k, the better our approximation
# Typically hope that a good approximation can be achieved with k << v.

# install.packages("NMF")
library(NMF)
install.extras('NMF')

# set memory on Unix
# ulimit  -s 97656 # change memory stack size to 10 times higher 
# R --slave -e 'Cstack_info()["size"]' # check effect
# get: 94999756

# For laptop: ulimit  -s 65532
# R --slave -e 'Cstack_info()["size"]'
# size = 63749529 

new_nmf <- new %>% select_if(colSums(.) > 2)
new_nmf$rows1 <- rowSums(new_nmf)
new_nmf2 <- (subset(new_nmf, rows1>2))
new_nmf3 <- subset(new_nmf2, select = -c(rows1))
dim(new_nmf3) # 331 samples x 1733 genes

sums_all2 <- rowSums(new_nmf3) # group labels
covariates_all <- data.frame(rownames(new_nmf3), sums_all2) # 
png("covariates_all.png", width=2500, height=2600, units="px")
aheatmap(new_nmf3, annCol=covariates_all) # draw heatmap
dev.off()

#estim.r.5 <- nmf(new_nmf3, c(5), nrun=2, seed=123456) # test

estim.r.04.16 <- nmf(new_nmf3, c(4:16), nrun=10, seed=123456)
png("NMF.estim.r.0416.png", width=1100, height=1100, units="px")
plot(estim.r.04.16) # drops when? at 18-19 but also 10?
dev.off()

covariates_all <- data.frame(rownames(new_nmf3), rowSums(new_nmf3)) # 
png("NMF.estim.r.0416.png", width=5500, height=5500, units="px")
consensusmap(estim.r.04.16, annCol=covariates_all, annColors=list(c="blue"),
             labCol="sample") # drops when?  
dev.off()

estim.r.16.27 <- nmf(new_nmf3, c(16:27), nrun=10, seed=123456)
png("NMF.estim.r.1627.png", width=1100, height=1100, units="px")
plot(estim.r.16.27) # drops when? at 18-19 but also 10?
dev.off()

png("NMF.estim.r.1627.png", width=5500, height=5500, units="px")
consensusmap(estim.r.16.27, annCol=covariates_all, annColors=list(c="blue"),
             labCol="sample") # drops when?  
dev.off()
 
# Cophenetic correlation coefficient: repeat NMF several times per rank and
# calculate how similar are the results: how stable are the clusters, give
# that the initial seed is random. Choose the max K before the cophenetic
# coefficient drops.

estim_6_all = nmf(new_nmf3, 6, seed='random') # estimate rank=7
# modeling the matrix X with g genes and s samples
# get product of a matrix h of g gene weights for k factors and a matrix w
# of s sample weights for k factors.
str(estim_6_all)  #  331 samples x 1733 genes

par(mfrow=c(1,1))
png("nmf.basismap.6.png", width=1000, height=3000, units="px")
basismap(estim_6_all, fontsize=11, cexRow=1, cexCol=1, Rowv=F, Colv=F)
dev.off()
pdf("nmf.basismap.6.pdf", width=11, height=25, bg="white")
basismap(estim_6_all, fontsize=11, cexRow=1, cexCol=1, Rowv=F, Colv=F)
dev.off()
png("nmf.coefmap.6.png", width=3000, height=1000, units="px")
coefmap(estim_6_all, fontsize=11, cexRow=1, cexCol=1, Rowv=F, Colv=F)
dev.off()
pdf("nmf.coefmap.6.pdf", width=31, height=5, bg="white")
coefmap(estim_6_all, fontsize=11, cexRow=1, cexCol=1, Rowv=F, Colv=F)
dev.off()

# X is a n x p matrix with rank r
w <- data.frame(basis(estim_6_all)) # basis matrix (sample profiles)
str(w) #  331 samples x 6 dimensions # metasamples? 
h <- data.frame(t(coef(estim_6_all))) # mixture coefficient matrix (gene profiles)
str(h) #  1733 genes x 6 dimensions # mixture coefficient, plasmid gene profiles
write.csv(w, "metasamples.csv")
write.csv(h, "mixture_coefficients.csv")

par(mar = c(4, 4, 4, 4))
pdf("metasamples.each.pdf", width=6, height=10, bg="white")
par(mfrow=c(3,2))
for (i in 1:6){
  print(hist(log10(w[,i]), ylim=c(0,370), breaks=49, xlab=i, main=i))
        grid() }
dev.off()

pdf("mixture_coefficients.each.pdf", width=6, height=10, bg="white")
par(mfrow=c(3,2))
for (i in 1:6){
  print(hist(log10(h[,i]), ylim=c(0,1400), breaks=49, xlab=i, main=i))
        grid() }
dev.off()

length(h[h > 0.001])/6 # average number of genes associated with each rank
 
length(w[w > 0.001])/6 # average number of samples associated with each rank
 

length(subset(w, X1 > 0.001)$X1) # 160 samples 
length(subset(w, X2 > 0.001)$X2) # 71 samples 
length(subset(w, X3 > 0.001)$X3) # 5 samples 
length(subset(w, X4 > 0.001)$X4) # 21 samples 
length(subset(w, X5 > 0.001)$X5) # 136 samples 
length(subset(w, X6 > 0.001)$X6) # 71 samples -> 160 + 71 + 5 + 21 + 136 + 71

new_nmf3_families <- new_nmf3 # analyse families
names1 <- c()
for (j2 in 1:length(rownames(new_nmf3))){  
  names1 <- c(names1, outc[[rownames(new_nmf3)[j2]]][6,1]) }
rownames(new_nmf3_families) <- make.names(names1, unique =T)
estim_6_fam = nmf(new_nmf3_families, 6, seed='random') # estimate rank=7
png("nmf.basismap.6.families.png", width=1000, height=3000, units="px")
basismap(estim_6_fam, fontsize=11, cexRow=1, cexCol=1, Rowv=F, Colv=F, main="")
dev.off()

# X is a n x p matrix with rank r
w2<- data.frame(basis(estim_6_fam)) # basis matrix (sample profiles)
str(w2) #  331 samples x 6 dimensions # metasamples? 
h2 <- data.frame(t(coef(estim_6_fam))) # mixture coefficient matrix (gene profiles)
str(h2) #  1733 genes x 6 dimensions # mixture coefficient, plasmid gene profiles
write.csv(w2, "metasamples.families.csv")
write.csv(h2, "mixture_coefficients.families.csv")

new_nmf3_orders <- new_nmf3 # analyse orders
names2 <- c()
for (j2 in 1:length(rownames(new_nmf3))){  
  names2 <- c(names2, outc[[rownames(new_nmf3)[j2]]][5,1]) }
rownames(new_nmf3_orders) <- make.names(names2, unique =T)
# View(data.frame(rownames(new_nmf3_orders)))
estim_6_orders = nmf(new_nmf3_orders, 6, seed='random') # estimate rank=7
png("nmf.basismap.6.orders.png", width=1000, height=3000, units="px")
basismap(nmf(new_nmf3_orders, 6, seed='random'), fontsize=11, Colv =NA, Rowv=F, main="")
dev.off()

length(subset(h, X1 > 0.001)$X1) # 286 genes 
length(subset(h, X2 > 0.001)$X2) # 584 genes 
length(subset(h, X3 > 0.001)$X3) # 1065 genes 
length(subset(h, X4 > 0.001)$X4) # 482 genes 
length(subset(h, X5 > 0.001)$X5) # 324 genes 
length(subset(h, X6 > 0.001)$X6) # 416 genes  -> 286, 584, 1,065, 482, 324 and 416 genes

p1_nmf <- ggplot(w, aes(x=w$X1, y=w$X2)) +
  geom_point(color ="blue", size=0.3, shape=1, (aes(alpha=0.1)), show.legend=F) +
  scale_x_continuous(name="Rank 1") + scale_y_continuous(name="Rank 2")  + 
  annotation_custom(grobTree(textGrob(paste("For ",length(w$X1),
                                            " bacterial samples",sep=""),
              x=0.22, y=.92, hjust=0, gp=gpar(col="black", fontsize=11)))) 
p11_nmf <- ggExtra::ggMarginal(p1_nmf, type = "densigram", size =6, colour="grey", bins=20)
pdf("NMF_1_2.samples.6.pdf", width=6, height=6, bg="white")
ggarrange(p11_nmf, ncol=1, nrow=1)
dev.off()

p1_nmf2 <- ggplot(h, aes(x=h$X1, y=h$X2)) +
  geom_point(color ="blue", size=0.3, shape=1, (aes(alpha=0.1)), show.legend=F) +
  scale_x_continuous(name="Rank 1") + scale_y_continuous(name="Rank 2")  + 
  annotation_custom(grobTree(textGrob(paste("For ",length(h$X1)," genes",sep=""),
              x=0.22, y=.92, hjust=0, gp=gpar(col="black", fontsize=11))))  
p11_nmf2 <- ggExtra::ggMarginal(p1_nmf2, type="densigram", size =6, colour="grey", bins=20)
pdf("NMF_1_2.genes.6.pdf", width=6, height=6, bg="white")
ggarrange(p11_nmf2, ncol=1, nrow=1)
dev.off()
# citation('NMF')

#################################################################
# install.packages("dendextend")
library(dendextend)
library(taxize)

hcl1 <- hclust(dist(scale(new)))
str(hcl1) # get Enterobacterales group to compare to NMF output
hcl_dist <- as.matrix(dist(scale(new),method="euclidean"))
str(hcl_dist)
# get_ids("Escherichia coli", db="ncbi") # provide vector
# get samples from plasmidPPI
inputc <- rownames(new)
#inputc <- c("Escherichia coli", "Klebsiella pneumoniae")
outc <- classification(inputc, db="ncbi") # provide vector
# outc$`Escherichia coli`$name # Phylum=3 Class=4 Order=5 Family=6
outc$`Buchnera_aphidicola_str._APS_` <- outc$`Sinorhizobium_fredii_NGR234_`
outc$`Escherichia_coli_O157:H7_str._EDL933_` <-  outc$`Escherichia_coli_CFT073_`
outc$`Buchnera_aphidicola_str._APS_`[4,1] <- "Gammaproteobacteria"
outc$`Buchnera_aphidicola_str._APS_`[5,1] <- "Hyphomicrobiales"
outc$`Buchnera_aphidicola_str._APS_`[6,1] <- "Rhizobiaceae"
outc$`Buchnera_aphidicola_str._Sg_` <- outc$`Buchnera_aphidicola_str._APS_`
outc$`Buchnera_aphidicola_str._Bp_` <- outc$`Buchnera_aphidicola_str._APS_`
outc$`Buchnera_aphidicola_str._Ua_` <- outc$`Buchnera_aphidicola_str._APS_`
outc$`Buchnera_aphidicola_str._Ak_` <- outc$`Buchnera_aphidicola_str._APS_`
outc$`Buchnera_aphidicola_str._G002_` <- outc$`Buchnera_aphidicola_str._APS_`
outc$`Pseudomonas_sp._URMO17WK12:I8_` <- outc$`Pseudomonas_aeruginosa_`
outc$`Bacillus_amyloliquefaciens_subsp._plantarum_str._FZB42_` <- outc$`Bacillus_anthracis_str._Ames_`
outc$`Clostridium_botulinum_C-D_str._BKT12695_` <- outc$`Clostridium_acetobutylicum_ATCC_824_`
outc$`[Eubacterium]_eligens_ATCC_27750_` <- outc$`Bacillus_methanolicus_MGA3_`
outc$`[Eubacterium]_eligens_ATCC_27750_`[4,1] <- "Clostridia"
outc$`[Eubacterium]_eligens_ATCC_27750_`[5,1] <- "Eubacteriales"
outc$`[Eubacterium]_eligens_ATCC_27750_`[6,1] <- "Lachnospiraceae"
outc$`Staphylococcus_caprae_M23864:W1_` <-  outc$`Staphylococcus_capitis_C87_`
outc$`Acinetobacter_sp._RUH2624_` <-  outc$`Acinetobacter_baumannii_`
outc$`Acinetobacter_lwoffii_NCTC_5866_=_CIP_64.10_` <-  outc$`Acinetobacter_baumannii_`
outc$`Erwinia_pyrifoliae_Ep1-96_` <- outc$`Erwinia_amylovora_CFBP1430_`
outc[[rownames(new)[180]]] <- outc[[rownames(new)[71]]]

# E coli O157 & Buchnera_aphidicola_str._Sg_ & Buchnera_aphidicola_str._APS_  &
# Bacillus_amyloliquefaciens_subsp._plantarum_str._FZB42_
# not found, get it from NCBI
# Bacteria; Proteobacteria; Gammaproteobacteria; Enterobacterales; Erwiniaceae; Buchnera

labels3 <- c()  # Phylum=3 Class=4 Order=5 Family=6
labels4 <- c()  # Phylum=3 Class=4 Order=5 Family=6
labels5 <- c()  # Phylum=3 Class=4 Order=5 Family=6
labels6 <- c()  # Phylum=3 Class=4 Order=5 Family=6
for (j2 in 1:length(rownames(new))){ # for each classification make labels
  print(j2)
  print(rownames(new)[j2])
  labels3 <- c(labels3, outc[[rownames(new)[j2]]][3,1])  
  labels4 <- c(labels4, outc[[rownames(new)[j2]]][4,1])   # end for  
  labels5 <- c(labels5, outc[[rownames(new)[j2]]][5,1])   # end for  
  labels6 <- c(labels6, outc[[rownames(new)[j2]]][6,1])  } # end for  

#create empty tables to store matching data # Phylum=3 Class=4 Order=5 Family=6
table_hcl3 <- data.frame(row1=character(), col1=character(), dist1=integer())
table_hcl4 <- data.frame(row1=character(), col1=character(), dist1=integer())
table_hcl5 <- data.frame(row1=character(), col1=character(), dist1=integer())
table_hcl6 <- data.frame(row1=character(), col1=character(), dist1=integer())

for (m1 in 1:length(rownames(hcl_dist))){ # sample list 1 - 685
   print(m1)
   for (m2 in 1:length(colnames(hcl_dist))){ # sample list 2
      if((m1 != m2) & (m1 < m2)){ 
       table_hcl3[length(table_hcl3$row1)+1,] <- c(outc[[rownames(new)[m1]]][3,1],
                              outc[[rownames(new)[m2]]][3,1], hcl_dist[m1,m2])
       table_hcl4[length(table_hcl4$row1)+1,] <- c(outc[[rownames(new)[m1]]][4,1],
                              outc[[rownames(new)[m2]]][4,1], hcl_dist[m1,m2])
       table_hcl5[length(table_hcl5$row1)+1,] <- c(outc[[rownames(new)[m1]]][5,1],
                              outc[[rownames(new)[m2]]][5,1], hcl_dist[m1,m2])
       table_hcl6[length(table_hcl6$row1)+1,] <- c(outc[[rownames(new)[m1]]][6,1],
                              outc[[rownames(new)[m2]]][6,1], hcl_dist[m1,m2])  
       } # end if valid
   }  # sample list 2
} # sample list 1
str(table_hcl3) # 121771 obs. of  3 variables:
str(table_hcl4) # 121771 obs. of  3 variables:
str(table_hcl5) # 121771 obs. of  3 variables:
str(table_hcl6) # 121771 obs. of  3 variables:

table_hcl3$dist1 <- as.numeric(table_hcl3$dist1)
table_hcl3$dist2 <- scale(table_hcl3$dist1)
hist((table_hcl3$dist2), breaks = 99) + grid()
table_hcl4$dist1 <- as.numeric(table_hcl4$dist1)
table_hcl4$dist2 <- scale(table_hcl4$dist1)
hist( (table_hcl4$dist2), breaks = 99) + grid()
table_hcl5$dist1 <- as.numeric(table_hcl5$dist1)
table_hcl5$dist2 <- scale(table_hcl5$dist1)
hist((table_hcl5$dist2), breaks = 99) + grid()
table_hcl6$dist1 <- as.numeric(table_hcl6$dist1)
table_hcl6$dist2 <- scale(table_hcl6$dist1)
hist(table_hcl6$dist2, breaks = 99) + grid() # same pattern across all

summary(table_hcl3)

length(unique(c(table_hcl3$row1, table_hcl3$col1))) # how many phyla? 10
length(unique(c(table_hcl4$row1, table_hcl4$col1))) # how many classes? 20
length(unique(c(table_hcl5$row1, table_hcl5$col1))) # how many orders? 46 
length(unique(c(table_hcl6$row1, table_hcl6$col1))) # how many families? 100

View(table_hcl3 %>% group_by(row1) %>% summarise(mean(dist2)))
View(table_hcl4 %>% group_by(row1) %>% summarise(mean(dist2)))
View(table_hcl5 %>% group_by(row1) %>% summarise(mean(dist2)))
View(table_hcl6 %>% group_by(row1) %>% summarise(mean(dist2)))

#library(circlize)
#par(mar = rep(0,4))
#circlize_dendrogram(color_labels(color_branches(dend2, k=3), k=3), cex=0.1,
#                   labels_track_height = NA, dend_track_height=.3) 
# https://cran.r-project.org/web/packages/dendextend/vignettes/FAQ.html

table11 <- c()
head(table_hcl3$dist2)
for (m1 in 1:length(rownames(hcl_dist3))){ # sample list 1 - 685
   print(m1)
   for (m2 in 1:length(colnames(hcl_dist3))){ # sample list 2
      if(m1==m2){ table11[m1] = table11[m1] + hcl_dist3[m2] }
  } # end for
} # end for

# draw phylogeny with families + orders + classes denoted
hcl1$labels <- labels3
dend2 <- as.dendrogram(hcl1)  # Phylum=3 Class=4 Order=5 Family=6
pdf("dendrogram.Phylum.pdf", width=11, height=78, bg="white")
par(mar = c(1, 1, 1, 15))
col2 <- ifelse(grepl("Proteo", labels(dend2)), "red", "blue")
dend2 <- assign_values_to_leaves_edgePar(dend=dend2, value=col2, edgePar="col")
plot(dend2, cex=0.1, cex.axis=0.9, cex.lab=0.9, horiz=T, main="")
dev.off()

hcl1$labels <- labels4
dend2 <- as.dendrogram(hcl1)  #   Class=4 Order=5 Family=6
pdf("dendrogram.Class.pdf", width=13, height=80, bg="white")
par(mar = c(1, 1, 1, 15))
col2 <- ifelse(grepl("Gammap", labels(dend2)), "red", "blue")
dend2 <- assign_values_to_leaves_edgePar(dend=dend2, value=col2, edgePar="col")
plot(dend2, cex=0.1, cex.axis=0.9, cex.lab=0.9, horiz=T, main="")
dev.off()

hcl1$labels <- labels5
dend2 <- as.dendrogram(hcl1) #  Order=5 Family=6
pdf("dendrogram.Order.pdf", width=12, height=99, bg="white")
par(mar = c(1, 1, 1, 25))
col2 <- ifelse(grepl("Enterobacterales", labels(dend2)), "red", "blue")
dend2 <- assign_values_to_leaves_edgePar(dend=dend2, value=col2, edgePar="col")
plot(dend2, cex=0.1, cex.axis=1, cex.lab=1, horiz=T, main="")
dev.off()

hcl1$labels <- labels6
dend2 <- as.dendrogram(hcl1)  #   Family=6
pdf("dendrogram.Family.pdf", width=11, height=80, bg="white")
par(mar = c(1, 1, 1, 15))
col2 <- ifelse(grepl("Enterobacteriaceae", labels(dend2)), "red", "blue")
dend2 <- assign_values_to_leaves_edgePar(dend=dend2, value=col2, edgePar="col")
plot(dend2, cex=0.1, cex.axis=0.8, cex.lab=0.8, horiz=T, main="")
dev.off()

########################################################################
########### LSA of plasmid-related genes #####################
########################################################################

# install.packages("lsa")
library(lsa)
dim(new)  # 497 samples x 2129 genes
mylsa <- lsa(new)

str(mylsa) # $tk = 1-497 samples (left singular vector)
# $dk 1-2129 genes (document vector) - variable data
# and diagnonal matrix $sk 1:360 - 360 components
# Can change Dk to other plasmid gene patterns
sum(mylsa$sk) # 10987.1
head(mylsa$sk) # 24.53281, 18.39363, etc

mycol<- c(rainbow(dim(new)[1], start=1, end=.1)) # group samples
png(file="LSA_1v2_acrosssamples.png")
par(mar = c(5, 5, 5, 5))
plot(mylsa$tk[,1:2], col=mycol, xlab=round(mylsa$sk[1]/sum(mylsa$sk),3), 
     ylab=round(mylsa$sk[2]/sum(mylsa$sk),3),
     main='LSA across samples', cex=0.1, cex.axis=1.5, cex.lab=1.5) + grid()
dev.off()

mylsa_t <- lsa(t(new))
mycol<- c(rainbow(dim(new)[1], start=0.1, end=2)) # group samples
png(file="LSA_1v2_acrossGenes.png")
par(mar = c(5, 5, 5, 5))
plot(mylsa_t$tk[,1:2], col=mycol, xlab=round(mylsa_t$sk[1]/sum(mylsa_t$sk),3), 
     ylab=round(mylsa_t$sk[2]/sum(mylsa_t$sk),3),
     main='LSA across genes', cex=0.1, cex.axis=1.5, cex.lab=1.5) + grid()
dev.off()

#################################################################
##### Plot plasmid and chromosomal genes PPI rates per sample ##
#################################################################

# CFT073 <- data.frame(read.csv("../OUTPUT7/Escherichia_coli_CFT073__genes_data.csv"))
# sum(subset(CFT073, Plasmid==1)$Interactions)/length((subset(CFT073, Plasmid==1)$Gene))
# plasmid proteins' PPIs 33.9
# sum(subset(CFT073, Plasmid==0)$Interactions)/length((subset(CFT073, Plasmid==0)$Gene))
# chromosomal proteins' PPIs 3.9

str(Plasmid_PPI_data) # 4377 obs. of  7 variables
Plasmid_PPI_data <- Plasmid_PPI_data[order(Plasmid_PPI_data$Sample),]
setwd("~/Desktop/OUTPUT7") # _genes_data.csv
for (k in 1:length(list.files())){
     print(k)
     name <- list.files()[k]
     try ( in2 <- data.frame(read.csv(name)) )
     in2 <- in2[,-c(1)]
     name = gsub("\\.(?=[^.]*\\.)", "", name, perl=T)
     name = gsub("\\:", "", name, perl=T)
     name = gsub("_genes_data.csv", "", name, perl=T)
     print(name) 
     for (kk in 1:length(Plasmid_PPI_data$Sample)){
       if(name == Plasmid_PPI_data[kk,]$Sample){
        Plasmid_PPI_data[kk,8] <-as.numeric(sum(subset(in2, Plasmid==1)$Interactions)/
                               length(subset(in2, Plasmid==1)$Gene)) # plasmid
        Plasmid_PPI_data[kk,9] <-as.numeric(sum(subset(in2, Plasmid==0)$Interactions)/
                               length(subset(in2, Plasmid==0)$Gene)) # chrom
        Plasmid_PPI_data[kk,10] <- as.numeric(Plasmid_PPI_data[kk,8])/ 
                                     as.numeric(Plasmid_PPI_data[kk,9])
       }
     } # end for
} # end for
# Plasmid_PPI_data <- Plasmid_PPI_data[,-c(8,9)]
# setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/RMDFILES/")
colnames(Plasmid_PPI_data)[8] <- c("Plasmid_Interations")
colnames(Plasmid_PPI_data)[9] <- c("Chrom_Interations")
colnames(Plasmid_PPI_data)[10] <- c("Ratio_Interations")
str(Plasmid_PPI_data)
View(Plasmid_PPI_data) # 4377 obs. of  10 variables
        # Sample, Plasmid_PPI, Chrom_PPI, Fraction, 
        #  inputProteins, interactions, Plasmid_genes
#write.csv(Plasmid_PPI_data, "Plasmid_PPI_data_full.csv")

# re-try
setwd("~/Desktop/OUTPUT7") # _genes_data.csv

table2 <- data.frame(name=character(), plasmidPPIs=numeric(), chromPPIs=numeric(),
                     plasmidPPIs_genes=numeric(), chromPPIs_genes=numeric(),
                     plasmidplasmidPPIs=numeric(), plasmidchromPPIs=numeric())
names5 <- list.files()
plasmidPPIs <- 0
chromPPIs <- 0
plasmidPPIs_genes <- 0
chromPPIs_genes <- 0
plasmidPPIs_Array <- c()
chromPPIs_Array <- c()

for (k in 1:length(names5)){
     name <-names5[k]
     print(k)
     try ( in2 <- data.frame(read.csv(name)) )
     in2 <- in2[,-c(1)]
     name = gsub("\\.(?=[^.]*\\.)", "", name, perl=T)
     name = gsub("\\:", "", name, perl=T)
     name = gsub("_genes_data.csv", "", name, perl=T)
     table2[k,1] <- name
     plasmidPPIs <- plasmidPPIs + sum(subset(in2, Plasmid>0)$Interactions)
     chromPPIs <-   chromPPIs + sum(subset(in2, Plasmid==0)$Interactions)
     plasmidPPIs_Array <- c(plasmidPPIs_Array, (subset(in2, Plasmid>0)$Interactions))
     chromPPIs_Array <- c(chromPPIs_Array, (subset(in2, Plasmid==0)$Interactions))
     plasmidPPIs_genes <- plasmidPPIs_genes + length(subset(in2, Plasmid>0)$Gene)
     chromPPIs_genes <-   chromPPIs_genes + length(subset(in2, Plasmid==0)$Gene)
     table2[k,2] <- sum(subset(in2, Plasmid>0)$Interactions)
     table2[k,3] <- sum(subset(in2, Plasmid==0)$Interactions) # chrom
     table2[k,4] <- length(subset(in2, Plasmid>0)$Gene)
     table2[k,5] <- length(subset(in2, Plasmid==0)$Gene)
} # end for
table2 <- data.frame(table2)
table2$plasmidPPIs[table2$plasmidPPIs == "NaN"] <- 0
table2$plasmidPPIs <- as.numeric(table2$plasmidPPIs)
table2$chromPPIs <- as.numeric(table2$chromPPIs)
table2$plasmidPPIs_genes <- as.numeric(table2$plasmidPPIs_genes)
table2$chromPPIs_genes <- as.numeric(table2$chromPPIs_genes)
str(table2)

setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/RMD_FILES/")
write.csv(table2, "Plasmid_PPI_rates.csv")

pdf("Plasmid_Gene_Interaction_Protein_Distribution.pdf", width=10, height=6)
par(mfrow=c(1,2))
hist.data = hist( log10(chromPPIs_Array), plot=F, breaks=31)
hist.data$counts = as.numeric(log10(hist.data$counts )) 
 hist.data$counts[is.na(hist.data$counts)] <- 0
  hist.data$counts[is.infinite(hist.data$counts)] <- 0
plot( hist.data,   col="orange", main="",  
     ylab="Number of chromosomal proteins", density =32, cex=1.1,
     xlab="Log10-scaled number of interactions per chromosomal protein")  
hist.data = hist( log10(plasmidPPIs_Array), plot=F, breaks=31)
hist.data$counts = as.numeric(log10(hist.data$counts )) 
 hist.data$counts[is.na(hist.data$counts)] <- 0
  hist.data$counts[is.infinite(hist.data$counts)] <- 0
plot( hist.data,   col="blue", add=T, 
      main="", cex=1.1, ylab="Number of plasmid-related proteins", density =51,
     xlab="Log10-scaled number of PPIs per plasmid-related protein") + grid()
dev.off()

pdf("Fig_4_Plasmid_Gene_Interaction_Rate_Distribution.pdf", width=11, height=5)
par(mfrow=c(1,3))
hist( (table2$plasmidPPIs/table2$plasmidPPIs_genes), breaks=888, col="green", 
      main="", cex.lab=1.3, ylab="Number of plasmid-related proteins", xlim=c(0,250), density =42,
     xlab="Average number of PPIs per plasmid-related protein") + grid()
hist( (table2$chromPPIs/table2$chromPPIs_genes), breaks=91, col="purple", main="",  
     ylab="Number of chromosomal proteins", xlim=c(0,250),  density =42, cex.lab=1.3,
     xlab="Average number of PPIs per chromosomal protein") + grid()
hist.data = hist( log10(chromPPIs_Array), plot=F, breaks=31)
hist.data$counts = as.numeric(log10(hist.data$counts )) 
 hist.data$counts[is.na(hist.data$counts)] <- 0
  hist.data$counts[is.infinite(hist.data$counts)] <- 0
plot( hist.data,   col="purple", main="",  
     ylab="Log10-scaled number of proteins", density =32, cex.lab=1.3,
     xlab="Log10-scaled number of PPIs per protein")  
hist.data = hist( log10(plasmidPPIs_Array), plot=F, breaks=31)
hist.data$counts = as.numeric(log10(hist.data$counts )) 
 hist.data$counts[is.na(hist.data$counts)] <- 0
  hist.data$counts[is.infinite(hist.data$counts)] <- 0
plot( hist.data,   col="green", add=T, cex.lab=1.3, density =51) + grid()
legend("topright", legend=c("Chromosomal proteins", "Plasmid-related proteins"),
       col=c("purple", "green"), fill=c("purple", "green"), cex=1.1, box.lty=0)
dev.off()
write.csv(table2, "Plasmid_PPI_rates.csv")

plasmidPPIs # 1,047,420
chromPPIs   # 600,858,987
plasmidPPIs_genes #  13,359
chromPPIs_genes   # 15,622,015
plasmidPPIs/plasmidPPIs_genes # 78.4
chromPPIs/chromPPIs_genes  # 38.5
summary(plasmidPPIs_Array) # median = 32 , sd=506.6
sd(plasmidPPIs_Array)  # PPIs per protein
hist(log10(plasmidPPIs_Array), breaks=111) + grid()
summary(chromPPIs_Array) # median = 15 , sd=284.5
sd(chromPPIs_Array)

# need output8/

# re-try
setwd("~/Desktop/OUTPUT8/") # _genes_data.csv

table3 <- data.frame(name=character(), plasmidPPIs=numeric(), plasmidchromPPIs=numeric(),
                     plasmidplasmidPPIs=numeric(), chromPPIs=numeric(),
                     plasmidPPIs_genes=numeric(), plasmidchromPPIs_genes=numeric(),
                     plasmidplasmidPPIs_genes=numeric(), chromPPIs_genes=numeric())
names6 <- list.files(pattern="chrom_data.csv")
str(names6)
plasmidPPIs <- 0
plasmidchromPPIs <- 0
plasmidplasmidPPIs <- 0
chromPPIs <- 0
plasmidPPIs_genes <- 0
plasmidchromPPIs_genes <- 0
plasmidplasmidPPIs_genes <- 0
chromPPIs_genes <- 0
plasmidPPIs_Array <- c()
chromPPIs_Array <- c()
plasmidchromPPIs_Array <- c()
plasmidplasmidPPIs_Array <- c()

for (k in 2509:length(names6)){ # check 796-798
     name <-names6[k]
     print(k)
     try ( in2 <- data.frame(read.csv(name)) )
     in2 <- in2[,-c(1)]
     name = gsub("\\.(?=[^.]*\\.)", "", name, perl=T)
     name = gsub("\\:", "", name, perl=T)
     name = gsub("_chrom_data.csv", "", name, perl=T)
     table3[k,1] <- name
     str(in2) # from to score from_ID to_ID c
     plasmidPPIs <- plasmidPPIs + length(subset(in2, c>0)$c)
     plasmidchromPPIs <- plasmidchromPPIs + length(subset(in2, c==1)$c)
     plasmidplasmidPPIs <- plasmidplasmidPPIs + length(subset(in2, c==2)$c)
     chromPPIs <-   chromPPIs + length(subset(in2, c==0)$c)
     plasmidPPIs_Array <- c(plasmidPPIs_Array, length(subset(in2, c>0)$c))
     plasmidchromPPIs_Array <- c(plasmidchromPPIs_Array, length(subset(in2, c==1)$c))
     plasmidplasmidPPIs_Array <- c(plasmidPPIs_Array, length(subset(in2, c==2)$c))
     chromPPIs_Array <- c(chromPPIs_Array, length(subset(in2, c==0)$c))
     plasmidPPIs_genes <- c(unique(sort(c(plasmidPPIs_genes, subset(in2, c>0)$from_ID))))
     plasmidchromPPIs_genes <- c(unique(sort(c(plasmidchromPPIs_genes, subset(in2, c==1)$from_ID))))
     plasmidplasmidPPIs_genes <- c(unique(sort(c(plasmidplasmidPPIs_genes, subset(in2, c==2)$from_ID))))
     chromPPIs_genes <- c(unique(sort(c(chromPPIs_genes, subset(in2, c==0)$from_ID))))
     table3[k,2] <- length(subset(in2, c>0)$c) # p-p or p-c PPI
     table3[k,3] <- length(subset(in2, c==1)$c) # p-c PPI
     table3[k,4] <- length(subset(in2, c==2)$c) # p-p PPI
     table3[k,5] <- length(subset(in2, c==0)$c) # c-c PPI
     table3[k,6] <- length(unique(sort(subset(in2, c>0)$from_ID))) # plasmid genes
     table3[k,7] <- length(unique(sort(subset(in2, c==1)$from_ID))) 
     table3[k,8] <- length(unique(sort(subset(in2, c==2)$from_ID)))
     table3[k,9] <- length(unique(sort(subset(in2, c==0)$from_ID))) # chrom genes
} # end for
table3 <- data.frame(table3)
table3$plasmidPPIs[table3$plasmidPPIs == "NaN"] <- 0
table3$plasmidchromPPIs[table3$plasmidchromPPIs == "NaN"] <- 0
table3$plasmidplasmidPPIs[table3$plasmidplasmidPPIs == "NaN"] <- 0
table3$chromPPIs[table3$chromPPIs == "NaN"] <- 0
table3$plasmidPPIs_genes[table3$plasmidPPIs_genes == "NaN"] <- 0
table3$plasmidchromPPIs_genes[table3$plasmidchromPPIs_genes == "NaN"] <- 0
table3$plasmidplasmidPPIs_genes[table3$plasmidplasmidPPIs_genes == "NaN"] <- 0
table3$chromPPIs_genes[table3$chromPPIs_genes == "NaN"] <- 0
table3$plasmidPPIs <- as.numeric(table3$plasmidPPIs)
table3$plasmidchromPPIs <- as.numeric(table3$plasmidchromPPIs)
table3$plasmidplasmidPPIs <- as.numeric(table3$plasmidplasmidPPIs)
table3$chromPPIs <- as.numeric(table3$chromPPIs)
table3$plasmidPPIs_genes <- as.numeric(table3$plasmidPPIs_genes)
table3$plasmidchromPPIs_genes <- as.numeric(table3$plasmidchromPPIs_genes)
table3$plasmidplasmidPPIs_genes <- as.numeric(table3$plasmidplasmidPPIs_genes)
table3$chromPPIs_genes <- as.numeric(table3$chromPPIs_genes)
str(table3)

# p + q vs p2 + 2pq + q2
# Given f(p) = 0.05 then f(!p) = 0.95
# Expect p-c PPIs = 2pq with p-p PPI = p2
# Deficiency in observed p-c PPIs compared to expected
# if F<0 then excess of p-c PPIs, if F>0 then deficit
table33$f <- c(0)
table33$fp2 <- c(0)
table33$fq2 <- c(0)
fp <- 0 # f(p) freq plasmid PPIs
fq <- 0 # f(q) freq chrom PPIs

for (k in 1:length(table33$plasmidPPIs)){ # F = 1 - obs/exp
  fp <- table33[k,6]/(table33[k,6]+table33[k,9])    # f(p)  0.008
  fq <- table33[k,9]/(table33[k,6]+table33[k,9])    # f(q)  0.991
  table33$fp2[k] <- 1 - (table33[k,4]/(table33[k,3]+table33[k,4]+table33[k,5]))/(fp*fp) # p-p PPI
  table33$fq2[k] <- 1 - (table33[k,5]/(table33[k,3]+table33[k,4]+table33[k,5]))/(fq*fq) # c-c PPI
  table33$f[k]   <- 1 - (table33[k,3]/(table33[k,3]+table33[k,4]+table33[k,5]))/(2*fp*fq) # p-c PPI
} # 1 - exp/obs

setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/Key_Results_Figures/")
table44 <- data.frame(read.xlsx("Table_S3_new.xlsx"))
str(table44)
table44$Plasmid_genes <- as.numeric(table44$Plasmid_genes )
table33 <- subset(table44, Plasmid_Plasmid_PPI!=0)
table33$fq2 <- as.numeric(table33$F_cc_PPI )
table33$fp2 <- as.numeric(table33$F_pp_PPI )
table33$f <- as.numeric(table33$F_pc_PPI )
str(table33)

hist(table33$f, breaks=41, col="pink", xlab="Plasmid-chromosome PPI F value per sample",
     main="", ylab="Number of samples") + grid()
hist(table33$fp2, breaks=41, col="blue", xlab="Plasmid-plasmid F value per sample",
     main="", ylab="Number of samples") + grid()
hist(table33$fq2, breaks=41, col="grey", xlab="Chromosome-chromosome F value per sample",
     main="", ylab="Number of samples") + grid()

setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/RMD_FILES/")
pdf("Fvalues.S24.2.pdf", width=6, height=6)
plot(table33$fp2, table33$f, col="black", pch=1, cex=0.2, xlab="Plasmid-plasmid PPI F value per sample",
     ylab="Plasmid-chromosome PPI F value per sample", xlim=c(-59,1), ylim=c(-59,1)) + grid()
dev.off()

gg1 <- ggExtra::ggMarginal(ggplot(table33, aes(x= table33$fp2, y=table33$f)) +
  geom_point(color ="darkred", size=0.2, shape=1, (aes(alpha=0.2)), show.legend=F) +
  scale_x_continuous(name="Plasmid-plasmid PPI F value per sample", limits =c(-59,1)) +
  scale_y_continuous(name="Plasmid-chromosome PPI F value per sample", limits =c(-4,1)),
  type="densigram",size=5,colour="darkred", bins=60)    
gg2 <- ggExtra::ggMarginal(ggplot(table33, aes(x= table33$fq2, y=table33$f)) +
  geom_point(color ="darkgreen", size=0.2, shape=1, (aes(alpha=0.2)), show.legend=F) +
  scale_x_continuous(name="Chromosomal PPI F value per sample", limits =c(-1,1)) +
  scale_y_continuous(name="Plasmid-chromosome PPI F value per sample", limits =c(-4,1)),
  type="densigram",size=5,colour="darkgreen", bins=60)     
pdf("Fvalues.S24.3.pdf", width=10, height=6)
ggarrange(gg1, gg2, ncol=2, nrow=1)
dev.off()

#plot(table33$f, table33$fq2, col="black", pch=1, cex=0.2, xlab="Plasmid-chromosome F value per sample",
 #    ylab="Chromosomal PPI F value per sample", xlim=c(0,1), ylim=c(0,1)) + grid()

pdf("Fvalues.S24.pdf", width=7, height=6)
c1 <- rgb(113,116,255, max = 255, alpha = 50, names = "lt.blue")
c2 <- rgb(255,112,103, max = 255, alpha = 50, names = "lt.pink")
c3 <- rgb(110,255,103, max = 255, alpha = 50, names = "lt.green")
hist(table33$f, breaks=71, col=c1, xlab="PPI F value per sample",
     main="", ylab="Number of samples", xlim=c(-59,1), ylim=c(0,40)) + grid()
hist(table33$fp2, breaks=771, col=c2, add=T)
#hist(table33$fq2, breaks=71, col=c3, add=T)
legend('topleft',c('Plasmid-chromosome PPIs','Plasmid-plasmid PPIs'), pch=1,
       col=c("blue", "red"), cex=0.8, border = NA)
text(0.25,100, "Excess of PPIs", col=alpha("black",0.7))
text(0.75,110, "Deficit of PPIs", col=alpha("black",0.7))
dev.off()

summary(table33$f)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.4863  0.8877  0.9127  0.8981  0.9355  0.9627 
sd(table33$f) # 0.898+-0.069
summary(table33$fp2)
# -4.0721  0.9428  0.9796  0.7295  1.0000  1.0000 
sd(table33$fp2) # 0.730+-0.94
summary(table33$fq2)
# -0.66942 -0.12938 -0.04916 -0.10112 -0.01830 -0.00181 
sd(table33$fq2) # -0.101+-0.130

plasmid_freq <- (table33[,6]/(table33[,6]+table33[,9]))
hets <- (table33[,3]/(table33[,3]+table33[,4]+table33[,5]))
ppp <- seq(0,1,0.001)
homs <- (table33[,4]/(table33[,3]+table33[,4]+table33[,5]))
plot(plasmid_freq, hets, col=alpha("black", 0.4), pch=1, cex=0.3, xlab="Plasmid-related protein frequency",
     ylab="Plasmid-chromosome PPI frequency") + grid()
abline(lm( hets ~ plasmid_freq), col="black", lty="dashed", lwd=1)
lines(ppp, 2*ppp*(1-ppp), col="grey", lty="dotted", lwd=2)
points(plasmid_freq, homs, col=alpha("blue",0.4), pch=1, cex=0.3)
abline(lm( homs ~ plasmid_freq), col="blue", lty="dashed", lwd=2)
lines(ppp, ppp*ppp, col="blue", lty="dotted", lwd=2)
legend("topleft", legend=c("Het expected rate", "Het observed rate", "Hom expected rate",
                           "Hom observed rate"),
       col=c("black", "grey", "blue", "blue"), lty=c("dashed","dotted"), lwd=2, cex=1)


#######
setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/RMD_FILES/OUTPUT7")
table22 <- data.frame(name=character(), plasmidPPIs=numeric(), chromPPIs=numeric(),
                     plasmidPPIs_genes=numeric(), chromPPIs_genes=numeric())
names5 <- list.files()
names497 <- subset(table2, table2$plasmidPPIs>0)$name # 497
plasmidPPIs <- 0
chromPPIs <- 0
plasmidPPIs_genes <- 0
chromPPIs_genes <- 0
counter <- 0
plasmidPPIs_Array <- c()
chromPPIs_Array <- c()

for (k in 1:length(names5)){
     name <-names5[k]
     try ( in2 <- data.frame(read.csv(name)) )
     in2 <- in2[,-c(1)]
     name = gsub("\\.(?=[^.]*\\.)", "", name, perl=T)
     name = gsub("\\:", "", name, perl=T)
     name = gsub("_genes_data.csv", "", name, perl=T)
     if(name %in% names497) { 
       counter = counter  + 1
       table2[k,1] <- name
     chromPPIs <-   chromPPIs + sum(subset(in2, Plasmid==0)$Interactions)
     chromPPIs_genes <-   chromPPIs_genes + length(subset(in2, Plasmid==0)$Gene)
     table22[k,3] <- sum(subset(in2, Plasmid==0)$Interactions) # chrom
     table22[k,5] <- length(subset(in2, Plasmid==0)$Gene) }
} # end for
print(counter)
table22 <- data.frame(table22)
table22$chromPPIs <- as.numeric(table22$chromPPIs)
table22$chromPPIs_genes <- as.numeric(table22$chromPPIs_genes)
str(table22)

chromPPIs   # 65,310,604
chromPPIs_genes   # 1,796,906
chromPPIs/chromPPIs_genes  # 38.5
summary(subset(table22, table22$plasmidPPIs >0 )$chromPPIs/
          subset(table22, table22$plasmidPPIs >0 )$chromPPIs_genes)
#    Min.   1st Qu.    Median      Mean   3rd Qu.      Max.  per sample
#    0.00093  25.88738  30.79655  32.25402  37.31209 162.54293 

pdf("Fig_S15_Plasmid_Gene_Interaction_Rate_Distribution.2.pdf", width=10, height=6)
par(mfrow=c(1,1))
hist( (table22$plasmidPPIs/table22$plasmidPPIs_genes), breaks=888, col="green",
      main="", cex=1.1, ylab="Number of plasmid-related proteins",xlim=c(0,250),   density =42,
     xlab="Average number of PPIs per plasmid-related protein") + grid()
dev.off()
 
# #################################################################
# RcppML::NMF for large sparse matrices: # doesn't work!!
# see also rsparse::WRMF
# https://github.com/zdebruine/RcppML
# install.packages("RcppML")
# library(RcppML)
# model <- RcppML::nmf(plasmidPPI_2, k = 10)
# str(model)
#################################################################

# look at interactions per protein per sample and chr/plasmid context
# Not possible, need more memory

setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/RMD_FILES/")
#Plasmid_PPI_data <- read.xlsx("FinalTable.xlsx")
#str(Plasmid_PPI_data) # 4377 obs. of  7 variables

setwd("~/Desktop/OUTPUT7")
# 4430 samples as Gene, Plasmid, Interactions
# "Escherichia_coli_str_K-12_substr_MG1655__genes_data.csv"

# samples 4,399 => 4 x 11 -> 100 times
p_dist <- c()
t_dist <- c()
names7 <- list.files(pattern="_genes_data.csv")

par(mfrow=c(1,1))
for (i in 1:length(names7)){ # get genes and plasmids
   name <- names7[i] 
   test1 <- read.csv(name) # read in samples  
   test1 <- test1[-c(1)] # remove extra column
   name <- str_replace_all(name, "_genes_data.csv", "") # remove
   print(i)
   
   if((length(subset(test1, Plasmid==0)$Interactions)>2) & 
     ((length(subset(test1, Plasmid==1)$Interactions)>2)) ){ # if ok
    p_dist <- c(p_dist, t.test(subset(test1, Plasmid==0)$Interactions,
            subset(test1, Plasmid==1)$Interactions)$p.value )
    t_dist <- c(t_dist, t.test(subset(test1, Plasmid==0)$Interactions,
            subset(test1, Plasmid==1)$Interactions)$statistic ) 
   
   p <- ggplot(test1, aes(x=factor(Plasmid),
                    y=log10(Interactions), color=Plasmid)) + 
    geom_violin(draw_quantiles=T, show.legend=F) +  
    labs(title=name, x="Genomic context", y="Log10-scaled number of interactions per gene")+
      scale_x_discrete(labels = c('Chromosomal','Plasmid')) +
    geom_jitter(shape=1, size=0.4, aes(alpha=0.1), position=position_jitter(0.3),
                show.legend=F) +
    stat_summary(fun.y=median, geom="point", size=7, color=c("black","blue"))
    pdf(paste0( name,".pdf",sep=""), width=4, height=7, bg="white")
      print(p) # add to list
    dev.off() # save file
    print(name) } # end if ok  
} # end for samples

p_dist2 <- p.adjust(p_dist, method="BH")
summary(p_dist2)
length(p_dist2[p_dist2<0.05]) # 65 out of 345
#   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#0.00000 0.09705 0.38113 0.38562 0.63053 0.99870 
summary(t_dist)
## Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
#-15.46954  -2.26582  -1.12046  -1.10578  -0.05381  13.79977 

pdf("FigS16.pdf", width=9, height=6, units="px")
par(mfrow=c(2,1))
hist( (p_dist2), breaks=55, main="", col="purple",  
 xlab="Difference in chrom vs plasmid interactions log10-scaled p value distribution")+
  grid()
hist(scale(t_dist), breaks=55, main="", col="darkgreen",  
  xlab="Difference in chrom vs plasmid interactions t-stat distribution") + grid()
dev.off()

################

gene_list <- c()
plasmid_status <- c()
new.env(hash=TRUE)
detach("package:STRINGdb", unload=TRUE)
library(hash)
genes <- hash()

for (i in 1:11){ # length(list.files())){ # get genes and plasmids
   test1 <- read.csv(list.files()[i]) # read in sample
   test1 <- test1[-c(1)] # remove extra column
   for (k in 1:length(test1$Gene)){ # get the genes and contexts
     genes[[test1$Gene[k]]] = test1$Plasmid[k]
     gene_list <- unique(c(gene_list, test1$Gene[k])) }
} # end for samples
#View(genes)
#genes[["ubix"]]
gene_list <- data.frame(gene_list)
str(gene_list)
View(gene_list)
write.csv(gene_list, "gene_list.csv")

# for(i in 1:length(gene_list)){ print(genes[[gene_list[i]]]) } # end for 

genelistout <- as.data.frame(matrix(0, ncol=length(gene_list$gene_list), #
                              nrow =length(list.files()))) # 4395 samples 
colnames(genelistout) <- gene_list$gene_list
gene_list2 <- str_replace_all(gene_list$gene_list, "_genes_data.csv", "") # remove
genelistout <- data.frame(genelistout)
rownames(genelistout) <- gene_list2
head(str(genelistout))


for (i in 1:2){ # length(list.files())){ # for each sample
   test1 <- read.csv(list.files()[i]) # read in sample
   test1 <- test1[-c(1)] # remove extra column, Gene, Plasmid, Interactions
   
   for (k in 1:length(colnames(genelistout))){ # for each possible gene
     if(colnames(genelistout)[k] == test1$Gene){ # if matching gene 
         genelistout[1,k] <- test1$Interaction
     } # end if matching gene
   } # end for genes
} # end for samples

plasmid_data <- c()
for (k in 1:length(colnames(genelistout))){ # for each
    plasmid_data <- c(plasmid_data, genes[[colnames(genelistout)[k]]])
} # end for plasmids

# add plasmid context
genelistout2 <- rbind(plasmid_data, genelistout)

########################################################################
########### all proteins per sample to get PPI info #####################
########################################################################

setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/OUTPUT7/")
#nn <- data.frame(read.csv("Escherichia_coli_O157_H7_str._EDL933__genes_data.csv"))
#nn <- nn[,-c(1)]
#str(nn)
# X,"Gene","Plasmid","Interactions"
# "1","aaea",1,4
# "2","aaeb",1,19

gene_list_all <- c()
gene_list_plasmid <- c()
gene_list_chr <- c()
gene_list_plasmid_only <- c()

files1 <- list.files(pattern = "\\.csv$")
for (k in 1:length(files1)){
     name <- files1[k]  # 4393 samples
     print(k)
     try ( in2 <- data.frame(read.csv(name)) )
     in2 <- in2[,-c(1)]
     name = gsub("\\.(?=[^.]*\\.)", "", name, perl=T)
     name = gsub("\\:", "", name, perl=T)
     name = gsub("_genes_data.csv", "", name, perl=T)

     gene_list_all <- unique(sort(c(gene_list_all, in2$Gene))) # get gene names
     gene_list_plasmid <- unique(sort(c(gene_list_plasmid, subset(in2, Plasmid==2)$Gene))) 
     gene_list_plasmid <- unique(sort(c(gene_list_plasmid, subset(in2, Plasmid==1)$Gene))) 
     gene_list_chr <- unique(sort(c(gene_list_plasmid, subset(in2, Plasmid==0)$Gene)))
} # end for

gene_list_all <- data.frame(unique(sort(gene_list_all)))
gene_list_all <- gene_list_all[-c(1:78),] # remove incorrect rows
str(gene_list_all) # 9,551,828 genes
# write.csv(gene_list_all, "../gene_list_all.csv")
gene_list_all <- read.csv("../gene_list_all.csv")

########################################################################
########### NMF of plasmid-related PPI rates per gene/sample #####################
########################################################################

# NMF part #
library(NMF)
install.extras('NMF')
setwd("/Volumes/GoogleDrive/My Drive/TDA_Summer2021/RMD_FILES/")

int_nmf0 <- new_int # %>% mutate(across(everything(), .fns = ~replace_na(.,0))) 
int_nmf <- (int_nmf0) %>% select_if(colSums(.) > 30)  
dim(int_nmf) # 494 2363
 int_nmf[is.na(int_nmf)] <- 0
summary(colSums(int_nmf))
#View(data.frame(colnames(int_nmf0), (colSums(int_nmf0))))
#View(data.frame(rownames(int_nmf0), (rowSums(int_nmf0))))
int_nmf$rows1 <- rowSums(int_nmf)
int_nmf2 <- (subset(int_nmf, rows1>10))
int_nmf3 <- subset(int_nmf2, select = -c(rows1))
summary(rowSums(int_nmf3))
int_nmf3[is.na(int_nmf3)] <- 0

#View(data.frame(colnames(int_nmf3), (colSums(int_nmf3))))
#View(data.frame(rownames(int_nmf3), (rowSums(int_nmf3))))
dim(int_nmf3) # 481 2363
int_nmf4 <- dplyr::bind_rows(int_nmf3)

dim(int_nmf4) # 481 2363
str(int_nmf4) # 481 2363
type(int_nmf4) # 481 2363

png("covariates_all.int.png", width=2500, height=2600, units="px")
aheatmap(int_nmf3, annCol=data.frame(rownames(int_nmf3),rowSums(int_nmf3))) # draw heatmap
dev.off()

# estim.r.5 <- nmf(int_nmf3, c(5), nrun=2, seed=123456)

estim.r.0515 <- nmf(int_nmf3, c(5:15), nrun=10, seed=123456)
png("NMF.estim.r.0515.interactions.all.png", width=1100, height=1100, units="px")
plot(estim.r.0515) # drops when?
dev.off()

estim.r.1725 <- nmf(int_nmf3, c(17:25), nrun=10, seed=123456)
png("NMF.estim.r.1725.interactions.all.png", width=1100, height=1100, units="px")
plot(estim.r.1725) # drops when?
dev.off()

# Cophenetic correlation coefficient: repeat NMF several times per rank and
# perform 2-rank NMF using default algorithm
estim_all = nmf(int_nmf3, 7, seed='random') # estimate rank=33
# modeling the matrix interactions_table3 with g genes and s samples
# get product of a matrix h of g gene weights for k factors and a matrix w
# of s sample weights for k factors.
str(estim_all)  #  XX samples x XX genes  

png("NMF.estim_6.interactions.all.png", width=2500, height=2600, units="px")
plot(estim_all)
dev.off()
png("nmf.basismap.6.interactions.all.png", width=12600, height=36500, units="px")
basismap(estim_all, fontsize=20, cexRow=2, cexCol=2)
dev.off()
pdf("nmf.basismap.6.interactions.all.pdf", width=56, height=156, bg="white")
basismap(estim_all, fontsize=20, cexRow=2, cexCol=2)
dev.off()
png("nmf.coefmap.6.interactions.all.png", width=36500, height=12600, units="px")
coefmap(estim_all, fontsize=20, cexRow=2, cexCol=2)
dev.off()
pdf("nmf.coefmap.29.interactions.all.pdf", width=156, height=56, bg="white")
coefmap(estim_all, fontsize=20, cexRow=2, cexCol=2)
dev.off()

# X is a n x p matrix with rank r=50
w_all <- data.frame(basis(estim_all)) # basis matrix (sample profiles)
str(w_all) # XX samples x 6 dimensions # metasamples? 
h_all <- data.frame(t(coef(estim_all))) # mixture coefficient matrix (gene profiles)
str(h_all) #  XX genes x 6 dimensions # mixture coefficient, plasmid gene profiles
write.csv(w_all, "metasamples.interactions.all.csv")
write.csv(h_all, "mixture_coefficients.interactions.all.csv")
hist(log10(w_all$X6), xlab="metasamples for PPIs all", breaks=49) + grid()

length(subset(w_all, X1 > 19999)$X1) # 5 samples
length(subset(w_all, X2 > 19999)$X2) # 482 samples
length(subset(w_all, X3 > 19999)$X3) # 10 samples
length(subset(w_all, X4 > 19999)$X4) # 50 samples
length(subset(w_all, X5 > 19999)$X5) # 13 samples
length(subset(w_all, X6 > 19999)$X6) # 78 samples
length(subset(w_all, X7 > 19999)$X7) # 5 samples
length(w_all[w_all > 19999])/7 # 91.9 average number of   associated with each rank

hist(log10(h_all$X6), xlab="metagenes for PPIs all", breaks=49) + grid()

length(subset(h_all, X1 > 0.00001)$X1) # 238 genes
length(subset(h_all, X2 > 0.00001)$X2) # 2993 genes
length(subset(h_all, X3 > 0.00001)$X3) # 183 genes
length(subset(h_all, X4 > 0.00001)$X4) # 374 genes
length(subset(h_all, X5 > 0.00001)$X5) # 909 genes
length(subset(h_all, X6 > 0.00001)$X6) # 387 genes
length(subset(h_all, X7 > 0.00001)$X7) # 152 genes

length(h_all[h_all > 0.00001])/7 # 748 average number of genes associated with each rank

pdf("metasamples.interactions.pdf", width=6, height=11, bg="white")
par(mfrow=c(3,3))
for (i in 1:7){
  print(hist(log10(w_all[,i]), ylim=c(0,500), breaks=49, xlab=i, main=i))
        grid() }
dev.off()

pdf("mixture_coefficients.interactions.pdf", width=6, height=11, bg="white")
par(mfrow=c(3,3))
for (i in 1:7){
  print(hist(log10(h_all[,i]), ylim=c(0,2600), breaks=49, xlab=i, main=i))
        grid() }
dev.off()
# #################################################################
# Compare NMF outputs

input4 <- data.frame(read.csv("Data_for_Fig_S11_metasamples.interactions.all.csv"))
# mixtures gene coefficients from NMF on plasmid gene presence-absence matrix
length(subset(input4, X6 > 0.01)$X) # 45 samples
length(subset(input4, X14 > 0.01)$X) # 130 samples
#length(subset(input4, X14 > 0.01 & X6 > 0.01)$X14) # 15 samples

length(intersect(subset(input2, X6 > 0.01)$X, subset(input4, X6 > 0.01)$X)) 
# 200 vs 45 => 25 intersection
25 / (200 + 45 - 25)  # 0.114
length(intersect(subset(input2, X14 > 0.01)$X, subset(input4, X6 > 0.01)$X)) 
# 130 vs 45 => 7 intersection
7 / (130 + 45 - 7)    # 0.042
length(intersect(subset(input2, X6 > 0.01)$X, subset(input4, X14 > 0.01)$X)) 
# 200 vs 130 => 42 intersection
42 / (200 + 130 - 42) # 0.146
length(intersect(subset(input2, X14 > 0.01)$X, subset(input4, X14 > 0.01)$X)) 
# 130 vs 130 => 42 intersection
42 / (130 + 130 - 42) # 0.193

mix2 <- data.frame(read.csv("Data_for_Fig_S8_mixture_coefficients.interactions.all.csv"))
# metasamples from NMF on plasmid gene PPI rates matrix
hist(log10(mix2$X6)) + grid()
length(subset(mix2, X6 > 0.00001)$X6) # 4,004 genes
length(subset(mix2, X14 > 0.00001)$X14) # 3,639 genes
length(intersect(subset(mix2, X6 > 0.00001)$X, subset(mix2, X14 > 0.00001)$X)) 
# 4004 vs 1288 => 2075 intersection
2075 / (4004 + 1288 - 2075)  # 0.645

mix4 <- data.frame(read.csv("Data_for_Fig_S12_mixture_coefficients.interactions.all.csv"))
# mixtures gene coefficients from NMF on plasmid gene PPI rates matrix
hist(log10(mix4$X6)) + grid()
length(subset(mix4, X6 > 0.00001)$X6) # 1288 genes

length(intersect(subset(mix2, X6 > 0.00001)$X, subset(mix4, X6 > 0.00001)$X)) 
# 4004 vs 1288 => 748 intersection
748 / (4004 + 1288 - 748)  # 0.165
length(intersect(subset(mix2, X14 > 0.00001)$X, subset(mix4, X6 > 0.00001)$X)) 
# 3639 vs 1288 => 690 intersection
690 / (3639 + 1288 - 690)  # 0.163

library(xlsx)
input3 <- read.xlsx("Key_Results_Figures/Table_S3.xlsx", sheetIndex=1)
str(input3)
length(subset(input3, Plasmid_genes==0)$Sample) # 1235 have no genes
length(subset(input3, Plasmid_PPI==0)$Sample) # 1231 have no interactions
check <- subset(input3, Plasmid_genes==0 & Plasmid_PPI>0) # 11 shit
View(check)
View(subset(input3, Plasmid_genes<20 & Plasmid_genes>1 & Plasmid_PPI>0))
# #################################################################
```
